<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>anacore.fusion &mdash; AnaCore  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/anacore.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/sphinx_highlight.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> AnaCore
            <img src="../../_static/anacore_logo_40deg.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../anacore.html">AnaCore package</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">AnaCore</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="../index.html">Module code</a></li>
      <li class="breadcrumb-item active">anacore.fusion</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for anacore.fusion</h1><div class="highlight"><pre>
<span></span><span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Classes and functions for reading/writing fusions callers&#39;s output and VCF containing fusions. Each record is converted to VCFRecord.</span>

<span class="sd">:Example:</span>

<span class="sd">    Read results form STAR-Fusion</span>

<span class="sd">    .. highlight:: python</span>
<span class="sd">    .. code-block:: python</span>

<span class="sd">        # File content&gt;</span>
<span class="sd">        # #FusionName	JunctionReadCount	SpanningFragCount	SpliceType	LeftGene	LeftBreakpoint	RightGene	RightBreakpoint	LargeAnchorSupport	FFPM	LeftBreakDinuc	LeftBreakEntropy	RightBreakDinuc	RightBreakEntropy	annots</span>
<span class="sd">        # MN1--BEND2	90	39	ONLY_REF_SPLICE	MN1^ENSG00000169184.6	chr22:27796763:-	BEND2^ENSG00000177324.14	chrX:18195442:-	YES_LDAS	38.9246	GT	1.9656	AG	1.7232	[&quot;INTERCHROMOSOMAL[chr22--chr9]&quot;]</span>

<span class="sd">        from anacore.fusion import FusionFileReader</span>

<span class="sd">        spl_name = &quot;spl1&quot;</span>
<span class="sd">        print(&quot;Breakpoint_1\\tGene_brkpt_1\\tBreakpoint_2\\tGene_brkpt_2\\tSpanning_pair\\tSplit_read&quot;)</span>
<span class="sd">        with FusionFileReader.factory(&quot;test.tsv&quot;, sample_name=spl_name, annot_field=&quot;ANN&quot;) as reader:</span>
<span class="sd">            for first, second in reader:</span>
<span class="sd">                print(</span>
<span class="sd">                    &quot;{}:{}&quot;.format(first.chrom, frist.pos),</span>
<span class="sd">                    &quot;,&quot;.join([elt[&quot;SYMBOL&quot;] for elt in first.info[&quot;ANN&quot;]]),</span>
<span class="sd">                    &quot;{}:{}&quot;.format(second.chrom, second.pos),</span>
<span class="sd">                    &quot;,&quot;.join([elt[&quot;SYMBOL&quot;] for elt in second.info[&quot;ANN&quot;]]),</span>
<span class="sd">                    first.samples[spl_name][&quot;PR&quot;],</span>
<span class="sd">                    first.samples[spl_name][&quot;SR&quot;],</span>
<span class="sd">                    sep=&quot;\\t&quot;</span>
<span class="sd">                )</span>

<span class="sd">        # Result&gt;</span>
<span class="sd">        # Breakpoint_1\tGene_brkpt_1\tBreakpoint_2\tGene_brkpt_2\tSpanning_pair\tSplit_read</span>
<span class="sd">        # chr22:27796763\tMN1\tchrX:18195442\tBEND2\t39\t99</span>

<span class="sd">    Read VCF</span>

<span class="sd">    .. highlight:: python</span>
<span class="sd">    .. code-block:: python</span>

<span class="sd">        from anacore.fusion import FusionFileReader</span>

<span class="sd">        # File content&gt;</span>
<span class="sd">        # ##fileformat=VCFv4.3</span>
<span class="sd">        # ##INFO=&lt;ID=MATEID,Number=A,Type=String,Description=&quot;ID of mate breakend.&quot;&gt;</span>
<span class="sd">        # ##INFO=&lt;ID=SVTYPE,Number=1,Type=String,Description=&quot;Type of structural variant.&quot;&gt;</span>
<span class="sd">        # #CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO</span>
<span class="sd">        # 2	321682	bnd_V	T	]13:123456]T	6	PASS	SVTYPE=BND;MATEID=bnd_U</span>
<span class="sd">        # 4	888888	.	T	G	.	PASS	.</span>
<span class="sd">        # 13	123456	bnd_U	C	C[2:321682[,C[17:198983[	6	PASS	SVTYPE=BND;MATEID=bnd_V,bnd_Z</span>
<span class="sd">        # 17	198983	bnd_Z	A	]13:123456]A	6	PASS	SVTYPE=BND;MATEID=bnd_U</span>

<span class="sd">        print(&quot;Breakpoint_1\\tBreakpoint_1_name\\tBreakpoint_2\\tBreakpoint_2_name&quot;)</span>
<span class="sd">        with FusionFileReader.factory(&quot;test.vcf&quot;) as reader:</span>
<span class="sd">            for first, second in reader:</span>
<span class="sd">                print(</span>
<span class="sd">                    &quot;{}:{}&quot;.format(first.chrom, frist.pos),</span>
<span class="sd">                    first.id,</span>
<span class="sd">                    &quot;{}:{}&quot;.format(second.chrom, second.pos),</span>
<span class="sd">                    second.id,</span>
<span class="sd">                    sep=&quot;\\t&quot;</span>
<span class="sd">                )</span>

<span class="sd">        # Result&gt;</span>
<span class="sd">        # Breakpoint_1\tBreakpoint_1_name\tBreakpoint_2\tBreakpoint_2_name</span>
<span class="sd">        # 2:321682\tbnd_V\t13:123456\tbnd_U_0</span>
<span class="sd">        # 13:123456\tbnd_U_1\t17:198983\tbnd_Z</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Alt representation in VCF specification:</span>

<span class="sd">    First in RNA and strand + =&gt; alt: N|</span>
<span class="sd">                    ------&gt;</span>
<span class="sd">                           |</span>
<span class="sd">        &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;     |     &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">                    ------&gt;</span>
<span class="sd">                           |</span>
<span class="sd">        &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;     |     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">    Second in RNA and strand + =&gt; alt: |N</span>
<span class="sd">                            ------&gt;</span>
<span class="sd">                           |</span>
<span class="sd">        &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;     |     &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">                            ------&gt;</span>
<span class="sd">                           |</span>
<span class="sd">        &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;     |     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">    First in RNA and strand - =&gt; alt: |N</span>
<span class="sd">                            &lt;------</span>
<span class="sd">                           |</span>
<span class="sd">        &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;     |     &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">                            &lt;------</span>
<span class="sd">                           |</span>
<span class="sd">        &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;     |     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">    Second in RNA and strand - =&gt; alt: N|</span>
<span class="sd">                    &lt;------</span>
<span class="sd">                           |</span>
<span class="sd">        &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;     |     &gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;&gt;</span>
<span class="sd">        ...................|.................</span>

<span class="sd">                    &lt;------</span>
<span class="sd">                           |</span>
<span class="sd">        &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;     |     &lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;&lt;</span>
<span class="sd">        ...................|.................</span>
<span class="sd">&quot;&quot;&quot;</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s1">&#39;Frederic Escudie&#39;</span>
<span class="n">__copyright__</span> <span class="o">=</span> <span class="s1">&#39;Copyright (C) 2019 IUCT-O&#39;</span>
<span class="n">__license__</span> <span class="o">=</span> <span class="s1">&#39;GNU General Public License&#39;</span>
<span class="n">__version__</span> <span class="o">=</span> <span class="s1">&#39;2.6.1&#39;</span>
<span class="n">__email__</span> <span class="o">=</span> <span class="s1">&#39;escudie.frederic@iuct-oncopole.fr&#39;</span>
<span class="n">__status__</span> <span class="o">=</span> <span class="s1">&#39;prod&#39;</span>

<span class="kn">import</span> <span class="nn">gzip</span>
<span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">uuid</span>
<span class="kn">from</span> <span class="nn">anacore.abstractFile</span> <span class="kn">import</span> <span class="n">isGzip</span>
<span class="kn">from</span> <span class="nn">anacore.annotVcf</span> <span class="kn">import</span> <span class="n">AnnotVCFIO</span>
<span class="kn">from</span> <span class="nn">anacore.sv</span> <span class="kn">import</span> <span class="n">HashedSVIO</span><span class="p">,</span> <span class="n">SVIO</span>
<span class="kn">from</span> <span class="nn">anacore.vcf</span> <span class="kn">import</span> <span class="n">decodeInfoValue</span><span class="p">,</span> <span class="n">getAlleleRecord</span><span class="p">,</span> <span class="n">VCFIO</span><span class="p">,</span> <span class="n">VCFRecord</span><span class="p">,</span> <span class="n">HeaderInfoAttr</span><span class="p">,</span> <span class="n">HeaderFilterAttr</span><span class="p">,</span> <span class="n">HeaderFormatAttr</span>


<div class="viewcode-block" id="getBNDInterval"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.getBNDInterval">[docs]</a><span class="k">def</span> <span class="nf">getBNDInterval</span><span class="p">(</span><span class="n">record</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return the start and end for BND record (CIPOS is taken into account).</span>

<span class="sd">    :param record: The breakend record.</span>
<span class="sd">    :type record: anacore.vcf.VCFRecord</span>
<span class="sd">    :return: Start and end for BND record (CIPOS is taken into account).</span>
<span class="sd">    :rtype: (int, int)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">start</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">pos</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">start</span>
    <span class="k">if</span> <span class="s2">&quot;CIPOS&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
        <span class="n">start</span> <span class="o">+=</span> <span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;CIPOS&quot;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">end</span> <span class="o">+=</span> <span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;CIPOS&quot;</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="s2">&quot;IMPRECISE&quot;</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">-=</span> <span class="mi">1</span>
    <span class="k">return</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span></div>


<div class="viewcode-block" id="getCoordDictFromCoordStr"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.getCoordDictFromCoordStr">[docs]</a><span class="k">def</span> <span class="nf">getCoordDictFromCoordStr</span><span class="p">(</span><span class="n">coord</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return elements from coordinates defined by a fusion caller (&quot;CHROM:POS:STRAND&quot;).</span>

<span class="sd">    :param coord: Coordinates defined by fusion caller in format &quot;CHROM:POS:STRAND&quot;.</span>
<span class="sd">    :type coord: str</span>
<span class="sd">    :return: Coordinates in dict format {&quot;chrom&quot;: &quot;CHROM&quot;, &quot;pos&quot;: &quot;POS&quot;, &quot;strand&quot;:&quot;STRAND&quot;}.</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">matches</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">match</span><span class="p">(</span><span class="s2">&quot;(.+):(\d+):([+-.])&quot;</span><span class="p">,</span> <span class="n">coord</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">matches</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;The coordinates fields </span><span class="si">{}</span><span class="s2"> cannot be parsed.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">coord</span><span class="p">))</span>
    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;chrom&quot;</span><span class="p">:</span> <span class="n">matches</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span>
        <span class="s2">&quot;pos&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">matches</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">2</span><span class="p">)),</span>
        <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">matches</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
    <span class="p">}</span></div>


<div class="viewcode-block" id="getAltFromCoord"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.getAltFromCoord">[docs]</a><span class="k">def</span> <span class="nf">getAltFromCoord</span><span class="p">(</span><span class="n">first_coord</span><span class="p">,</span> <span class="n">second_coord</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return VCF alternative field value for the breakends from the 5&#39; and 3&#39; shard of fusion.</span>

<span class="sd">    :param first_coord: Coordinates of the breakend for the first 5&#39; shard in fusion.</span>
<span class="sd">    :type first_coord: dict</span>
<span class="sd">    :param second_coord: Coordinates of the breakend for the first 3&#39; shard in fusion.</span>
<span class="sd">    :type second_coord: dict</span>
<span class="sd">    :return: Alternative for the first shard in fusion and alternative for the second shard in fusion.</span>
<span class="sd">    :rtype: (str, str)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">first_alt</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="n">second_alt</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span> <span class="ow">or</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;getAltFromCoord cannot process coordinates with unknown strand.&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>  <span class="c1"># First shard is + and second is -</span>
            <span class="n">first_alt</span> <span class="o">=</span> <span class="s2">&quot;N]</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
            <span class="n">second_alt</span> <span class="o">=</span> <span class="s2">&quot;N]</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">]&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>   <span class="c1"># First shard is + and second is +</span>
            <span class="n">first_alt</span> <span class="o">=</span> <span class="s2">&quot;N[</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">[&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
            <span class="n">second_alt</span> <span class="o">=</span> <span class="s2">&quot;]</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">]N&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">:</span>  <span class="c1"># First shard is - and second is -</span>
            <span class="n">first_alt</span> <span class="o">=</span> <span class="s2">&quot;]</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">]N&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
            <span class="n">second_alt</span> <span class="o">=</span> <span class="s2">&quot;N[</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">[&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>  <span class="c1"># First shard is - and second is +</span>
            <span class="n">first_alt</span> <span class="o">=</span> <span class="s2">&quot;[</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">[N&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
            <span class="n">second_alt</span> <span class="o">=</span> <span class="s2">&quot;[</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">[N&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">first_alt</span><span class="p">,</span> <span class="n">second_alt</span></div>


<div class="viewcode-block" id="getStrand"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.getStrand">[docs]</a><span class="k">def</span> <span class="nf">getStrand</span><span class="p">(</span><span class="n">breakend</span><span class="p">,</span> <span class="n">is_first</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return strand from breakend record.</span>

<span class="sd">    :param breakend: Breakend defining a part of an RNA fusion.</span>
<span class="sd">    :type breakend: anacore.vcf.VCFRecord</span>
<span class="sd">    :param is_first: The shard is the first in RNA.</span>
<span class="sd">    :type is_first: boolean</span>
<span class="sd">    :return: Strand from breakend record.</span>
<span class="sd">    :rtype: str</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">is_first</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">is_first</span> <span class="o">=</span> <span class="kc">True</span> <span class="k">if</span> <span class="s2">&quot;RNA_FIRST&quot;</span> <span class="ow">in</span> <span class="n">breakend</span><span class="o">.</span><span class="n">info</span> <span class="k">else</span> <span class="kc">False</span>
    <span class="n">strand</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">if</span> <span class="n">breakend</span><span class="o">.</span><span class="n">alt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;[&quot;</span><span class="p">)</span> <span class="ow">or</span> <span class="n">breakend</span><span class="o">.</span><span class="n">alt</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;]&quot;</span><span class="p">):</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span> <span class="k">if</span> <span class="n">is_first</span> <span class="k">else</span> <span class="s2">&quot;+&quot;</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;+&quot;</span> <span class="k">if</span> <span class="n">is_first</span> <span class="k">else</span> <span class="s2">&quot;-&quot;</span>
    <span class="k">return</span> <span class="n">strand</span></div>


<div class="viewcode-block" id="getCoordStr"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.getCoordStr">[docs]</a><span class="k">def</span> <span class="nf">getCoordStr</span><span class="p">(</span><span class="n">breakend</span><span class="p">,</span> <span class="n">is_first</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Return coordinates elements (chrom, pos and strand) a from breakend record.</span>

<span class="sd">    :param breakend: Breakend defining a part of an RNA fusion.</span>
<span class="sd">    :type breakend: anacore.vcf.VCFRecord</span>
<span class="sd">    :param is_first: The shard is the first in RNA.</span>
<span class="sd">    :type is_first: boolean</span>
<span class="sd">    :return: Coordinates in dict format {&quot;chrom&quot;: &quot;CHROM&quot;, &quot;pos&quot;: &quot;POS&quot;, &quot;strand&quot;:&quot;STRAND&quot;}.</span>
<span class="sd">    :rtype: dict</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;chrom&quot;</span><span class="p">:</span> <span class="n">breakend</span><span class="o">.</span><span class="n">chrom</span><span class="p">,</span> <span class="s2">&quot;pos&quot;</span><span class="p">:</span> <span class="n">breakend</span><span class="o">.</span><span class="n">pos</span><span class="p">,</span> <span class="s2">&quot;strand&quot;</span><span class="p">:</span> <span class="n">getStrand</span><span class="p">(</span><span class="n">breakend</span><span class="p">,</span> <span class="n">is_first</span><span class="p">)}</span></div>


<div class="viewcode-block" id="FusionFileReader"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.FusionFileReader">[docs]</a><span class="k">class</span> <span class="nc">FusionFileReader</span><span class="p">(</span><span class="nb">object</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Factory to identify and return the dedicated to the software used to produce the fusions file.&quot;&quot;&quot;</span>

<div class="viewcode-block" id="FusionFileReader.factory"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.FusionFileReader.factory">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">factory</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the parser dedicated to the software used to produce the fusions file.</span>

<span class="sd">        :param filepath: Path to the file (format: TSV).</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :param args: Additional arguments.</span>
<span class="sd">        :type args: list</span>
<span class="sd">        :param kwargs: Additional keyword arguments.</span>
<span class="sd">        :type kwargs: dict</span>
<span class="sd">        :return: The parser dedicated to the software used to produce the fusions file.</span>
<span class="sd">        :rtype: ArribaIO or FusionCatcherIO or STARFusionIO</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">STARFusionIO</span><span class="o">.</span><span class="n">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">STARFusionIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">FusionCatcherIO</span><span class="o">.</span><span class="n">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">FusionCatcherIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">ArribaIO</span><span class="o">.</span><span class="n">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">ArribaIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">elif</span> <span class="n">BreakendVCFIO</span><span class="o">.</span><span class="n">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">BreakendVCFIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">IOError</span><span class="p">(</span><span class="s2">&quot;The file </span><span class="si">{}</span><span class="s2"> does not have a valid format for &#39;FusionFileReader&#39;.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filepath</span><span class="p">))</span></div></div>


<div class="viewcode-block" id="FusionCatcherIO"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.FusionCatcherIO">[docs]</a><span class="k">class</span> <span class="nc">FusionCatcherIO</span><span class="p">(</span><span class="n">HashedSVIO</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class to read and write fusions in fusionCatcher TSV output format (see https://github.com/ndaniel/fusioncatcher/blob/master/doc/manual.md#output-data). The records are in anacore.vcf.VCFRecord format and a fusion is represented by 2 records (breakend for the 5&#39; shard and breakend for the 3&#39; shard).&quot;&quot;&quot;</span>

    <span class="n">titles</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;Gene_1_symbol(5end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Gene_2_symbol(3end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusion_description&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Counts_of_common_mapping_reads&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Spanning_pairs&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Spanning_unique_reads&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Longest_anchor_found&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusion_finding_method&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusion_point_for_gene_1(5end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusion_point_for_gene_2(3end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Gene_1_id(5end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Gene_2_id(3end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Exon_1_id(5end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Exon_2_id(3end_fusion_partner)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusion_sequence&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Predicted_effect&quot;</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">annot_field</span><span class="o">=</span><span class="s2">&quot;FCANN&quot;</span><span class="p">,</span> <span class="n">sample_name</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build and return an instance of FusionCatcherIO.</span>

<span class="sd">        :param filepath: The filepath.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :param mode: Mode to open the file (&#39;r&#39;, &#39;w&#39;, &#39;a&#39;).</span>
<span class="sd">        :type mode: str</span>
<span class="sd">        :param annot_field: The tag for the field used to store annotations.</span>
<span class="sd">        :type annot_field: str</span>
<span class="sd">        :param sample_name: The name of the sample.</span>
<span class="sd">        :type sample_name: str</span>
<span class="sd">        :return: The new instance.</span>
<span class="sd">        :rtype: FusionCatcherIO</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span> <span class="o">=</span> <span class="n">annot_field</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">sample_name</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">titles</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">titles</span> <span class="o">=</span> <span class="n">FusionCatcherIO</span><span class="o">.</span><span class="n">titles</span>

    <span class="k">def</span> <span class="nf">_parseLine</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a structured record from the current line.</span>

<span class="sd">        :return: The record defined by the current line.</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fusion_record</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">_parseLine</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_dictToBNDRecords</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_dictToBNDRecords</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fusion_record</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the two breakends records from the fusion record.</span>

<span class="sd">        :param fusion_record: The fusion record.</span>
<span class="sd">        :type fusion_record: dict</span>
<span class="sd">        :return: The breakend record for the 5&#39; shard of fusion and the breakend record for the 3&#39; shard.</span>
<span class="sd">        :rtype: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">filters_field</span> <span class="o">=</span> <span class="p">[]</span> <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_description&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="nb">sorted</span><span class="p">([</span><span class="n">elt</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_description&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)])</span>
        <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Spanning_unique_reads&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">:</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Imprecise&quot;</span><span class="p">)</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">sample_field</span> <span class="o">=</span> <span class="p">{</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">:</span> <span class="p">{</span>
                <span class="s2">&quot;PR&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Spanning_pairs&quot;</span><span class="p">]),</span>
                <span class="s2">&quot;SR&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Spanning_unique_reads&quot;</span><span class="p">]),</span>
                <span class="s2">&quot;CM&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Counts_of_common_mapping_reads&quot;</span><span class="p">]),</span>
                <span class="s2">&quot;LA&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Longest_anchor_found&quot;</span><span class="p">])</span>
            <span class="p">}</span>
        <span class="p">}</span>
        <span class="n">format_field</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">sample_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">info_sources</span> <span class="o">=</span> <span class="kc">None</span> <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_finding_method&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span> <span class="k">else</span> <span class="p">[</span><span class="n">elt</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_finding_method&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)]</span>
        <span class="n">first_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">())</span>
        <span class="n">first_coord</span> <span class="o">=</span> <span class="n">getCoordDictFromCoordStr</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_point_for_gene_1(5end_fusion_partner)&quot;</span><span class="p">])</span>
        <span class="n">second_id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">())</span>
        <span class="n">second_coord</span> <span class="o">=</span> <span class="n">getCoordDictFromCoordStr</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_point_for_gene_2(3end_fusion_partner)&quot;</span><span class="p">])</span>
        <span class="n">first_alt</span><span class="p">,</span> <span class="n">second_alt</span> <span class="o">=</span> <span class="n">getAltFromCoord</span><span class="p">(</span><span class="n">first_coord</span><span class="p">,</span> <span class="n">second_coord</span><span class="p">)</span>
        <span class="c1"># First breakend</span>
        <span class="n">first_info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;MATEID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">second_id</span><span class="p">],</span>
            <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;BND&quot;</span><span class="p">,</span>
            <span class="s2">&quot;SOURCES&quot;</span><span class="p">:</span> <span class="n">info_sources</span><span class="p">,</span>
            <span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">:</span> <span class="p">[{</span>
                <span class="s2">&quot;SYMBOL&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Gene_1_symbol(5end_fusion_partner)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Gene_1_id(5end_fusion_partner)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;EXON&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Exon_1_id(5end_fusion_partner)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;Effect&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Predicted_effect&quot;</span><span class="p">]</span>
            <span class="p">}]</span>
        <span class="p">}</span>
        <span class="n">first_bnd</span> <span class="o">=</span> <span class="n">VCFRecord</span><span class="p">(</span>
            <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
            <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
            <span class="n">first_id</span><span class="p">,</span>
            <span class="s2">&quot;N&quot;</span><span class="p">,</span>  <span class="c1"># ref</span>
            <span class="p">[</span><span class="n">first_alt</span><span class="p">],</span>  <span class="c1"># alt</span>
            <span class="n">pFilter</span><span class="o">=</span><span class="n">filters_field</span><span class="p">,</span>
            <span class="n">info</span><span class="o">=</span><span class="n">first_info</span><span class="p">,</span>
            <span class="n">pFormat</span><span class="o">=</span><span class="n">format_field</span><span class="p">,</span>
            <span class="n">samples</span><span class="o">=</span><span class="n">sample_field</span>
        <span class="p">)</span>
        <span class="c1"># Second breakends</span>
        <span class="n">second_info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;MATEID&quot;</span><span class="p">:</span> <span class="p">[</span><span class="n">first_id</span><span class="p">],</span>
            <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;BND&quot;</span><span class="p">,</span>
            <span class="s2">&quot;SOURCES&quot;</span><span class="p">:</span> <span class="n">info_sources</span><span class="p">,</span>
            <span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Fusion_sequence&quot;</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">:</span> <span class="p">[{</span>
                <span class="s2">&quot;SYMBOL&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Gene_2_symbol(3end_fusion_partner)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Gene_2_id(3end_fusion_partner)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;EXON&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Exon_2_id(3end_fusion_partner)&quot;</span><span class="p">],</span>
                <span class="s2">&quot;Effect&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;Predicted_effect&quot;</span><span class="p">]</span>
            <span class="p">}]</span>
        <span class="p">}</span>
        <span class="n">second_bnd</span> <span class="o">=</span> <span class="n">VCFRecord</span><span class="p">(</span>
            <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
            <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
            <span class="n">second_id</span><span class="p">,</span>
            <span class="s2">&quot;N&quot;</span><span class="p">,</span>  <span class="c1"># ref</span>
            <span class="p">[</span><span class="n">second_alt</span><span class="p">],</span>  <span class="c1"># alt</span>
            <span class="n">pFilter</span><span class="o">=</span><span class="n">filters_field</span><span class="p">,</span>
            <span class="n">info</span><span class="o">=</span><span class="n">second_info</span><span class="p">,</span>
            <span class="n">pFormat</span><span class="o">=</span><span class="n">format_field</span><span class="p">,</span>
            <span class="n">samples</span><span class="o">=</span><span class="n">sample_field</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">first_bnd</span><span class="p">,</span> <span class="n">second_bnd</span><span class="p">]</span>

<div class="viewcode-block" id="FusionCatcherIO.write"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.FusionCatcherIO.write">[docs]</a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">breakends</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write record line in file.</span>

<span class="sd">        :param record: The first breakend and the second breakend in pair.</span>
<span class="sd">        :type record: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_BNDRecordsToDict</span><span class="p">(</span><span class="n">breakends</span><span class="p">))</span></div>

    <span class="k">def</span> <span class="nf">_BNDRecordsToDict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">breakends</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the FusionCatcher dict from the two breakends of a fusion.</span>

<span class="sd">        :param breakends: The two breakends of a fusion.</span>
<span class="sd">        :type breakends: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        :return: The FusionCatcher dict.</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="o">=</span> <span class="n">breakends</span>
        <span class="n">first_coord</span> <span class="o">=</span> <span class="n">getCoordStr</span><span class="p">(</span><span class="n">first</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">second_coord</span> <span class="o">=</span> <span class="n">getCoordStr</span><span class="p">(</span><span class="n">second</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;Gene_1_symbol(5end_fusion_partner)&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Gene_2_symbol(3end_fusion_partner)&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Fusion_description&quot;</span><span class="p">:</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">elt</span> <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">filter</span> <span class="k">if</span> <span class="n">elt</span> <span class="o">!=</span> <span class="s2">&quot;Imprecise&quot;</span><span class="p">]),</span>
            <span class="s2">&quot;Counts_of_common_mapping_reads&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;CM&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Spanning_pairs&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;PR&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Spanning_unique_reads&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SR&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Longest_anchor_found&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;LA&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Fusion_finding_method&quot;</span><span class="p">:</span> <span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;SOURCES&quot;</span><span class="p">]),</span>
            <span class="s2">&quot;Fusion_point_for_gene_1(5end_fusion_partner)&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
                <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;Fusion_point_for_gene_2(3end_fusion_partner)&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
                <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;Gene_1_id(5end_fusion_partner)&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Gene&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Gene_2_id(3end_fusion_partner)&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Gene&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Exon_1_id(5end_fusion_partner)&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;EXON&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Exon_2_id(3end_fusion_partner)&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;EXON&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Fusion_sequence&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">],</span>
            <span class="s2">&quot;Predicted_effect&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Effect&quot;</span><span class="p">]</span>
        <span class="p">}</span>

<div class="viewcode-block" id="FusionCatcherIO.isValid"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.FusionCatcherIO.isValid">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return true if the file is in FusionCatcher output format.</span>

<span class="sd">        :param filepath: The file path.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :return: True if the file is in FusionCatcher output format.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">SVIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">reader</span><span class="p">:</span>
                <span class="n">mandatory_fields</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">FusionCatcherIO</span><span class="o">.</span><span class="n">titles</span><span class="p">)</span> <span class="o">-</span> <span class="p">{</span><span class="s2">&quot;Predicted_effect&quot;</span><span class="p">}</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mandatory_fields</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">reader</span><span class="o">.</span><span class="n">titles</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">return</span> <span class="n">is_valid</span></div>

<div class="viewcode-block" id="FusionCatcherIO.setVCFHeader"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.FusionCatcherIO.setVCFHeader">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">setVCFHeader</span><span class="p">(</span><span class="n">vcf_io</span><span class="p">,</span> <span class="n">annotation_field</span><span class="o">=</span><span class="s2">&quot;FCANN&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Set header of an AnnotVCFIO to write FusionCatcher records in VCF output. This function pre-set ANN_titles, info and format used by FusionCatcherIO records.</span>

<span class="sd">        :param vcf_io: The VCF used to write FusionCatcher records.</span>
<span class="sd">        :type vcf_io: anacore.annotVcf.AnnotVCFIO</span>
<span class="sd">        :param annotation_field: INFO field used for store annotations.</span>
<span class="sd">        :type annotation_field: str</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># INFO</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;MATEID&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;MATEID&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;ID of mate breakend.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;SVTYPE&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Type of structural variant.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SOURCES&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;SOURCES&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Aligning method used for mapping the reads and finding the fusion genes. Here are two methods used which are: (i) BOWTIE = only Bowtie aligner is used for mapping the reads on the genome and exon-exon fusion junctions, (ii) BOWTIE+BLAT = Bowtie aligner is used for mapping reads on the genome and BLAT is used for mapping reads for finding the fusion junction, (iii) BOWTIE+STAR = Bowtie aligner is used for mapping reads on the genome and STAR is used for mapping reads for finding the fusion junction, (iv) BOWTIE+BOWTIE2 = Bowtie aligner is used for mapping reads on the genome and Bowtie2 is used for mapping reads for finding the fusion junction.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Flag&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;For RNA fusions, this break-end is 5&#39; in the fusion transcript.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;The sequence of the breakend spanning contig.&quot;</span><span class="p">),</span>
            <span class="n">annotation_field</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="n">annotation_field</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Consequence annotations. Format: SYMBOL|Gene|EXON|Effect.&quot;</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="c1"># ANN_titles</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">ANN_titles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;EXON&quot;</span><span class="p">,</span> <span class="s2">&quot;Effect&quot;</span><span class="p">]</span>
        <span class="c1"># FORMAT</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;PR&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;PR&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Count of pairs of reads supporting the fusion (including also the multimapping reads).&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SR&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SR&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Count of unique reads (i.e. unique mapping positions) mapping on the fusion junction. Shortly, here are counted all the reads which map on fusion junction minus the PCR duplicated reads.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;CM&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;CM&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Count of reads mapping simultaneously on both genes which form the fusion gene. This is an indication how similar are the DNA/RNA sequences of the genes forming the fusion gene (i.e. what is their homology because highly homologous genes tend to appear show as candidate fusion genes). In case of completely different sequences of the genes involved in forming a fusion gene then here it is expected to have the value zero.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;LA&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;LA&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Longest anchor (hangover) found among the unique reads mapping on the fusion junction.&quot;</span><span class="p">)</span>
        <span class="p">}</span></div></div>


<span class="n">starfusion_filters</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;GTEx_recurrent_StarF2019&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;GTEx_recurrent_StarF2019&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusions found recurrently in GTEx as mined using STAR-Fusion v1.5.0&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;BodyMap&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;BodyMap&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusions found by STAR-Fusion as applied to the Illumina Human Body Map reference data&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;DGD_PARALOGS&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;DGD_PARALOGS&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Duplicated genes as per the Duplicated Genes Database&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;HGNC_GENEFAM&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;HGNC_GENEFAM&quot;</span><span class="p">,</span>
        <span class="s2">&quot;HGNC gene family membership as per ftp://ftp.ebi.ac.uk/pub/databases/genenames/genefam_list.txt.gz&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;Greger_Normal&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;Greger_Normal&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fusion transcripts (mostly from tandem genes) detected based on analysis of RNA-Seq from 1000 genomes project samples. List derived from Greger et al. PLOS One, 2014&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;Babiceanu_Normal&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;Babiceanu_Normal&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Recurrent chimeric fusion RNAs in non-cancer tissues and cells as per Babiceanu et al. NAR, 2016&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;ConjoinG&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;ConjoinG&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Fused transcripts derived from the Conjoined Gene Database&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;Imprecise&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;Imprecise&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RNA fusion candidates for which no spanning contig was found.&quot;</span>
    <span class="p">)</span>
<span class="p">}</span>


<div class="viewcode-block" id="STARFusionIO"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.STARFusionIO">[docs]</a><span class="k">class</span> <span class="nc">STARFusionIO</span><span class="p">(</span><span class="n">HashedSVIO</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class to read and write fusions in STAR-Fusion TSV output format (see https://github.com/STAR-Fusion/STAR-Fusion/wiki#Outputs). The records are in anacore.vcf.VCFRecord format and a fusion is represented by 2 records (breakend for the 5&#39; shard and breakend for the 3&#39; shard).&quot;&quot;&quot;</span>

    <span class="n">titles</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;FusionName&quot;</span><span class="p">,</span>
        <span class="s2">&quot;JunctionReadCount&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SpanningFragCount&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SpliceType&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LeftGene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LeftBreakpoint&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RightGene&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RightBreakpoint&quot;</span><span class="p">,</span>
        <span class="s2">&quot;JunctionReads&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SpanningFrags&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LargeAnchorSupport&quot;</span><span class="p">,</span>
        <span class="s2">&quot;FFPM&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LeftBreakDinuc&quot;</span><span class="p">,</span>
        <span class="s2">&quot;LeftBreakEntropy&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RightBreakDinuc&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RightBreakEntropy&quot;</span><span class="p">,</span>
        <span class="s2">&quot;annots&quot;</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">annot_field</span><span class="o">=</span><span class="s2">&quot;FCANN&quot;</span><span class="p">,</span> <span class="n">sample_name</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build and return an instance of STARFusionIO.</span>

<span class="sd">        :param filepath: The filepath.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :param mode: Mode to open the file (&#39;r&#39;, &#39;w&#39;, &#39;a&#39;).</span>
<span class="sd">        :type mode: str</span>
<span class="sd">        :param annot_field: The tag for the field used to store annotations.</span>
<span class="sd">        :type annot_field: str</span>
<span class="sd">        :param sample_name: The name of the sample.</span>
<span class="sd">        :type sample_name: str</span>
<span class="sd">        :return: The new instance.</span>
<span class="sd">        :rtype: STARFusionIO</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">,</span> <span class="n">separator</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">title_starter</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span> <span class="o">=</span> <span class="n">annot_field</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">sample_name</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">titles</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">titles</span> <span class="o">=</span> <span class="n">STARFusionIO</span><span class="o">.</span><span class="n">titles</span>

    <span class="k">def</span> <span class="nf">_parseLine</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a structured record from the current line.</span>

<span class="sd">        :return: The record defined by the current line.</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fusion_record</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">_parseLine</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_dictToBNDRecords</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_dictToBNDRecords</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fusion_record</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the two breakends records from the fusion record.</span>

<span class="sd">        :param fusion_record: The fusion record.</span>
<span class="sd">        :type fusion_record: dict</span>
<span class="sd">        :return: The breakend record for the 5&#39; shard of fusion and the breakend record for the 3&#39; shard.</span>
<span class="sd">        :rtype: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">breakends</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">large_anchor_support</span> <span class="o">=</span> <span class="s2">&quot;1&quot;</span> <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;LargeAnchorSupport&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;YES_LDAS&quot;</span> <span class="k">else</span> <span class="s2">&quot;0&quot;</span>
        <span class="n">info_annot_tags</span> <span class="o">=</span> <span class="p">[</span><span class="n">elt</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;annots&quot;</span><span class="p">])]</span>
        <span class="n">filters_field</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">starfusion_filters</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">info_annot_tags</span><span class="p">)))</span>
        <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;JunctionReadCount&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;0&quot;</span><span class="p">:</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Imprecise&quot;</span><span class="p">)</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filters_field</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">filters_field</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">]</span>
        <span class="n">coord_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">side</span><span class="p">,</span> <span class="n">mate_side</span> <span class="ow">in</span> <span class="p">[(</span><span class="s2">&quot;Left&quot;</span><span class="p">,</span> <span class="s2">&quot;Right&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;Right&quot;</span><span class="p">,</span> <span class="s2">&quot;Left&quot;</span><span class="p">)]:</span>
            <span class="nb">id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">())</span>
            <span class="n">coord</span> <span class="o">=</span> <span class="n">getCoordDictFromCoordStr</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="n">side</span> <span class="o">+</span> <span class="s2">&quot;Breakpoint&quot;</span><span class="p">])</span>
            <span class="n">coord_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">coord</span><span class="p">)</span>
            <span class="c1"># INFO</span>
            <span class="n">gene_symbol</span><span class="p">,</span> <span class="n">gene_id</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="n">side</span> <span class="o">+</span> <span class="s2">&quot;Gene&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;^&quot;</span><span class="p">)</span>
            <span class="n">info</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;BND&quot;</span><span class="p">,</span>
                <span class="s2">&quot;BREAK_DINUC&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="n">side</span> <span class="o">+</span> <span class="s2">&quot;BreakDinuc&quot;</span><span class="p">],</span>
                <span class="s2">&quot;BREAK_ENTROPY&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="n">side</span> <span class="o">+</span> <span class="s2">&quot;BreakEntropy&quot;</span><span class="p">],</span>
                <span class="s2">&quot;SPLICE_TYPE&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;SpliceType&quot;</span><span class="p">],</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">:</span> <span class="p">[{</span>
                    <span class="s2">&quot;SYMBOL&quot;</span><span class="p">:</span> <span class="n">gene_symbol</span><span class="p">,</span>
                    <span class="s2">&quot;Gene&quot;</span><span class="p">:</span> <span class="n">gene_id</span><span class="p">,</span>
                    <span class="s2">&quot;Tags&quot;</span><span class="p">:</span> <span class="s2">&quot;&amp;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info_annot_tags</span><span class="p">)</span>
                <span class="p">}]</span>
            <span class="p">}</span>
            <span class="c1"># SAMPLES</span>
            <span class="n">samples_field</span> <span class="o">=</span> <span class="p">{</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;PR&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;SpanningFragCount&quot;</span><span class="p">]),</span>
                    <span class="s2">&quot;SR&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;JunctionReadCount&quot;</span><span class="p">]),</span>
                    <span class="s2">&quot;hasLAS&quot;</span><span class="p">:</span> <span class="n">large_anchor_support</span><span class="p">,</span>
                    <span class="s2">&quot;FFPM&quot;</span><span class="p">:</span> <span class="nb">float</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;FFPM&quot;</span><span class="p">])</span>
                <span class="p">}</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">side</span> <span class="o">==</span> <span class="s2">&quot;Left&quot;</span><span class="p">:</span>
                <span class="n">info</span><span class="p">[</span><span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">if</span> <span class="s2">&quot;JunctionReads&quot;</span> <span class="ow">in</span> <span class="n">fusion_record</span><span class="p">:</span>
                    <span class="n">samples_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;JRL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;JunctionReads&quot;</span><span class="p">]</span>
                <span class="k">if</span> <span class="s2">&quot;SpanningFrags&quot;</span> <span class="ow">in</span> <span class="n">fusion_record</span><span class="p">:</span>
                    <span class="n">samples_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SFL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;SpanningFrags&quot;</span><span class="p">]</span>
            <span class="c1"># Record</span>
            <span class="n">breakends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">VCFRecord</span><span class="p">(</span>
                    <span class="n">coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
                    <span class="n">coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                    <span class="nb">id</span><span class="p">,</span>
                    <span class="s2">&quot;N&quot;</span><span class="p">,</span>  <span class="c1"># ref</span>
                    <span class="p">[],</span>  <span class="c1"># alt</span>
                    <span class="n">info</span><span class="o">=</span><span class="n">info</span><span class="p">,</span>
                    <span class="n">pFilter</span><span class="o">=</span><span class="n">filters_field</span><span class="p">,</span>
                    <span class="n">pFormat</span><span class="o">=</span><span class="nb">sorted</span><span class="p">(</span><span class="n">samples_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                    <span class="n">samples</span><span class="o">=</span><span class="n">samples_field</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="n">left_alt</span><span class="p">,</span> <span class="n">right_alt</span> <span class="o">=</span> <span class="n">getAltFromCoord</span><span class="p">(</span><span class="o">*</span><span class="n">coord_list</span><span class="p">)</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alt</span> <span class="o">=</span> <span class="p">[</span><span class="n">left_alt</span><span class="p">]</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">alt</span> <span class="o">=</span> <span class="p">[</span><span class="n">right_alt</span><span class="p">]</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">breakends</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">breakends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">breakends</span>

<div class="viewcode-block" id="STARFusionIO.write"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.STARFusionIO.write">[docs]</a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">breakends</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write record line in file.</span>

<span class="sd">        :param record: The first breakend and the second breakend in pair.</span>
<span class="sd">        :type record: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_BNDRecordsToDict</span><span class="p">(</span><span class="n">breakends</span><span class="p">))</span></div>

    <span class="k">def</span> <span class="nf">_BNDRecordsToDict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">breakends</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the STAR-fusion dict from the two breakends of a fusion.</span>

<span class="sd">        :param breakends: The two breakends of a fusion.</span>
<span class="sd">        :type breakends: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        :return: The STAR-fusion dict.</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="o">=</span> <span class="n">breakends</span>
        <span class="n">first_coord</span> <span class="o">=</span> <span class="n">getCoordStr</span><span class="p">(</span><span class="n">first</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">second_coord</span> <span class="o">=</span> <span class="n">getCoordStr</span><span class="p">(</span><span class="n">second</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;FusionName&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">--</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
                <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
            <span class="p">),</span>
            <span class="s2">&quot;JunctionReadCount&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SR&quot;</span><span class="p">],</span>
            <span class="s2">&quot;SpanningFragCount&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;PR&quot;</span><span class="p">],</span>
            <span class="s2">&quot;SpliceType&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;SPLICE_TYPE&quot;</span><span class="p">],</span>
            <span class="s2">&quot;LeftGene&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">^</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
                <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Gene&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;LeftBreakpoint&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
                <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;RightGene&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">^</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
                <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Gene&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;RightBreakpoint&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
                <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;JunctionReads&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="s2">&quot;JRL&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span> <span class="k">else</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;JRL&quot;</span><span class="p">],</span>
            <span class="s2">&quot;SpanningFrags&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span> <span class="k">if</span> <span class="s2">&quot;SFL&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span> <span class="k">else</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SFL&quot;</span><span class="p">],</span>
            <span class="s2">&quot;LargeAnchorSupport&quot;</span><span class="p">:</span> <span class="s2">&quot;YES_LDAS&quot;</span> <span class="k">if</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;hasLAS&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span> <span class="k">else</span> <span class="s2">&quot;NO_LDAS&quot;</span><span class="p">,</span>
            <span class="s2">&quot;FFPM&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;FFPM&quot;</span><span class="p">],</span>
            <span class="s2">&quot;LeftBreakDinuc&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;BREAK_DINUC&quot;</span><span class="p">],</span>
            <span class="s2">&quot;LeftBreakEntropy&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;BREAK_ENTROPY&quot;</span><span class="p">],</span>
            <span class="s2">&quot;RightBreakDinuc&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;BREAK_DINUC&quot;</span><span class="p">],</span>
            <span class="s2">&quot;RightBreakEntropy&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;BREAK_ENTROPY&quot;</span><span class="p">],</span>
            <span class="s2">&quot;annots&quot;</span><span class="p">:</span> <span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Tags&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;&amp;&quot;</span><span class="p">))</span>
        <span class="p">}</span>

<div class="viewcode-block" id="STARFusionIO.isValid"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.STARFusionIO.isValid">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return true if the file is in STAR-Fusion output format.</span>

<span class="sd">        :param filepath: The file path.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :return: True if the file is in STAR-Fusion output format.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">SVIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">title_starter</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">reader</span><span class="p">:</span>
                <span class="n">mandatory_fields</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">STARFusionIO</span><span class="o">.</span><span class="n">titles</span><span class="p">)</span> <span class="o">-</span> <span class="p">{</span><span class="s2">&quot;JunctionReads&quot;</span><span class="p">,</span> <span class="s2">&quot;SpanningFrags&quot;</span><span class="p">}</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mandatory_fields</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">reader</span><span class="o">.</span><span class="n">titles</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># All mandatory fields of Arriba are in reader</span>
                    <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">return</span> <span class="n">is_valid</span></div>

<div class="viewcode-block" id="STARFusionIO.setVCFHeader"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.STARFusionIO.setVCFHeader">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">setVCFHeader</span><span class="p">(</span><span class="n">vcf_io</span><span class="p">,</span> <span class="n">annotation_field</span><span class="o">=</span><span class="s2">&quot;FCANN&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Set header of an AnnotVCFIO to write STAR-Fusion records in VCF output. This function pre-set ANN_titles, info and format used by STARFusionIO records.</span>

<span class="sd">        :param vcf_io: The VCF used to write STAR-Fusion records.</span>
<span class="sd">        :type vcf_io: anacore.annotVcf.AnnotVCFIO</span>
<span class="sd">        :param annotation_field: INFO field used for store annotations.</span>
<span class="sd">        :type annotation_field: str</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># FILTERS</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">filter</span> <span class="o">=</span> <span class="n">starfusion_filters</span>
        <span class="c1"># INFO</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;MATEID&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;MATEID&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;ID of mate breakend.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;SVTYPE&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Type of structural variant.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Flag&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;For RNA fusions, this breakend is 5&#39; in the fusion transcript.&quot;</span><span class="p">),</span>  <span class="c1"># In the fusion transcript, the &#39;LeftGene&#39; is always 5&#39; relative to the &#39;RightGene&#39; (https://groups.google.com/forum/#!topic/star-fusion/9rBLT-v5JHI)</span>
            <span class="s2">&quot;BREAK_DINUC&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;BREAK_DINUC&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Dinucleotides flanking the breakend of the fragment excluded by the fusion.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;BREAK_ENTROPY&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;BREAK_ENTROPY&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Float&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Shannon entropy of the 15 exonic bases flanking the breakpoint. The maximum entropy is 2, representing highest complexity. The lowest would be zero (involving a 15 base mononucleotide run). Low entropy sites should generally be treated as less confident breakpoints.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SPLICE_TYPE&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;SPLICE_TYPE&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Whether the proposed breakpoint occurs at reference exon junctions as provided by the reference transcript structure annotations (ex. gencode).&quot;</span><span class="p">),</span>
            <span class="n">annotation_field</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="n">annotation_field</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Annotation generated by FusionAnnotator (see: https://github.com/FusionAnnotator/CTAT_HumanFusionLib/wiki). Format: SYMBOL|Gene|Tags&quot;</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="c1"># ANN_titles</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">ANN_titles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="s2">&quot;Gene&quot;</span><span class="p">,</span> <span class="s2">&quot;Tags&quot;</span><span class="p">]</span>
        <span class="c1"># FORMAT</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;PR&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;PR&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Number of RNA-Seq fragments that encompass the fusion junction such that one read of the pair aligns to a different gene than the other paired-end read of that fragment (SpanningFragCount).&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SR&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SR&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Number of RNA-Seq fragments containing a read that aligns as a split read at the site of the putative fusion junction (JunctionReadCount).&quot;</span><span class="p">),</span>
            <span class="s2">&quot;hasLAS&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;hasLAS&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;This column indicates whether there are split reads that provide &#39;long&#39; (set to length of 25 bases) alignments on both sides of the putative breakpoint (LargeAnchorSupport).&quot;</span><span class="p">),</span>
            <span class="s2">&quot;FFPM&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;FFPM&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Float&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Normalized measures of the fusion-supporting rna-seq fragments (fusion fragments per million total reads).&quot;</span><span class="p">),</span>
            <span class="s2">&quot;JRL&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;JRL&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;RNA-Seq fragments containing a read that aligns as a split read at the site of the putative fusion junction.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SFL&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SFL&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;RNA-Seq fragments that encompass the fusion junction such that one read of the pair aligns to a different gene than the other paired-end read of that fragment.&quot;</span><span class="p">)</span>
        <span class="p">}</span></div></div>


<span class="n">arriba_event_level_filters</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;merge_adjacent&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;merge_adjacent&quot;</span><span class="p">,</span>
        <span class="s2">&quot;When multiple alternative alignments with equal quality are possible, STAR does not always choose the same one. This leads to multiple breakpoints in close proximity, all of which arise from one and the same event. This filter merges all breakpoints in a window of 5 bp to the breakpoint with the highest number of supporting reads.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;non_coding_neighbors&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;non_coding_neighbors&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Genes which are not well studied suffer from incomplete annotation. Many exons are annotated as separate genes even though they might actually be part of one and the same gene. Predicted genes named RP11-... are common examples for this. When poorly understood genes lie next to each other on the same strand, this would frequently lead to false positive predictions of deletions, because the transcripts that span both genes give rise to reads, which resemble focal deletions. This filter discards deletions, which are predicted between two neighboring genes, if both genes are non-coding or one breakpoint is intergenic.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;intragenic_exonic&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;intragenic_exonic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Since exons usually make up only a small fraction of a gene, it is more likely that a genomic rearrangement starts and ends in intronic regions. On the transcriptomic level, this manifests as breakpoints at splice-sites or in introns. Many candidates found by STAR have both breakpoints within exons of the same gene. This is particularly true for intragenic events, which are prone to in vitro artifacts. This filter removes intragenic events, if both breakpoints are in exons and more than 80</span><span class="si">% o</span><span class="s2">f the region between the breakpoints is intronic, such that it should be very unlikely that both breakpoints are located inside exons.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;min_support&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;min_support&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter discards all events with few reads.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;relative_support&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;relative_support&quot;</span><span class="p">,</span>
        <span class="s2">&quot;With increasing expression of a gene the number of chimeric reads mapping to the gene increases, too. Arriba assumes a polynomial relationship between the number of events and the number of supporting reads for a given event. This assumption is based on empirical evidence. An expected value (e-value) is calculated for every event, reflecting how many events with the given number of supporting reads are expected by chance. This filter selects only those events with a low e-value, i.e., that have a high number of supporting reads relative to the overall number of events in a gene. Multiple covariates are taken into account, such as whether a breakpoint is at a splice-site or whether the event is intragenic. The parameters of the polynomial relationship are fixed and were estimated from several hundred RNA-Seq samples of various cancer types.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;intronic&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;intronic&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter discards an event, when none of the supporting reads overlaps an exon (i.e., all alignments are in introns or intergenic regions). True events most often have at least one breakpoint in an exon or at an exon boundary.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;known_fusions&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;known_fusions&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Events with too few supporting reads but known in cancer.&quot;</span>
    <span class="p">),</span>  <span class="c1"># Protection</span>
    <span class="s2">&quot;pcr_fusions&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;pcr_fusions&quot;</span><span class="p">,</span>
        <span class="s2">&quot;In some tissues certain genes are expressed at very high levels, for example hemoglobin and fibrinogen in blood or collagens in connective tissue. Presumably, the abundance of fragments from such genes increases the chance of unrelated molecules sticking together, which can serve as undesired primers for PCR or may cause the reverse transcriptase enzyme to switch templates. These processes generate a large amount of chimeric fragments in vitro. Such artifactual fusions can be recognized as an extraordinary number of events with breakpoints within exons (rather than at exon boundaries, which is more common for true predictions). This filter eliminates events with genes that are highly expressed (top 0.2%) and have an unbalanced number of split-reads vs. discordant mates or that have an excessive amount of intra-exonic breakpoints.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;spliced&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;spliced&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter recovers events discarded due to a low number of supporting reads, given that both breakpoints of the event are at splice-sites and there is at least one additional event linking the same pair of genes.&quot;</span>
    <span class="p">),</span>  <span class="c1"># Protection</span>
    <span class="s2">&quot;select_best&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;select_best&quot;</span><span class="p">,</span>
        <span class="s2">&quot;If there are multiple breakpoints detected between the same pair of genes, this filter discards all but the most credible one. Events with split reads in both genes are preferred over events with only discordant mates, because in the latter case, the precise breakpoint is unknown. Moreover, events with a higher number of supporting reads are favored.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;many_spliced&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;many_spliced&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Occassionally, a genomic rearrangement produces multiple alternatively spliced transcripts, all of which are low expressed and therefore discarded by the filters relative_support or min_support. This filter recovers events that were discarded due to too few supporting reads, under the circumstances that there are many events between a pair of genes with at least one of the breakpoints at splice-sites. All events between the pair of genes with at least one spliced breakpoint are recovered.&quot;</span>
    <span class="p">),</span>  <span class="c1"># Protection</span>
    <span class="s2">&quot;no_genomic_support&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;no_genomic_support&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter removes events with low confidence, if they are not confirmed by structural variant calls obtained from whole-genome sequencing.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;blacklist&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;blacklist&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter removes events with breakpoint coordinates matching entries in the blacklist. If an event has no split-reads (but only discordant mates), the precise breakpoint coordinates are unknown. In this case the event is discarded, if the breakpoints are within a range of the insert size around the blacklisted coordinates.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;short_anchor&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;short_anchor&quot;</span><span class="p">,</span>
        <span class="s2">&quot;A chimeric read aligns to some part in one of the fused genes and to some part to the other gene. The anchor of a read is the longer aligned segment. It is more likely to be aligned correctly. True positives often have anchors in both fused genes, whereas alignment artifacts are frequently characterized by only a small segment aligning to one of the genes and all anchors to the other.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;end_to_end&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;end_to_end&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Theoretically, it should be impossible to observe fusions which only retain the 3&#39; ends of the fusion partners, because the promoter would be missing. In reality, end-to-end fused transcripts are observed, albeit rarely. Most of them are false positives. Therefore, this filter discards such events, unless they have a lot of supporting reads, i.e., split-reads in both fusion partners or split-reads and discordant mates.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;no_coverage&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;no_coverage&quot;</span><span class="p">,</span>
        <span class="s2">&quot;For intronic and intragenic breakpoints as well as read-through fusions, this filter checks, if there is some coverage in the vicinity of the breakpoint in the normal alignments. Only reads that are not chimeric are considered. When there are no non-chimeric reads near the breakpoint, this is indicative of an alignment artifact.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;homologs&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;homologs&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter discards events between genes that have high sequence homology, which frequently leads to erroneous alignments. Homology is quantified by counting the number of shared 16-mers.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;mismappers&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;mismappers&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Many alignment artifacts are caused by an excessive number of reference mismatches in close proximity due to sequencing errors or adjacent SNPs. STAR tends to clip reads when it encounters a cluster of mismatches. The clipped segment is then used for a chimeric alignment, which occassionally aligns elsewhere in the genome. The filter mismappers performs a sensitive realignment of both segments. If both segments can be aligned to the same gene (while allowing more mismatches than STAR does), the chimeric alignment is considered to be an artifact. When an important proportion of the supporting reads are classified as being aligned incorrectly, the event is discarded.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;genomic_support&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;genomic_support&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter recovers events which were discarded by previous filters due to few supporting reads, but which can be explained by genomic rearrangements as evidenced by structural variant calls obtained from whole-genome sequencing data. Arriba considers structural variant calls to match with breakpoints seen in transcriptomic data, when the breakpoints are less then 100 kb apart and the orientation of the genomic and transcriptomic breakpoints are identical.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;isoforms&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;isoforms&quot;</span><span class="p">,</span>
        <span class="s2">&quot;This filter searches for additional isoforms for those gene pairs that are predicted to be fused in proper orientation. Typically there is a major isoform which is expressed at a high level and a few low expressed isoforms.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;Imprecise&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;Imprecise&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RNA fusion candidates for which no spanning contig was found.&quot;</span>
    <span class="p">),</span>
    <span class="s2">&quot;Unstranded&quot;</span><span class="p">:</span> <span class="n">HeaderFilterAttr</span><span class="p">(</span>
        <span class="s2">&quot;Unstranded&quot;</span><span class="p">,</span>
        <span class="s2">&quot;RNA fusion is not stranded.&quot;</span>
    <span class="p">)</span>
<span class="p">}</span>


<div class="viewcode-block" id="ArribaIO"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.ArribaIO">[docs]</a><span class="k">class</span> <span class="nc">ArribaIO</span><span class="p">(</span><span class="n">HashedSVIO</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class to read and write fusions in Arriba TSV output format (see https://github.com/suhrig/arriba/blob/master/documentation/output-files.md). The records are in anacore.vcf.VCFRecord format and a fusion is represented by 2 records (breakend for the 5&#39; shard and breakend for the 3&#39; shard).&quot;&quot;&quot;</span>

    <span class="n">titles</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;gene1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gene2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;strand1(gene/fusion)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;strand2(gene/fusion)&quot;</span><span class="p">,</span>
        <span class="s2">&quot;breakpoint1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;breakpoint2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;site1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;site2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;type&quot;</span><span class="p">,</span>
        <span class="s2">&quot;direction1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;direction2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;split_reads1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;split_reads2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;discordant_mates&quot;</span><span class="p">,</span>
        <span class="s2">&quot;coverage1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;coverage2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;confidence&quot;</span><span class="p">,</span>
        <span class="s2">&quot;closest_genomic_breakpoint1&quot;</span><span class="p">,</span>
        <span class="s2">&quot;closest_genomic_breakpoint2&quot;</span><span class="p">,</span>
        <span class="s2">&quot;filters&quot;</span><span class="p">,</span>
        <span class="s2">&quot;fusion_transcript&quot;</span><span class="p">,</span>
        <span class="s2">&quot;reading_frame&quot;</span><span class="p">,</span>
        <span class="s2">&quot;peptide_sequence&quot;</span><span class="p">,</span>
        <span class="s2">&quot;read_identifiers&quot;</span>
    <span class="p">]</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">annot_field</span><span class="o">=</span><span class="s2">&quot;FCANN&quot;</span><span class="p">,</span> <span class="n">sample_name</span><span class="o">=</span><span class="s2">&quot;sample&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build and return an instance of ArribaIO.</span>

<span class="sd">        :param filepath: The filepath.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :param mode: Mode to open the file (&#39;r&#39;, &#39;w&#39;, &#39;a&#39;).</span>
<span class="sd">        :type mode: str</span>
<span class="sd">        :param annot_field: The tag for the field used to store annotations.</span>
<span class="sd">        :type annot_field: str</span>
<span class="sd">        :param sample_name: The name of the sample.</span>
<span class="sd">        :type sample_name: str</span>
<span class="sd">        :return: The new instance.</span>
<span class="sd">        :rtype: ArribaIO</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="n">mode</span><span class="p">,</span> <span class="n">separator</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">title_starter</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span> <span class="o">=</span> <span class="n">annot_field</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span> <span class="o">=</span> <span class="n">sample_name</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">titles</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">titles</span> <span class="o">=</span> <span class="n">ArribaIO</span><span class="o">.</span><span class="n">titles</span>

    <span class="k">def</span> <span class="nf">_parseLine</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a structured record from the current line.</span>

<span class="sd">        :return: The record defined by the current line.</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">fusion_record</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">_parseLine</span><span class="p">()</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_dictToBNDRecords</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_dictToBNDRecords</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fusion_record</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the two breakends records from the fusion record.</span>

<span class="sd">        :param fusion_record: The fusion record.</span>
<span class="sd">        :type fusion_record: dict</span>
<span class="sd">        :return: The breakend record for the 5&#39; shard of fusion and the breakend record for the 3&#39; shard.</span>
<span class="sd">        :rtype: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_unstranded</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;strand1(gene/fusion)&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span> <span class="ow">or</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;strand2(gene/fusion)&quot;</span><span class="p">][</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;.&quot;</span>
        <span class="n">event_level_filters</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">arriba_event_level_filters</span><span class="p">)</span>
        <span class="n">filters_from_file</span> <span class="o">=</span> <span class="p">{</span><span class="n">elt</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">elt</span> <span class="ow">in</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;filters&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)}</span>
        <span class="n">filters_field</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">filters_from_file</span> <span class="o">&amp;</span> <span class="n">event_level_filters</span><span class="p">))</span>
        <span class="n">read_level_filters</span> <span class="o">=</span> <span class="nb">sorted</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">filters_from_file</span> <span class="o">-</span> <span class="n">event_level_filters</span><span class="p">))</span>
        <span class="n">is_imprecise</span> <span class="o">=</span> <span class="kc">False</span> <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;split_reads1&quot;</span><span class="p">])</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;split_reads2&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="k">else</span> <span class="kc">True</span>
        <span class="k">if</span> <span class="n">is_imprecise</span><span class="p">:</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Imprecise&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">is_unstranded</span><span class="p">:</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;Unstranded&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">is_imprecise</span> <span class="ow">or</span> <span class="n">is_unstranded</span><span class="p">:</span>
            <span class="n">filters_field</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">breakends</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">coord_list</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">side</span><span class="p">,</span> <span class="n">mate_side</span> <span class="ow">in</span> <span class="p">[(</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;2&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;2&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">)]:</span>
            <span class="nb">id</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">uuid</span><span class="o">.</span><span class="n">uuid4</span><span class="p">())</span>
            <span class="n">strand</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span> <span class="o">+</span> <span class="n">side</span> <span class="o">+</span> <span class="s2">&quot;(gene/fusion)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">is_unstranded</span><span class="p">:</span>
                <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;+&quot;</span>
                <span class="k">if</span> <span class="n">side</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;direction1&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;downstream&quot;</span><span class="p">:</span>  <span class="c1"># The first in fusion go to the breakpoint: ----&gt;| or |&lt;----</span>
                        <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;direction2&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;upstream&quot;</span><span class="p">:</span>  <span class="c1"># The second in fusion go from the breakpoint: &lt;----| or |----&gt;</span>
                        <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>
            <span class="n">curr_breakpoint</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;breakpoint&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="n">strand</span>
            <span class="n">coord</span> <span class="o">=</span> <span class="n">getCoordDictFromCoordStr</span><span class="p">(</span><span class="n">curr_breakpoint</span><span class="p">)</span>
            <span class="n">coord_list</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">coord</span><span class="p">)</span>
            <span class="c1"># INFO</span>
            <span class="n">has_frameshift</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;in-frame&quot;</span><span class="p">:</span> <span class="s2">&quot;false&quot;</span><span class="p">,</span> <span class="s2">&quot;out-of-frame&quot;</span><span class="p">:</span> <span class="s2">&quot;true&quot;</span><span class="p">,</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span><span class="p">}</span>
            <span class="n">info</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;BND&quot;</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">:</span> <span class="p">[{</span>
                    <span class="s2">&quot;SYMBOL&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;gene&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">],</span>
                    <span class="s2">&quot;STRAND&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span> <span class="o">+</span> <span class="n">side</span> <span class="o">+</span> <span class="s2">&quot;(gene/fusion)&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
                    <span class="s2">&quot;Site&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;site&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">],</span>
                    <span class="s2">&quot;Type&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;GENE_SHARD&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;direction&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;stream&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">),</span>
                    <span class="s2">&quot;FRAMESHIFT&quot;</span><span class="p">:</span> <span class="n">has_frameshift</span><span class="p">[</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;reading_frame&quot;</span><span class="p">]],</span>
                    <span class="s2">&quot;Protein_contig&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;peptide_sequence&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">,</span> <span class="s2">&quot;@&quot;</span><span class="p">)</span>
                <span class="p">}]</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">is_imprecise</span><span class="p">:</span>
                <span class="n">info</span><span class="p">[</span><span class="s2">&quot;IMPRECISE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="n">is_unstranded</span><span class="p">:</span>
                <span class="n">info</span><span class="p">[</span><span class="s2">&quot;UNSTRANDED&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;closest_genomic_breakpoint&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
                <span class="n">info</span><span class="p">[</span><span class="s2">&quot;GBP&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;closest_genomic_breakpoint&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">]</span>
            <span class="c1"># SAMPLES</span>
            <span class="n">samples_field</span> <span class="o">=</span> <span class="p">{</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s2">&quot;PR&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;discordant_mates&quot;</span><span class="p">]),</span>
                    <span class="s2">&quot;SR&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;split_reads1&quot;</span><span class="p">])</span> <span class="o">+</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;split_reads2&quot;</span><span class="p">]),</span>
                    <span class="s2">&quot;SR1&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;split_reads1&quot;</span><span class="p">]),</span>
                    <span class="s2">&quot;SR2&quot;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;split_reads2&quot;</span><span class="p">]),</span>
                    <span class="s2">&quot;CFD&quot;</span><span class="p">:</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;confidence&quot;</span><span class="p">],</span>
                    <span class="s2">&quot;RFIL&quot;</span><span class="p">:</span> <span class="n">read_level_filters</span>
                <span class="p">}</span>
            <span class="p">}</span>
            <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
                <span class="n">samples_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;DPS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;coverage&quot;</span> <span class="o">+</span> <span class="n">side</span><span class="p">])</span>
            <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;read_identifiers&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
                <span class="n">samples_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SRL&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;read_identifiers&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">side</span> <span class="o">==</span> <span class="s2">&quot;1&quot;</span><span class="p">:</span>
                <span class="n">info</span><span class="p">[</span><span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;fusion_transcript&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;.&quot;</span><span class="p">:</span>
                    <span class="n">info</span><span class="p">[</span><span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fusion_record</span><span class="p">[</span><span class="s2">&quot;fusion_transcript&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;|&quot;</span><span class="p">,</span> <span class="s2">&quot;@&quot;</span><span class="p">)</span>
            <span class="c1"># Record</span>
            <span class="n">breakends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">VCFRecord</span><span class="p">(</span>
                    <span class="n">coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span>
                    <span class="n">coord</span><span class="p">[</span><span class="s2">&quot;pos&quot;</span><span class="p">],</span>
                    <span class="nb">id</span><span class="p">,</span>
                    <span class="s2">&quot;N&quot;</span><span class="p">,</span>  <span class="c1"># ref</span>
                    <span class="p">[],</span>  <span class="c1"># alt</span>
                    <span class="n">pFilter</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;PASS&quot;</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">filters_field</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">filters_field</span><span class="p">,</span>
                    <span class="n">info</span><span class="o">=</span><span class="n">info</span><span class="p">,</span>
                    <span class="n">pFormat</span><span class="o">=</span><span class="nb">sorted</span><span class="p">(</span><span class="n">samples_field</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span>
                    <span class="n">samples</span><span class="o">=</span><span class="n">samples_field</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="n">left_alt</span><span class="p">,</span> <span class="n">right_alt</span> <span class="o">=</span> <span class="n">getAltFromCoord</span><span class="p">(</span><span class="o">*</span><span class="n">coord_list</span><span class="p">)</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alt</span> <span class="o">=</span> <span class="p">[</span><span class="n">left_alt</span><span class="p">]</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">alt</span> <span class="o">=</span> <span class="p">[</span><span class="n">right_alt</span><span class="p">]</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">breakends</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
        <span class="n">breakends</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">breakends</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">id</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">breakends</span>

<div class="viewcode-block" id="ArribaIO.write"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.ArribaIO.write">[docs]</a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">breakends</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write record line in file.</span>

<span class="sd">        :param record: The first breakend and the second breakend in pair.</span>
<span class="sd">        :type record: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_BNDRecordsToDict</span><span class="p">(</span><span class="n">breakends</span><span class="p">))</span></div>

    <span class="k">def</span> <span class="nf">_BNDRecordsToDict</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">breakends</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the Arriba dict from the two breakends of a fusion.</span>

<span class="sd">        :param breakends: The two breakends of a fusion.</span>
<span class="sd">        :type breakends: (anacore.vcf.VCFRecord, anacore.vcf.VCFRecord)</span>
<span class="sd">        :return: The Arriba dict.</span>
<span class="sd">        :rtype: dict</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">first</span><span class="p">,</span> <span class="n">second</span> <span class="o">=</span> <span class="n">breakends</span>
        <span class="n">first_coord</span> <span class="o">=</span> <span class="n">getCoordStr</span><span class="p">(</span><span class="n">first</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="n">second_coord</span> <span class="o">=</span> <span class="n">getCoordStr</span><span class="p">(</span><span class="n">second</span><span class="p">,</span> <span class="kc">False</span><span class="p">)</span>
        <span class="n">fs_to_reading_frame</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;.&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;false&quot;</span><span class="p">:</span> <span class="s2">&quot;in-frame&quot;</span><span class="p">,</span> <span class="s2">&quot;true&quot;</span><span class="p">:</span> <span class="s2">&quot;out-of-frame&quot;</span><span class="p">}</span>
        <span class="n">arriba_filters</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="n">first</span><span class="o">.</span><span class="n">filter</span><span class="p">:</span>
            <span class="n">excluded</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;Imprecise&quot;</span><span class="p">,</span> <span class="s2">&quot;Unstranded&quot;</span><span class="p">}</span>
            <span class="n">arriba_filters</span> <span class="o">=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">filter</span><span class="p">)</span> <span class="o">-</span> <span class="n">excluded</span><span class="p">))</span>
        <span class="k">if</span> <span class="s2">&quot;RFIL&quot;</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span> <span class="ow">and</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;RFIL&quot;</span><span class="p">]:</span>
            <span class="n">arriba_filters</span> <span class="o">+=</span> <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;RFIL&quot;</span><span class="p">])</span>
        <span class="k">return</span> <span class="p">{</span>
            <span class="s2">&quot;gene1&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
            <span class="s2">&quot;gene2&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">],</span>
            <span class="s2">&quot;strand1(gene/fusion)&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">/</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;STRAND&quot;</span><span class="p">],</span>
                <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="s2">&quot;UNSTRANDED&quot;</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span> <span class="k">else</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;strand2(gene/fusion)&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">/</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;STRAND&quot;</span><span class="p">],</span>
                <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="s2">&quot;UNSTRANDED&quot;</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span> <span class="k">else</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
            <span class="p">),</span>
            <span class="s2">&quot;breakpoint1&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">first_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]),</span>
            <span class="s2">&quot;breakpoint2&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2">:</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">],</span> <span class="n">second_coord</span><span class="p">[</span><span class="s2">&quot;chrom&quot;</span><span class="p">]),</span>
            <span class="s2">&quot;site1&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Site&quot;</span><span class="p">],</span>
            <span class="s2">&quot;site2&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Site&quot;</span><span class="p">],</span>
            <span class="s2">&quot;type&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Type&quot;</span><span class="p">],</span>
            <span class="s2">&quot;direction1&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GENE_SHARD&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;stream&quot;</span><span class="p">,</span>
            <span class="s2">&quot;direction2&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;GENE_SHARD&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;stream&quot;</span><span class="p">,</span>
            <span class="s2">&quot;split_reads1&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SR1&quot;</span><span class="p">],</span>
            <span class="s2">&quot;split_reads2&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SR2&quot;</span><span class="p">],</span>
            <span class="s2">&quot;discordant_mates&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;PR&quot;</span><span class="p">],</span>
            <span class="s2">&quot;coverage1&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;DPS&quot;</span><span class="p">],</span>
            <span class="s2">&quot;coverage2&quot;</span><span class="p">:</span> <span class="n">second</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;DPS&quot;</span><span class="p">],</span>
            <span class="s2">&quot;confidence&quot;</span><span class="p">:</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;CFD&quot;</span><span class="p">],</span>
            <span class="s2">&quot;closest_genomic_breakpoint1&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="s2">&quot;GBP&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span> <span class="k">else</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;GBP&quot;</span><span class="p">],</span>
            <span class="s2">&quot;closest_genomic_breakpoint2&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="s2">&quot;GBP&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span> <span class="k">else</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;GBP&quot;</span><span class="p">],</span>
            <span class="s2">&quot;filters&quot;</span><span class="p">:</span> <span class="n">arriba_filters</span><span class="p">,</span>
            <span class="s2">&quot;fusion_transcript&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="s2">&quot;RNA_CONTIG&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span> <span class="k">else</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;@&quot;</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span><span class="p">),</span>
            <span class="s2">&quot;reading_frame&quot;</span><span class="p">:</span> <span class="n">fs_to_reading_frame</span><span class="p">[</span><span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;FRAMESHIFT&quot;</span><span class="p">]],</span>
            <span class="s2">&quot;peptide_sequence&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Protein_contig&quot;</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">first</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">annot_field</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;Protein_contig&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;@&quot;</span><span class="p">,</span> <span class="s2">&quot;|&quot;</span><span class="p">),</span>
            <span class="s2">&quot;read_identifiers&quot;</span><span class="p">:</span> <span class="s2">&quot;.&quot;</span> <span class="k">if</span> <span class="s2">&quot;SRL&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">]</span> <span class="ow">or</span> <span class="nb">len</span><span class="p">(</span><span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SRL&quot;</span><span class="p">])</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">first</span><span class="o">.</span><span class="n">samples</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">sample_name</span><span class="p">][</span><span class="s2">&quot;SRL&quot;</span><span class="p">],</span>
        <span class="p">}</span>

<div class="viewcode-block" id="ArribaIO.isValid"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.ArribaIO.isValid">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return true if the file is in Arriba output format.</span>

<span class="sd">        :param filepath: The file path.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :return: True if the file is in Arriba output format.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">SVIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">title_starter</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">reader</span><span class="p">:</span>
                <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">ArribaIO</span><span class="o">.</span><span class="n">titles</span><span class="p">)</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">reader</span><span class="o">.</span><span class="n">titles</span><span class="p">))</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># All mandatory fields of Arriba are in reader</span>
                    <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">return</span> <span class="n">is_valid</span></div>

<div class="viewcode-block" id="ArribaIO.setVCFHeader"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.ArribaIO.setVCFHeader">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">setVCFHeader</span><span class="p">(</span><span class="n">vcf_io</span><span class="p">,</span> <span class="n">annotation_field</span><span class="o">=</span><span class="s2">&quot;FCANN&quot;</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Set header of an AnnotVCFIO to write Arriba records in VCF output. This function pre-set ANN_titles, info and format used by ArribaIO records.</span>

<span class="sd">        :param vcf_io: The VCF used to write Arriba records.</span>
<span class="sd">        :type vcf_io: anacore.annotVcf.AnnotVCFIO</span>
<span class="sd">        :param annotation_field: INFO field used for store annotations.</span>
<span class="sd">        :type annotation_field: str</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># INFO</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">info</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;GBP&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;GBP&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;The coordinates of the genomic breakpoint which is closest to the transcriptomic breakpoint.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;IMPRECISE&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;IMPRECISE&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Flag&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Imprecise structural variation.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;MATEID&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;MATEID&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;ID of mate breakend.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;RNA_CONTIG&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;The transcript sequence assembled from the supporting reads of the most highly expressed transcript.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;RNA_FIRST&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Flag&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;For RNA fusions, this break-end is 5&#39; in the fusion transcript.&quot;</span><span class="p">),</span>  <span class="c1"># In the fusion transcript, the &#39;LeftGene&#39; is always 5&#39; relative to the &#39;RightGene&#39; (https://groups.google.com/forum/#!topic/star-fusion/9rBLT-v5JHI)</span>
            <span class="s2">&quot;SVTYPE&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;SVTYPE&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Type of structural variant.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;UNSTRANDED&quot;</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="s2">&quot;UNSTRANDED&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Flag&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Unstranded structural variation.&quot;</span><span class="p">),</span>
            <span class="n">annotation_field</span><span class="p">:</span> <span class="n">HeaderInfoAttr</span><span class="p">(</span><span class="n">annotation_field</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Consequence annotations. Format: SYMBOL|STRAND|Site|Type|GENE_SHARD|FRAMESHIFT|Protein_contig&quot;</span><span class="p">)</span>
        <span class="p">}</span>
        <span class="c1"># ANN_titles</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">ANN_titles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;SYMBOL&quot;</span><span class="p">,</span> <span class="s2">&quot;STRAND&quot;</span><span class="p">,</span> <span class="s2">&quot;Site&quot;</span><span class="p">,</span> <span class="s2">&quot;Type&quot;</span><span class="p">,</span> <span class="s2">&quot;GENE_SHARD&quot;</span><span class="p">,</span> <span class="s2">&quot;FRAMESHIFT&quot;</span><span class="p">,</span> <span class="s2">&quot;Protein_contig&quot;</span><span class="p">]</span>
        <span class="c1"># FILTERS</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">filter</span> <span class="o">=</span> <span class="n">arriba_event_level_filters</span>
        <span class="c1"># FORMAT</span>
        <span class="n">vcf_io</span><span class="o">.</span><span class="n">format</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;PR&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;PR&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Number of RNA-Seq fragments that encompass the fusion junction such that one read of the pair aligns to a different gene than the other paired-end read of that fragment.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SR&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SR&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Number of RNA-Seq fragments containing a read that aligns as a split read at the site of the putative fusion junction.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SR1&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SR1&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Number of RNA-Seq fragments containing a read that aligns as a split read at the site of the putative fusion junction with an anchor on first shard.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SR2&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SR2&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Number of RNA-Seq fragments containing a read that aligns as a split read at the site of the putative fusion junction with an anchor on second shard.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;CFD&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;CFD&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Each prediction is assigned one of the confidences low, medium, or high. Several characteristics are taken into account, including: the number of supporting reads, the balance of split reads and discordant mates, the distance between the breakpoints, the type of event, whether the breakpoints are intragenic or not, and whether there are other events which corroborate the prediction, e.g. multiple isoforms or balanced translocations.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;DPS&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;DPS&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;Integer&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Coverage near breakpoint. The coverage is calculated as the number of fragments near the breakpoint on the side of the breakpoint that is retained in the fusion transcript. Note that the coverage calculation counts all fragments (even duplicates).&quot;</span><span class="p">),</span>
            <span class="s2">&quot;RFIL&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;RFIL&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;Filters which removed one or more of the supporting reads. The number of filtered reads is given in parantheses after the name of the filter. If a filter discarded the event as a whole (all reads), the number of filtered reads is missing.&quot;</span><span class="p">),</span>
            <span class="s2">&quot;SRL&quot;</span><span class="p">:</span> <span class="n">HeaderFormatAttr</span><span class="p">(</span><span class="s2">&quot;SRL&quot;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="s2">&quot;String&quot;</span><span class="p">,</span> <span class="n">number</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="s2">&quot;The names of the supporting reads.&quot;</span><span class="p">)</span>
        <span class="p">}</span></div></div>


<div class="viewcode-block" id="BreakendVCFIO"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO">[docs]</a><span class="k">class</span> <span class="nc">BreakendVCFIO</span><span class="p">(</span><span class="n">AnnotVCFIO</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Read and write VCF file containing breakends. Each iteration return a couple of breakends (the first and the second in fusion).&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="n">annot_field</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return instance of BreakendVCFIO.</span>

<span class="sd">        :param filepath: The filepath.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :param mode: Mode to open the file (&#39;r&#39;, &#39;w&#39;, &#39;a&#39;, &#39;i&#39;). The mode &#39;i&#39; allow to open file in indexed mode to fetch by region (tabix).</span>
<span class="sd">        :type mode: str</span>
<span class="sd">        :return: The new instance.</span>
<span class="sd">        :rtype: anacore.fusionVcf.BreakendVCFIO</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sv_rec_by_id</span> <span class="o">=</span> <span class="p">{}</span>  <span class="c1"># By record ID the byte position of start and end of line</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="n">mode</span> <span class="ow">in</span> <span class="p">{</span><span class="s2">&quot;r&quot;</span><span class="p">,</span> <span class="s2">&quot;i&quot;</span><span class="p">}:</span>
            <span class="k">if</span> <span class="n">isGzip</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span> <span class="o">=</span> <span class="n">gzip</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;rt&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">loadIndex</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_iter_already_processed</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">mode</span><span class="p">,</span> <span class="n">annot_field</span><span class="p">)</span>

<div class="viewcode-block" id="BreakendVCFIO.loadIndex"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.loadIndex">[docs]</a>    <span class="k">def</span> <span class="nf">loadIndex</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Parse file and store in index the byte positions (start and end) of each record by record ID.&quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">sv_rec_by_id</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">next_start</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">line_nb</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
        <span class="k">while</span> <span class="n">line</span><span class="p">:</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">next_start</span>
            <span class="n">next_start</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span><span class="o">.</span><span class="n">tell</span><span class="p">()</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="n">end</span> <span class="o">=</span> <span class="n">next_start</span> <span class="o">-</span> <span class="mi">1</span>
                <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="kc">None</span><span class="p">,</span> <span class="mi">3</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">sv_rec_by_id</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">line_nb</span><span class="p">)</span>
            <span class="n">line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="n">line_nb</span> <span class="o">+=</span> <span class="mi">1</span></div>

<div class="viewcode-block" id="BreakendVCFIO.close"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.close">[docs]</a>    <span class="k">def</span> <span class="nf">close</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Close file handle.&quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span><span class="o">.</span><span class="n">close</span><span class="p">()</span></div>

<div class="viewcode-block" id="BreakendVCFIO.get"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.get">[docs]</a>    <span class="k">def</span> <span class="nf">get</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">id</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return record with provided ID.</span>

<span class="sd">        :param id: record ID.</span>
<span class="sd">        :type id: str</span>
<span class="sd">        :return: The record.</span>
<span class="sd">        :rtype: anacore.vcf.VCFRecord</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Get current pos</span>
        <span class="n">current</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;line&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_line</span><span class="p">,</span>
            <span class="s2">&quot;line_nb&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">current_line_nb</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="c1"># Get new record</span>
        <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">,</span> <span class="n">line_nb</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">sv_rec_by_id</span><span class="p">[</span><span class="nb">id</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_line</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">pick_reader</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_line_nb</span> <span class="o">=</span> <span class="n">line_nb</span>
        <span class="n">record</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_parseLine</span><span class="p">()</span>
        <span class="c1"># Jump to previous record</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_line</span> <span class="o">=</span> <span class="n">current</span><span class="p">[</span><span class="s2">&quot;line&quot;</span><span class="p">]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current_line_nb</span> <span class="o">=</span> <span class="n">current</span><span class="p">[</span><span class="s2">&quot;line_nb&quot;</span><span class="p">]</span>
        <span class="c1"># Return</span>
        <span class="k">return</span> <span class="n">record</span></div>

<div class="viewcode-block" id="BreakendVCFIO.getSub"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.getSub">[docs]</a>    <span class="k">def</span> <span class="nf">getSub</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return generator on records overlapping the specified region.</span>

<span class="sd">        .. warning::</span>
<span class="sd">            This method can only be used with an instance of anacore.VCFIO opened</span>
<span class="sd">            with mode &quot;i&quot;.</span>

<span class="sd">        :param chr: Chromosome name of the selected region.</span>
<span class="sd">        :type chr: str</span>
<span class="sd">        :param start: Start of the selected region (1-based).</span>
<span class="sd">        :type start: int</span>
<span class="sd">        :param end: End of the selected region (1-based).</span>
<span class="sd">        :type end: int</span>
<span class="sd">        :return: Fusions (record and his mates) overlappind the specified region.</span>
<span class="sd">        :rtype: generator for (anacore.vcf.VCFRecord, [anacore.vcf.VCFRecord, ...])</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">getSub</span><span class="p">(</span><span class="nb">chr</span><span class="p">,</span> <span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">):</span>
            <span class="n">mates</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">mate_id</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]:</span>
                <span class="n">mates</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mate_id</span><span class="p">))</span>
            <span class="k">yield</span> <span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">mates</span><span class="p">)</span></div>

<div class="viewcode-block" id="BreakendVCFIO.isRecordLine"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.isRecordLine">[docs]</a>    <span class="k">def</span> <span class="nf">isRecordLine</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">line</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return True if the line corresponds to a new record (it is not a comment, an header line or an already processed fusion).</span>

<span class="sd">        :param line: The evaluated line.</span>
<span class="sd">        :type line: str.</span>
<span class="sd">        :return: True if the line corresponds to a record.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_record</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">isRecordLine</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">is_record</span><span class="p">:</span>
            <span class="nb">id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="mi">3</span><span class="p">)[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="n">mate_info</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="mi">8</span><span class="p">)[</span><span class="mi">7</span><span class="p">]</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="n">match</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;MATEID\=([^;]+);?&quot;</span><span class="p">,</span> <span class="n">mate_info</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">match</span><span class="p">:</span>
                <span class="n">mate_id</span> <span class="o">=</span> <span class="n">decodeInfoValue</span><span class="p">(</span><span class="n">match</span><span class="o">.</span><span class="n">groups</span><span class="p">()[</span><span class="mi">0</span><span class="p">])</span>
                <span class="n">fusion_id</span> <span class="o">=</span> <span class="s2">&quot; @@ &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="nb">id</span><span class="p">,</span> <span class="n">mate_id</span><span class="p">]))</span>
                <span class="k">if</span> <span class="n">fusion_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_iter_already_processed</span><span class="p">:</span>  <span class="c1"># Partial pre-filtering for records without multi-alternatives</span>
                    <span class="n">is_record</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">is_record</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="n">is_record</span></div>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__iter__</span><span class="p">():</span>
            <span class="k">for</span> <span class="n">mate_idx</span><span class="p">,</span> <span class="n">mate_id</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]):</span>
                <span class="n">mate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mate_id</span><span class="p">)</span>
                <span class="n">fusion_id</span> <span class="o">=</span> <span class="s2">&quot; @@ &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">sorted</span><span class="p">([</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">mate</span><span class="o">.</span><span class="n">id</span><span class="p">]))</span>
                <span class="k">if</span> <span class="n">fusion_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_iter_already_processed</span><span class="p">:</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_iter_already_processed</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">fusion_id</span><span class="p">)</span>
                    <span class="c1"># Change ID</span>
                    <span class="n">record_new_id</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span>
                    <span class="n">alt_record</span> <span class="o">=</span> <span class="n">record</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">alt</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">record_new_id</span> <span class="o">+=</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">mate_idx</span><span class="p">)</span>  <span class="c1"># Record must be splitted for each mate</span>
                        <span class="n">alt_record</span> <span class="o">=</span> <span class="n">getAlleleRecord</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">record</span><span class="p">,</span> <span class="n">mate_idx</span><span class="p">)</span>
                        <span class="n">alt_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">][</span><span class="n">mate_idx</span><span class="p">]]</span>
                    <span class="n">mate_new_id</span> <span class="o">=</span> <span class="n">mate</span><span class="o">.</span><span class="n">id</span>
                    <span class="n">alt_mate</span> <span class="o">=</span> <span class="n">mate</span>
                    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">mate</span><span class="o">.</span><span class="n">alt</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="n">record_idx</span> <span class="o">=</span> <span class="n">mate</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
                        <span class="n">mate_new_id</span> <span class="o">+=</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">record_idx</span><span class="p">)</span>  <span class="c1"># Record must be splitted for each mate</span>
                        <span class="n">alt_mate</span> <span class="o">=</span> <span class="n">getAlleleRecord</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mate</span><span class="p">,</span> <span class="n">record_idx</span><span class="p">)</span>
                        <span class="n">alt_mate</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">mate</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">][</span><span class="n">record_idx</span><span class="p">]]</span>
                    <span class="n">alt_record</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">record_new_id</span>
                    <span class="n">alt_mate</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">record_new_id</span><span class="p">]</span>
                    <span class="n">alt_mate</span><span class="o">.</span><span class="n">id</span> <span class="o">=</span> <span class="n">mate_new_id</span>
                    <span class="n">alt_record</span><span class="o">.</span><span class="n">info</span><span class="p">[</span><span class="s2">&quot;MATEID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">mate_new_id</span><span class="p">]</span>
                    <span class="c1"># Order by transcript</span>
                    <span class="n">first</span> <span class="o">=</span> <span class="n">alt_record</span>
                    <span class="n">second</span> <span class="o">=</span> <span class="n">alt_mate</span>
                    <span class="k">if</span> <span class="s2">&quot;RNA_FIRST&quot;</span> <span class="ow">in</span> <span class="n">second</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
                        <span class="n">first</span> <span class="o">=</span> <span class="n">alt_mate</span>
                        <span class="n">second</span> <span class="o">=</span> <span class="n">alt_record</span>
                    <span class="c1"># Return</span>
                    <span class="k">yield</span> <span class="n">first</span><span class="p">,</span> <span class="n">second</span>

<div class="viewcode-block" id="BreakendVCFIO.isValid"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.isValid">[docs]</a>    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">isValid</span><span class="p">(</span><span class="n">filepath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return true if the file is in VCF output format.</span>

<span class="sd">        :param filepath: The file path.</span>
<span class="sd">        :type filepath: str</span>
<span class="sd">        :return: True if the file is in VCF output format.</span>
<span class="sd">        :rtype: bool</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="k">with</span> <span class="n">VCFIO</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span> <span class="k">as</span> <span class="n">reader</span><span class="p">:</span>
                <span class="k">if</span> <span class="s2">&quot;SVTYPE&quot;</span> <span class="ow">in</span> <span class="n">reader</span><span class="o">.</span><span class="n">info</span> <span class="ow">and</span> <span class="s2">&quot;MATEID&quot;</span> <span class="ow">in</span> <span class="n">reader</span><span class="o">.</span><span class="n">info</span><span class="p">:</span>
                    <span class="n">is_valid</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
            <span class="k">raise</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">pass</span>
        <span class="k">return</span> <span class="n">is_valid</span></div>

<div class="viewcode-block" id="BreakendVCFIO.write"><a class="viewcode-back" href="../../anacore.html#anacore.fusion.BreakendVCFIO.write">[docs]</a>    <span class="k">def</span> <span class="nf">write</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">first</span><span class="p">,</span> <span class="n">second</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Write fusion record in VCF.</span>

<span class="sd">        :param first: The fisrt breakend.</span>
<span class="sd">        :type first: anacore.vcf.VCFRecord</span>
<span class="sd">        :param second: The second breakend.</span>
<span class="sd">        :type second: anacore.vcf.VCFRecord</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">first</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">second</span><span class="p">)</span></div></div>
</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2019, Frédéric Escudié.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>