Search.setIndex({"docnames": ["anacore", "index"], "filenames": ["anacore.rst", "index.rst"], "titles": ["AnaCore package", "AnaCore - managing standard file formats and objects from NGS"], "terms": {"class": 0, "funcion": 0, "read": [0, 1], "write": [0, 1], "text": 0, "file": 0, "filepath": 0, "mode": 0, "r": 0, "sourc": 0, "object": 0, "inherit": 0, "build": 0, "return": 0, "an": 0, "instanc": 0, "paramet": 0, "str": 0, "The": 0, "open": 0, "w": 0, "new": 0, "type": [0, 1], "close": [0, 1], "handl": 0, "true": 0, "i": [0, 1], "retrun": 0, "bool": 0, "isrecordlin": [0, 1], "line": 0, "correspond": 0, "record": 0, "comment": 0, "header": 0, "evalu": 0, "all": 0, "from": 0, "list": 0, "isempti": [0, 1], "path": 0, "doe": 0, "exist": 0, "empti": 0, "process": 0, "isgzip": [0, 1], "gzipe": 0, "function": 0, "exampl": 0, "import": 0, "annotvcfio": [0, 1], "content": 0, "fileformat": 0, "vcfv4": 0, "3": 0, "info": 0, "id": 0, "csq": 0, "number": 0, "string": 0, "descript": 0, "consequ": 0, "ensembl": 0, "vep": 0, "format": 0, "allel": 0, "impact": 0, "symbol": 0, "hgvsp": 0, "ad": 0, "A": 0, "integ": 0, "depth": 0, "dp": 0, "1": 0, "total": 0, "chrom": 0, "po": 0, "ref": 0, "alt": 0, "qual": 0, "sample_test": 0, "17": 0, "3675200": 0, "c": 0, "g": 0, "4845": 0, "9": 0, "pass": 0, "4560": 0, "1500": 0, "7675200": 0, "31799": 0, "6": 0, "missense_vari": 0, "moder": 0, "tp53": 0, "np_000537": 0, "3ap": 0, "ala138pro": 0, "np_001119584": 0, "print": 0, "alter": 0, "sep": 0, "t": 0, "test": 0, "gz": 0, "annot_field": 0, "reader": 0, "getnam": [0, 1], "els": 0, "result": 0, "p": 0, "ann": 0, "writer": 0, "my_sampl": 0, "self": 0, "headerinfoattr": [0, 1], "af": 0, "headerformatattr": [0, 1], "frequenc": 0, "float": 0, "writehead": [0, 1], "vcf_record_list": 0, "0": 0, "33": 0, "vcfio": [0, 1], "manag": 0, "contain": [0, 1], "variant": 0, "allow": 0, "index": [0, 1], "fetch": 0, "tabix": 0, "tag": 0, "field": 0, "us": 0, "store": 0, "default": 0, "copyhead": [0, 1], "model": 0, "copi": 0, "specifi": 0, "rectovcflin": [0, 1], "vcfrecord": [0, 1], "vepvcfio": [0, 1], "produc": 0, "code": 0, "bedio": [0, 1], "isvalid": [0, 1], "txt": 0, "name": 0, "refer": 0, "start": 0, "end": [0, 1], "strand": 0, "chr1": 0, "158": 0, "367": 0, "target_01": 0, "54899": 0, "54999": 0, "target_02": 0, "whole": 0, "getsortedareasbychr": [0, 1], "areas_by_chr": 0, "areas_on_curr_chr": 0, "item": [0, 1], "curr_area": 0, "bedrecord": [0, 1], "column": 0, "curr_region": 0, "recfromregion": [0, 1], "157": 0, "54898": 0, "write_nb_col": 0, "http": 0, "genom": 0, "ucsc": 0, "edu": 0, "faq": 0, "faqformat": 0, "html": 0, "format1": 0, "int": 0, "when": 0, "bedrecordtobedlin": [0, 1], "static": 0, "getmaxnbcol": [0, 1], "in_b": 0, "maximum": 0, "per": 0, "must": 0, "consist": 0, "throughout": 0, "ani": 0, "singl": 0, "set": 0, "data": 0, "track": 0, "first": 0, "can": 0, "none": 0, "score": 0, "thickstart": 0, "thickend": 0, "itemrgb": 0, "blockcount": 0, "blocksiz": 0, "blockstart": 0, "chromosom": 0, "which": 0, "ha": 0, "been": 0, "defin": 0, "posit": 0, "thi": 0, "ascend": 0, "between": 0, "1000": 0, "featur": 0, "drawn": 0, "thickli": 0, "codon": 0, "gene": [0, 1], "displai": 0, "stop": 0, "rgb": 0, "valu": [0, 1], "form": 0, "b": 0, "e": 0, "255": 0, "block": 0, "exon": [0, 1], "size": 0, "should": 0, "calcul": 0, "rel": 0, "chromstart": 0, "region_record": 0, "follow": 0, "getarea": [0, 1], "area": 0, "regionlist": [0, 1], "getareasbychr": [0, 1], "each": 0, "dict": 0, "sort": 0, "util": 0, "work": 0, "human": 0, "assemblyaccess": 0, "assembli": 0, "compar": 0, "two": 0, "differ": 0, "first_assembli": 0, "gcf_000001405": 0, "refseq": 0, "first_assembly_std": 0, "tohumannam": 0, "grch37": 0, "second_assembli": 0, "hg19": 0, "second_assembly_std": 0, "second_chr": 0, "same": 0, "human_by_acc": 0, "gca_000001405": 0, "10": 0, "11": 0, "12": 0, "13": 0, "14": 0, "15": 0, "grch38": 0, "16": 0, "18": 0, "19": 0, "2": 0, "20": 0, "21": 0, "22": 0, "23": 0, "24": 0, "25": 0, "26": 0, "27": 0, "28": 0, "4": 0, "5": 0, "7": 0, "8": 0, "grch36": 0, "29": 0, "30": 0, "31": 0, "32": 0, "34": 0, "35": 0, "36": 0, "37": 0, "38": 0, "39": 0, "grcb36": 0, "grcb37": 0, "hg18": 0, "hg38": 0, "ncbi36": 0, "human_by_genbank": 0, "human_by_refseq": 0, "assembly_nam": 0, "standard": 0, "chraccess": 0, "first_grch38_po": 0, "nc_000001": 0, "154545": 0, "first_chr": 0, "first_po": 0, "split": 0, "first_chr_std": 0, "second_grch38_po": 0, "second_po": 0, "second_chr_std": 0, "cm000663": 0, "cm000664": 0, "cm000665": 0, "cm000666": 0, "cm000667": 0, "cm000668": 0, "cm000669": 0, "cm000670": 0, "cm000671": 0, "cm000672": 0, "cm000673": 0, "cm000674": 0, "cm000675": 0, "cm000676": 0, "cm000677": 0, "cm000678": 0, "cm000679": 0, "cm000680": 0, "cm000681": 0, "cm000682": 0, "cm000683": 0, "cm000684": 0, "cm000685": 0, "x": 0, "cm000686": 0, "y": 0, "j01415": 0, "mt": 0, "m": 0, "nc_000002": 0, "nc_000003": 0, "nc_000004": 0, "nc_000005": 0, "nc_000006": 0, "nc_000007": 0, "nc_000008": 0, "nc_000009": 0, "nc_000010": 0, "nc_000011": 0, "nc_000012": 0, "nc_000013": 0, "nc_000014": 0, "nc_000015": 0, "nc_000016": 0, "nc_000017": 0, "nc_000018": 0, "nc_000019": 0, "nc_000020": 0, "nc_000021": 0, "nc_000022": 0, "nc_000023": 0, "nc_000024": 0, "nc_0018007": 0, "nc_012920": 0, "chr10": 0, "chr11": 0, "chr12": 0, "chr13": 0, "chr14": 0, "chr15": 0, "chr16": 0, "chr17": 0, "chr18": 0, "chr19": 0, "chr2": 0, "chr20": 0, "chr21": 0, "chr22": 0, "chr3": 0, "chr4": 0, "chr5": 0, "chr6": 0, "chr7": 0, "chr8": 0, "chr9": 0, "chrm": 0, "chrmt": 0, "chrx": 0, "chry": 0, "human_by_human_chr": 0, "human_by_human_wo_chr": 0, "chr_name": 0, "complex": 0, "like": 0, "includ": 0, "combin": 0, "oper": [0, 1], "getter": 0, "aggreg": [0, 1], "ratio": 0, "action": 0, "select": 0, "one": 0, "patient": 0, "ag": 0, "treatment": 0, "placebo": 0, "group": 0, "20ng": 0, "ip": 0, "d": [0, 1], "pc": 0, "children": 0, "children_filt": 0, "curr": 0, "eval": [0, 1], "treatment_filt": 0, "exclud": 0, "least": 0, "group_filt": 0, "nb": 0, "filterscombin": [0, 1], "": 0, "threshold": 0, "callabl": 0, "properti": 0, "appli": 0, "get": [0, 1], "see": 0, "setgett": [0, 1], "rule": 0, "It": 0, "determin": 0, "element": 0, "fit": 0, "ar": 0, "kept": 0, "reject": 0, "fromdict": [0, 1], "filter_desc": 0, "dictionari": 0, "getrecordvalu": [], "_getter": [], "setaggreg": [], "chang": 0, "have": 0, "ok": [], "superior": 0, "than": 0, "setfct": [0, 1], "accord": 0, "retriev": 0, "take": 0, "If": 0, "himself": 0, "call": 0, "compat": 0, "todict": [0, 1], "pars": 0, "specif": 0, "setoper": [], "wrap": 0, "uniqu": 0, "sever": 0, "age_filt": 0, "young_with_treat": 0, "With": 0, "valid": 0, "xor": 0, "onli": 0, "todo": 0, "regard": 0, "filtersfromdict": [0, 1], "descriptor": 0, "caller": 0, "output": 0, "convert": 0, "star": 0, "fusionnam": 0, "junctionreadcount": 0, "spanningfragcount": 0, "splicetyp": 0, "leftgen": 0, "leftbreakpoint": 0, "rightgen": 0, "rightbreakpoint": 0, "largeanchorsupport": 0, "ffpm": 0, "leftbreakdinuc": 0, "leftbreakentropi": 0, "rightbreakdinuc": 0, "rightbreakentropi": 0, "mn1": 0, "bend2": 0, "90": 0, "only_ref_splic": 0, "ensg00000169184": 0, "27796763": 0, "ensg00000177324": 0, "18195442": 0, "yes_lda": 0, "9246": 0, "gt": 0, "9656": 0, "7232": 0, "interchromosom": 0, "fusionfileread": [0, 1], "spl_name": 0, "spl1": 0, "breakpoint_1": 0, "tgene_brkpt_1": 0, "tbreakpoint_2": 0, "tgene_brkpt_2": 0, "tspanning_pair": 0, "tsplit_read": 0, "factori": [0, 1], "tsv": 0, "sample_nam": 0, "second": 0, "frist": 0, "join": 0, "elt": 0, "pr": 0, "sr": 0, "gene_brkpt_1": 0, "breakpoint_2": 0, "gene_brkpt_2": 0, "spanning_pair": 0, "split_read": 0, "99": 0, "mateid": 0, "mate": 0, "breakend": 0, "svtype": 0, "structur": 0, "321682": 0, "bnd_v": 0, "123456": 0, "bnd": 0, "bnd_u": 0, "888888": 0, "198983": 0, "bnd_z": 0, "tbreakpoint_1_nam": 0, "tbreakpoint_2_nam": 0, "breakpoint_1_nam": 0, "breakpoint_2_nam": 0, "bnd_u_0": 0, "bnd_u_1": 0, "arribaio": [0, 1], "fcann": 0, "hashedsvio": [0, 1], "arriba": 0, "github": 0, "com": 0, "suhrig": 0, "blob": 0, "master": 0, "document": 0, "md": 0, "repres": 0, "shard": 0, "setvcfhead": [0, 1], "vcf_io": 0, "annotation_field": 0, "pre": 0, "ann_titl": 0, "titl": [0, 1], "gene1": 0, "gene2": 0, "strand1": 0, "strand2": 0, "breakpoint1": 0, "breakpoint2": 0, "site1": 0, "site2": 0, "direction1": 0, "direction2": 0, "split_reads1": 0, "split_reads2": 0, "discordant_m": 0, "coverage1": 0, "coverage2": 0, "confid": 0, "closest_genomic_breakpoint1": 0, "closest_genomic_breakpoint2": 0, "fusion_transcript": 0, "reading_fram": 0, "peptide_sequ": 0, "read_identifi": 0, "pair": 0, "breakendvcfio": [0, 1], "iter": 0, "coupl": 0, "fusionvcf": 0, "provid": 0, "getsub": [0, 1], "chr": 0, "gener": 0, "overlap": 0, "method": 0, "hi": 0, "overlappind": 0, "alreadi": 0, "loadindex": [0, 1], "byte": 0, "fisrt": 0, "fusioncatcherio": [0, 1], "fusioncatch": 0, "ndaniel": 0, "doc": 0, "manual": 0, "gene_1_symbol": 0, "5end_fusion_partn": 0, "gene_2_symbol": 0, "3end_fusion_partn": 0, "fusion_descript": 0, "counts_of_common_mapping_read": 0, "spanning_unique_read": 0, "longest_anchor_found": 0, "fusion_finding_method": 0, "fusion_point_for_gene_1": 0, "fusion_point_for_gene_2": 0, "gene_1_id": 0, "gene_2_id": 0, "exon_1_id": 0, "exon_2_id": 0, "fusion_sequ": 0, "predicted_effect": 0, "identifi": 0, "dedic": 0, "softwar": 0, "arg": 0, "kwarg": 0, "parser": 0, "addit": 0, "argument": 0, "keyword": 0, "starfusionio": [0, 1], "wiki": 0, "junctionread": 0, "spanningfrag": 0, "getaltfromcoord": [0, 1], "first_coord": 0, "second_coord": 0, "altern": 0, "coordin": 0, "getbndinterv": [0, 1], "cipo": 0, "taken": 0, "account": 0, "getcoorddictfromcoordstr": [0, 1], "coord": 0, "getcoordstr": [0, 1], "is_first": 0, "part": 0, "rna": 0, "boolean": 0, "getstrand": [0, 1], "manipul": 0, "transcript": [0, 1], "protein": [0, 1], "regul": 0, "cd": [0, 1], "parent": 0, "regiontre": [0, 1], "sub": 0, "intron": [0, 1], "aalength": [0, 1], "length": [0, 1], "amino": 0, "acid": 0, "sum": 0, "eval_region": 0, "getcdsfromtranscript": [0, 1], "order": 0, "getcodoninfo": [0, 1], "ref_po": 0, "sequence_read": 0, "idxfastaio": [0, 1], "getcodonrefpo": [0, 1], "aa_po": 0, "nt": 0, "getcodonseqfromprotpo": [0, 1], "nucleotid": 0, "getntposfromrefpo": [0, 1], "chr_po": 0, "utr": 0, "getntposfromregionpo": [0, 1], "protein_po": 0, "codon_po": 0, "getposonref": [0, 1], "getposonregion": [0, 1], "getsubfromregionpo": [0, 1], "its": 0, "hasoverlap": [0, 1], "operon": 0, "readthrough": 0, "addprotein": [0, 1], "add": 0, "link": 0, "delprotein": [0, 1], "remov": 0, "transcript_po": 0, "getsubfromrefpo": [0, 1], "where": 0, "locat": 0, "gff3io": [0, 1], "separ": 0, "title_start": 0, "has_titl": 0, "gff3": 0, "decodedvalu": [0, 1], "encoded_valu": 0, "attribut": 0, "readabl": 0, "decod": 0, "rfc": 0, "3986": 0, "percent": 0, "encod": 0, "encodedvalu": [0, 1], "dirty_valu": 0, "after": 0, "clean": 0, "url": 0, "escap": 0, "charact": 0, "space": 0, "tab": 0, "replac": 0, "quot": 0, "delet": 0, "gff3record": [0, 1], "seq_id": 0, "phase": 0, "comind": 0, "landmark": 0, "establish": 0, "system": 0, "current": 0, "free": 0, "qualifi": 0, "intend": 0, "describ": 0, "algorithm": 0, "procedur": 0, "typic": 0, "piec": 0, "genescan": 0, "databas": 0, "genbank": 0, "In": 0, "effect": 0, "extend": 0, "ontologi": 0, "creat": 0, "composit": 0, "subclass": 0, "previous": 0, "constrain": 0, "either": 0, "term": 0, "lite": 0, "sofa": 0, "latter": 0, "distinguish": 0, "syntax": 0, "so": 0, "000000": 0, "given": 0, "alwai": 0, "less": 0, "equal": 0, "As": 0, "earlier": 0, "version": 0, "semant": 0, "ill": 0, "strongli": 0, "recommend": 0, "similar": 0, "ab": 0, "initio": 0, "predict": 0, "For": 0, "indic": 0, "begin": 0, "frame": 0, "reach": 0, "next": 0, "other": 0, "word": [0, 1], "third": 0, "NOT": 0, "confus": 0, "simpli": 0, "modulo": 0, "forward": 0, "count": 0, "revers": 0, "addtoannot": [0, 1], "annott": 0, "togff": [0, 1], "corrspond": 0, "gtfio": [0, 1], "recordtolin": [0, 1], "loadmodel": [0, 1], "gtf_path": 0, "feature_handl": 0, "restrict_to": 0, "tree": 0, "directli": 0, "author": 0, "retrict": 0, "seqnam": 0, "mutalyz": 0, "ncbi": 0, "np_001245144": 0, "inform": 0, "acces": 0, "fromstr": [0, 1], "hgvs_str": 0, "ispredict": [0, 1], "come": 0, "opposit": 0, "experimet": 0, "evid": 0, "hgvsprotchang": [0, 1], "start_aa": [0, 1], "start_po": 0, "end_aa": [0, 1], "end_po": 0, "new_aa": [0, 1], "evt": 0, "new_elt": 0, "proteic": 0, "ex": 0, "val600glu": 0, "val": 0, "gln": 0, "gln746_lys747in": 0, "63": 0, "il": 0, "ile327argf": 0, "600": 0, "746": 0, "327": 0, "last": 0, "ly": 0, "747": 0, "f": 0, "ext": 0, "even": 0, "del": 0, "dup": 0, "ins": 0, "insdel": 0, "glu": 0, "without": 0, "experiment": 0, "analys": 0, "aa_nb_lett": 0, "letter": 0, "By": 0, "tri": 0, "automat": 0, "inscouldbeident": [0, 1], "hgvs_in": 0, "hgvs_repeat": 0, "insert": 0, "could": 0, "ident": 0, "repeat": 0, "l16_i17insgttl": 0, "v": 0, "g13_l16dup": 0, "case": 0, "bewtenn": 0, "notat": 0, "g13_l16": 0, "isinframein": [0, 1], "duplic": 0, "frameshift": 0, "extens": 0, "fals": 0, "val582_asn583insxxxxx": 0, "mutalyzerbatch": [0, 1], "queri": 0, "nl": 0, "command": 0, "namecheck": 0, "proxy_url": 0, "submiss": 0, "batch": 0, "server": 0, "mutalyzer_url": 0, "25245351c": 0, "ng_012337": 0, "sdhd_v001": 0, "274g": 0, "curr_r": 0, "syncrequest": [0, 1], "re": 0, "job": 0, "choos": 0, "syntaxcheck": 0, "positionconvert": 0, "snpconvert": 0, "proxi": 0, "necessari": 0, "getexectim": [0, 1], "execut": 0, "time": 0, "getparsedrespons": [0, 1], "time_to_upd": 0, "submit": [0, 1], "wait": 0, "befor": 0, "retri": 0, "respons": 0, "statu": [0, 1], "becom": 0, "complet": 0, "few": 0, "updat": 0, "getrequesturl": [0, 1], "request": 0, "getrespons": [0, 1], "first_time_to_upd": 0, "updatestatu": [0, 1], "progress": 0, "runmutalyzerdescript": [0, 1], "description_data": 0, "proteindescript": 0, "transcriptdescript": 0, "runmutalyz": 0, "light": 0, "runmutalyzerlegend": [0, 1], "getbyaccess": [0, 1], "id_by_nam": 0, "kit_v001": 0, "nm_000222": 0, "kit_i001": 0, "np_000213": 0, "legend_data": 0, "legend": 0, "getidbynam": [0, 1], "getprotbytr": [0, 1], "mafio": [0, 1], "mutat": 0, "gdc": 0, "cancer": [0, 1], "gov": 0, "file_format": 0, "maf_format": 0, "2d": 0, "distancematrixio": [0, 1], "cast_fct": 0, "distanc": 0, "fct": 0, "cast": 0, "msiannot": [0, 1], "introduc": 0, "addlocirestospl": [0, 1], "msi_spl": 0, "data_by_locu": 0, "addlociresult": [0, 1], "msi_locu": 0, "data_by_r": 0, "getcastedvalu": [0, 1], "type_nam": 0, "getlocusannotdict": [0, 1], "in_annot": 0, "locusclassifi": [0, 1], "locus_id": 0, "method_nam": 0, "classifi": 0, "model_method_nam": 0, "data_method_nam": 0, "msisampl": [0, 1], "synopsi": 0, "clf": 0, "train_dataset": 0, "test_dataset": 0, "predict_proba": 0, "set_statu": 0, "loci": 0, "target": 0, "analysi": 0, "sklearn": 0, "expect": 0, "distribut": 0, "extract": 0, "train": 0, "locusr": [0, 1], "ecpect": 0, "These": 0, "eaxmpl": 0, "np": 0, "arrai": 0, "estim": 0, "probabl": 0, "msisplr": [0, 1], "librari": [0, 1], "authorizedvalu": 0, "stabil": 0, "stabl": 0, "mss": 0, "undetermin": 0, "unstabl": 0, "getnbsupport": [0, 1], "report": 0, "in_report": 0, "kei": [0, 1], "locus_nam": 0, "support": 0, "msi_object": 0, "retun": 0, "json": 0, "dump": 0, "recurs": 0, "hubblediff": [0, 1], "ref_path": 0, "differenti": 0, "fasta": [0, 1], "hubbledist": [0, 1], "len_limit": 0, "100": 0, "max": 0, "length_distribut": 0, "nbrepeatfromstart": [0, 1], "ref_read": 0, "motif": 0, "occurr": 0, "pattern": 0, "parsehubbleresult": [0, 1], "sequences_path": 0, "summary_path": 0, "differential_path": 0, "distributions_path": 0, "unstable_rate_threshold": 0, "about": 0, "microsatellit": 0, "summari": 0, "diffrenti": 0, "over": 0, "rate": 0, "msispl": 0, "nr21": 0, "delresult": 0, "result_id": 0, "hash": 0, "load": 0, "locusdatadistrib": [0, 1], "ct_by_len": 0, "fragment": 0, "fromdens": 0, "dens": 0, "li": 0, "more": 0, "often": 0, "getcount": 0, "getdensecount": 0, "absolut": 0, "betwen": 0, "also": 0, "minimum": 0, "getdenseprct": 0, "percentag": 0, "reduc": 0, "getmaxlength": 0, "getminlength": 0, "getmostrepres": 0, "most": 0, "larger": 0, "getnbpeak": 0, "min_count": 0, "peak": 0, "lower": 0, "tupl": 0, "spars": 0, "anlysi": 0, "getrefseqinfo": [0, 1], "ref_fh": 0, "microsat": 0, "flank_siz": 0, "flank": 0, "occur": 0, "binari": 0, "msingsanalysisio": [0, 1], "analyz": 0, "bitbucket": 0, "org": 0, "uwlabm": 0, "msingsanalysi": 0, "msingsev": [0, 1], "locus_distrib": 0, "highest_peak_ratio": 0, "05": 0, "height": 0, "peak_height_cutoff": 0, "highest_peak_height": 0, "highest": 0, "consid": 0, "getthresholdfromnbpeak": 0, "models_nb_peak": 0, "std_dev_rat": 0, "popul": 0, "deviat": 0, "around": 0, "averag": 0, "msingsreport": [0, 1], "count_msi_sampl": 0, "msisensor": 0, "pro": 0, "baselineio": [0, 1], "baselin": 0, "baselinerecord": [0, 1], "repeat_tim": 0, "repeat_unit_bas": 0, "left_flank_bas": 0, "right_flank_bas": 0, "support_sampl": 0, "unit": 0, "frommodel": 0, "model_method": 0, "left_flank_binari": 0, "represent": 0, "repeat_unit_binari": 0, "repeat_unit_length": 0, "right_flank_binari": 0, "distio": [0, 1], "nextlocu": 0, "metadata": 0, "meta": 0, "proeval": [0, 1], "getslippagescor": 0, "ref_len": 0, "getthreshold": 0, "getthresholdfromscor": 0, "models_scor": 0, "ddof": 0, "due": 0, "point": 0, "mean": 0, "delta": 0, "degre": 0, "freedom": 0, "divisor": 0, "n": 0, "proio": [0, 1], "present": 0, "pro_p_threshold": 0, "binarytostr": [0, 1], "len": 0, "origin": 0, "parseproresult": [0, 1], "all_path": 0, "determined_rate_threshold": 0, "min_support": 0, "rsult": 0, "di": 0, "under": 0, "cannot": 0, "ist": 0, "stringtobinari": [0, 1], "getincompletemodel": 0, "in_path": 0, "min_support_model": 0, "unsuffici": 0, "msireport": 0, "msi_sampl": 0, "out_path": 0, "context": 0, "instabl": 0, "level": 0, "addlocu": 0, "delloci": 0, "dellocu": 0, "getlocimethod": 0, "getnbdetermin": 0, "getnbloci": 0, "getnbprocess": 0, "getnbstabl": 0, "getnbundetermin": 0, "getnbunst": 0, "setscor": 0, "undetermined_weight": 0, "locus_weight_is_scor": 0, "formula": 0, "nb_loci_undetermin": 0, "weight": 0, "wheight": 0, "setstatusbyinstabilityratio": 0, "min_voting_loci": 0, "instability_threshold": 0, "vote": 0, "param": 0, "parent_nod": 0, "children_nod": 0, "addchild": [0, 1], "child": 0, "fromclusternod": [0, 1], "id_to_nam": 0, "distance_tag": 0, "dist": 0, "_parent_dist_from_root": 0, "_root_dist_from_leav": 0, "scipi": 0, "cluster": 0, "hierarchi": 0, "clusternod": 0, "numpi": 0, "root": 0, "view": 0, "leav": 0, "other_nod": 0, "option": 0, "getancestor": [0, 1], "ancestor": 0, "getchildbynam": [0, 1], "search": 0, "getdepth": [0, 1], "branch": 0, "getdescend": [0, 1], "descend": 0, "grandchildren": 0, "getleav": [0, 1], "haschild": [0, 1], "toextendednewick": [0, 1], "newick": [0, 1], "represet": 0, "tonewick": [0, 1], "picard": 0, "tool": 0, "picardread": [0, 1], "castcol": [0, 1], "col_typ": 0, "among": 0, "getcoltyp": [0, 1], "getvaltyp": [0, 1], "interval_str": 0, "interv": 0, "15746211": 0, "5746311": 0, "ref_nam": 0, "getcoordinatesstr": [0, 1], "printabl": 0, "getmindist": [0, 1], "region_po": 0, "hasstrandedoverlap": [0, 1], "setrefer": [0, 1], "strandedcontain": [0, 1], "getcontain": [0, 1], "getnearest": [0, 1], "select_fct": 0, "nearest": 0, "getoverlap": [0, 1], "hierarch": 0, "relat": 0, "sortchildren": [0, 1], "apparit": 0, "siblings_idx": 0, "consolid": [0, 1], "merge_contigu": 0, "trace": 0, "contigu": 0, "merg": 0, "merged_region": 0, "merge_traceback": 0, "iteroverlap": [0, 1], "subject": 0, "check_ref": 0, "you": 0, "thei": 0, "check": 0, "ensur": 0, "care": 0, "gain": 0, "comput": 0, "certain": 0, "iteroverlappedbyregion": [0, 1], "queries_by_chr": 0, "subject_by_chr": 0, "mergedregion": [0, 1], "One": 0, "splittedbyref": [0, 1], "region_list": 0, "aa3lettersalphabet": [0, 1], "alphabet": [0, 1], "three": 0, "codons_by_aa": [0, 1], "ala": 0, "gct": 0, "gcc": 0, "gca": 0, "gcg": 0, "cgt": 0, "cgc": 0, "cga": 0, "cgg": 0, "aga": 0, "agg": 0, "asn": 0, "aat": 0, "aac": 0, "asp": 0, "gat": 0, "gac": 0, "cy": 0, "tgt": 0, "tgc": 0, "caa": 0, "cag": 0, "gaa": 0, "gag": 0, "gly": 0, "ggt": 0, "ggc": 0, "gga": 0, "ggg": 0, "cat": 0, "cac": 0, "att": 0, "atc": 0, "ata": 0, "leu": 0, "ctt": 0, "ctc": 0, "cta": 0, "ctg": 0, "tta": 0, "ttg": 0, "aaa": 0, "aag": 0, "met": 0, "atg": 0, "phe": 0, "ttt": 0, "ttc": 0, "cct": 0, "ccc": 0, "cca": 0, "ccg": 0, "ser": 0, "tct": 0, "tcc": 0, "tca": 0, "tcg": 0, "agt": 0, "agc": 0, "ter": 0, "taa": 0, "tga": 0, "thr": 0, "act": 0, "acc": 0, "aca": 0, "acg": 0, "trp": 0, "tgg": 0, "tyr": 0, "tat": 0, "tac": 0, "gtt": 0, "gtc": 0, "gta": 0, "gtg": 0, "seq": 0, "one_by_thre": [0, 1], "asx": 0, "q": 0, "glx": 0, "z": 0, "h": 0, "l": [0, 1], "k": 0, "pyl": 0, "o": 0, "sec": 0, "u": 0, "xaa": 0, "xle": 0, "j": 0, "toonelett": [0, 1], "codonalphabet": [0, 1], "aa_by_codon": [0, 1], "translat": [0, 1], "dnaalphabet": [0, 1], "dna": 0, "complement": [0, 1], "revcom": [0, 1], "rnaalphabet": [0, 1], "qualiti": 0, "dnarevcom": [0, 1], "rnarevcom": [0, 1], "getshortestrepeatunit": [0, 1], "shortest": 0, "reproduc": 0, "entir": 0, "faidx": [0, 1], "svio": [0, 1], "www": 0, "htslib": 0, "readbyid": [0, 1], "faiitem": 0, "faidxrecord": [0, 1], "offset": 0, "line_bas": 0, "line_width": 0, "qual_offset": 0, "fastq": 0, "newlin": 0, "within": 0, "fastaio": [0, 1], "en": 0, "wikipedia": 0, "fasta_format": 0, "nbseq": [0, 1], "nbseqandnt": [0, 1], "nextseq": [0, 1], "seqtofastalin": [0, 1], "sequence_record": 0, "fastqio": [0, 1], "fastq_format": 0, "qualoffset": [0, 1], "sanger": 0, "64": 0, "solexa": 0, "seqtofastqlin": [0, 1], "use_cach": 0, "fai_path": 0, "memori": 0, "speed": 0, "up": 0, "success": 0, "fai": 0, "out": 0, "sequencefileread": [0, 1], "getstrandedseqfrompo": [0, 1], "pos_on_ref": 0, "complement_alphabet": 0, "continu": 0, "pos_list": 0, "unsort": 0, "discontinu": 0, "log": 0, "jan": 0, "06": 0, "04": 0, "49": 0, "map": 0, "46": 0, "finish": 0, "million": 0, "hour": 0, "237": 0, "83": 0, "input": 0, "13477055": 0, "150": 0, "7988755": 0, "59": 0, "47": 0, "splice": 0, "2900295": 0, "sjdb": 0, "2866020": 0, "2870546": 0, "gc": 0, "22895": 0, "AT": 0, "ac": 0, "2759": 0, "non": 0, "canon": 0, "4095": 0, "mismatch": 0, "01": 0, "00": 0, "41": 0, "multi": 0, "multipl": 0, "4957483": 0, "78": 0, "too": 0, "mani": 0, "23143": 0, "unmap": 0, "short": 0, "72": 0, "csv": 0, "metadata_start": 0, "etc": 0, "row": 0, "title_metadata": 0, "tophatfusionio": [0, 1], "ccb": 0, "jhu": 0, "tophat": 0, "fusion_manu": 0, "shtml": 0, "alt_freq": 0, "getaf": [0, 1], "curr_spl": 0, "n01": 0, "n02": 0, "n03": 0, "128": 0, "q10": 0, "headerfilterattr": [0, 1], "below": 0, "dbsnp": 0, "membership": 0, "129": 0, "flag": 0, "extra_head": 0, "myimputationprogramv3": 0, "partial": 0, "headerattr": [0, 1], "headerdescattr": [0, 1], "requir": 0, "headertypedattr": [0, 1], "bit": 0, "sign": 0, "ieee": 0, "754": 0, "char": 0, "headersampleattr": [0, 1], "knownsnpid": 0, "refallel": 0, "altallel": 0, "pfilter": 0, "pformat": 0, "contig": 0, "semi": 0, "colon": 0, "avail": 0, "possibl": 0, "state": 0, "wa": 0, "were": 0, "filter_nam": 0, "additionn": 0, "containsindel": [0, 1], "note": 0, "substitut": 0, "aa": 0, "doubl": 0, "fastdownstream": [0, 1], "seq_handl": 0, "pad": 0, "500": 0, "simplifi": 0, "move": 0, "downstream": 0, "postit": 0, "inspect": 0, "upstream": 0, "movement": 0, "limit": 0, "faststandard": [0, 1], "unecessari": 0, "concern": 0, "getad": [0, 1], "getafbysampl": [0, 1], "missing_replac": 0, "miss": 0, "getaltad": [0, 1], "getaltaf": [0, 1], "getdp": [0, 1], "getemptyallelemark": [0, 1], "marker": 0, "intern": 0, "getmostdownstream": [0, 1], "ref_seq": 0, "normal": 0, "normalizesingleallel": [0, 1], "warn": 0, "getmostupstream": [0, 1], "getpopaltad": [0, 1], "compos": 0, "getpopaltaf": [0, 1], "getpopdp": [0, 1], "getpoprefad": [0, 1], "getpoprefaf": [0, 1], "isdelet": [0, 1], "isindel": [0, 1], "isinsanddel": [0, 1], "isinsert": [0, 1], "empty_mark": 0, "minim": 0, "tg": 0, "aagc": 0, "atac": 0, "ta": 0, "refend": [0, 1], "affect": 0, "std": 0, "aagt": 0, "aatc": 0, "aacgt": 0, "tc": 0, "tt": 0, "aaat": 0, "refstart": [0, 1], "snp": 0, "indel": 0, "variat": 0, "decodeinfovalu": [0, 1], "special": 0, "unescap": 0, "3a": 0, "3b": 0, "3d": 0, "2c": 0, "encodeinfovalu": [0, 1], "getallelerecord": [0, 1], "fh_vcf": 0, "idx_alt": 0, "handler": 0, "getfreqmatrix": [0, 1], "vcf_path": 0, "accept_miss": 0, "nan": 0, "found": 0, "except": 0, "rais": 0, "getheaderattr": [0, 1], "header_lin": 0, "declar": 0, "anapath": 1, "core": 1, "packag": 1, "bed": 1, "gff": 1, "gtf": 1, "maf": 1, "vcf": 1, "pip": 1, "abstractfil": 1, "annotvcf": 1, "db": 1, "homo_sapien": 1, "access": 1, "filter": 1, "fusion": 1, "genomicregion": 1, "hgv": 1, "illumina": 1, "matrix": 1, "msi": 1, "annot": 1, "base": 1, "hubbl": 1, "locu": 1, "msing": 1, "msisensorpro": 1, "reportio": 1, "sampl": 1, "node": 1, "picardio": 1, "region": 1, "sequenc": 1, "sequenceio": 1, "starlog": 1, "sv": 1, "tophatfus": 1, "2017": [0, 1], "laboratoir": 1, "anatomo": 1, "cytopathologi": 1, "de": 1, "institut": 1, "universitair": 1, "du": 1, "toulous": 1, "oncopol": 1, "stat": 0, "barcod": 0, "greater": 0, "smallest": 0, "demultstatfactori": 0, "demult_stat": 0, "my_demultiplex_fold": 0, "unexpectedbarcod": 0, "ct": 0, "10000": 0, "14879": 0, "aatgc": 0, "ttta": 0, "10457": 0, "agcgc": 0, "ttga": 0, "abstractdemultstat": 0, "statist": 0, "lane": 0, "spla": 0, "atgc": 0, "demultstat": 0, "samplescount": 0, "atgg": 0, "154782": 0, "255567": 0, "undeterminedcount": 0, "udi": 0, "skipped_spl": 0, "great": [], "detect": 0, "find": 0, "neg": 0, "control": 0, "demultlog": 0, "demultipex": 0, "intens": 0, "basecal": 0, "bcl": 0, "folder_path": 0, "folder": 0, "out_fold": 0, "demult_path": 0, "undet_path": 0, "demultiplex_stat": 0, "top_unknown_barcod": 0, "undeter_path": 0, "compliant": 0, "filenam": 0, "getinffromseqid": [0, 1], "read_id": 0, "eas139": 0, "136": 0, "fc706vj": 0, "2104": 0, "15343": 0, "197393": 0, "atgcata": 0, "ctagc": 0, "sequencer_id": 0, "run_id": 0, "flowcell_id": 0, "lane_id": 0, "tile_id": 0, "x_po": 0, "y_po": 0, "umi": 0, "platform": 0, "instrument": 0, "serial": 0, "getplatformfromserialnumb": [0, 1], "hiseq": 0, "getilluminanam": [0, 1], "getinffromseqdesc": [0, 1], "seq_desc": 0, "flowcel": 0, "getlibnamefromreadspath": [0, 1], "seq_path": 0, "instrument_id": 0, "runinfo": [0, 1], "runparamet": [0, 1], "my_run": 0, "step": 0, "issequenc": 0, "iscopi": 0, "isend": 0, "completedjobinfo": [0, 1], "date_format": 0, "dt": 0, "xml": 0, "date": 0, "rtacomplet": [0, 1], "temp": 0, "post": 0, "rr": 0, "unparamet": 0, "etreetodict": [0, 1], "etre": 0, "elementtre": 0, "getrunfolderinfo": [0, 1], "run_fold": 0, "configur": 0, "samplesheetfactori": [0, 1], "spl": 0, "library_basenam": 0, "sapmlea_s1": 0, "sampleb_s2": 0, "samplec_s3": 0, "abstractsamplesheet": [0, 1], "abstract": 0, "cleanedend": 0, "invalid": 0, "comma": 0, "obtain": 0, "samplesheetad": [0, 1], "samplesheetv1": [0, 1], "design": 0, "amplicond": 0, "filterpanel": 0, "selected_manifest": 0, "manifest": 0, "panel": 0, "findsplfil": 0, "directori": 0, "end_pattern": 0, "basenam": 0, "setsplfil": 0, "v1": 0, "dam": 0, "market": 0, "product": 0, "technot": 0, "sheet": 0, "technic": 0, "970": 0, "004": 0, "pdf": 0, "demultiplex_sect": 0, "required_sect": 0, "samples_sect": 0, "samplsheet": 0, "samplesheetv2": [0, 1], "v2": 0, "sapac": 0, "system_document": 0, "nextseq2000": 0, "samplesheet_v2_templ": 0, "bclconvert_data": 0, "bclconvert_set": 0, "demultiplex": 1, "bcl2fastq": 1, "bclconvert": 1, "run": 1, "samplesheet": 1, "abstractatomicfilt": [0, 1], "emptyiterfilt": [0, 1], "wout_gp_filt": 0, "deep": 0, "accessor": 0, "dot": [], "attr_nam": 0, "need": 0, "increas": 0, "kind": 0, "args_list": 0, "tokens_chain": 0, "getterpathiterabletoken": [], "sub_gett": 0, "getterpathkeytoken": [], "getterpathmethodtoken": [], "tatcttcagc": 0, "index2": 0, "cgaatattgg": 0, "tgcacggata": 0, "gttgtctgcg": 0, "ggttgataga": 0, "tctgctccga": 0, "sheet_index": 0, "splid": 0, "_": 0, "splsheetindex": 0, "molecular": 0, "cgca": 0, "skip": 0, "enrich_soli": 0, "tumor_v3": 0, "project": 0, "rimo": 0, "section": 0, "spla_s3": 0, "samplead": [0, 1], "amplicon": 0, "vcfsymbaltrecord": [0, 1], "implic": 0, "fromstdrecord": [0, 1], "initi": 0, "uniq": 0, "isinvers": [0, 1], "sv_len": [0, 1], "longer": 0, "shorter": 0, "inv": 0, "getterpath": 1, "pathogen": 0, "layer": 0, "shape": 0, "squar": 0, "rectang": 0, "scaleup": 0, "abov": 0, "comparison": 0, "bewteen": 0, "ilter": 0, "drug": 0, "simpl": 0, "highlight": [], "python": [], "doctr": 0, "eq": 0, "ne": 0, "inequ": 0, "le": 0, "ge": 0, "lt": 0, "substr": 0, "token": 0, "dictpath": [], "wich": 0, "undefin": 0, "iterabletoken": [0, 1], "iteat": 0, "deeper": 0, "keytoken": [0, 1], "methodtoken": [0, 1], "null": 0, "brief": 0, "otherwis": 0, "classif": 0, "clinvar": 0, "cosmic": 0, "unknown_signific": 0, "tostr": [0, 1]}, "objects": {"anacore": [[0, 0, 0, "-", "STARLog"], [0, 0, 0, "-", "abstractFile"], [0, 0, 0, "-", "annotVcf"], [0, 0, 0, "-", "bed"], [0, 0, 0, "-", "db"], [0, 0, 0, "-", "filters"], [0, 0, 0, "-", "fusion"], [0, 0, 0, "-", "genomicRegion"], [0, 0, 0, "-", "getterPath"], [0, 0, 0, "-", "gff"], [0, 0, 0, "-", "gtf"], [0, 0, 0, "-", "hgvs"], [0, 0, 0, "-", "illumina"], [0, 0, 0, "-", "maf"], [0, 0, 0, "-", "matrix"], [0, 0, 0, "-", "msi"], [0, 0, 0, "-", "node"], [0, 0, 0, "-", "picardIO"], [0, 0, 0, "-", "region"], [0, 0, 0, "-", "sequence"], [0, 0, 0, "-", "sequenceIO"], [0, 0, 0, "-", "sv"], [0, 0, 0, "-", "tophatFusion"], [0, 0, 0, "-", "vcf"]], "anacore.STARLog": [[0, 1, 1, "", "STARLog"]], "anacore.abstractFile": [[0, 1, 1, "", "AbstractFile"], [0, 3, 1, "", "isEmpty"], [0, 3, 1, "", "isGzip"]], "anacore.abstractFile.AbstractFile": [[0, 2, 1, "", "close"], [0, 2, 1, "", "closed"], [0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "read"]], "anacore.annotVcf": [[0, 1, 1, "", "AnnotVCFIO"], [0, 1, 1, "", "VEPVCFIO"]], "anacore.annotVcf.AnnotVCFIO": [[0, 2, 1, "", "copyHeader"], [0, 2, 1, "", "recToVCFLine"], [0, 2, 1, "", "writeHeader"]], "anacore.bed": [[0, 1, 1, "", "BEDIO"], [0, 1, 1, "", "BEDRecord"], [0, 3, 1, "", "getAreas"], [0, 3, 1, "", "getAreasByChr"], [0, 3, 1, "", "getSortedAreasByChr"]], "anacore.bed.BEDIO": [[0, 2, 1, "", "BEDRecordToBEDLine"], [0, 2, 1, "", "getMaxNbCol"], [0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "write"]], "anacore.bed.BEDRecord": [[0, 2, 1, "", "recFromRegion"]], "anacore.db": [[0, 0, 0, "-", "homo_sapiens"]], "anacore.db.homo_sapiens": [[0, 0, 0, "-", "accession"]], "anacore.db.homo_sapiens.accession": [[0, 1, 1, "", "AssemblyAccession"], [0, 1, 1, "", "ChrAccession"]], "anacore.db.homo_sapiens.accession.AssemblyAccession": [[0, 4, 1, "", "HUMAN_BY_ACC"], [0, 4, 1, "", "HUMAN_BY_GENBANK"], [0, 4, 1, "", "HUMAN_BY_REFSEQ"], [0, 2, 1, "", "toHumanName"]], "anacore.db.homo_sapiens.accession.ChrAccession": [[0, 4, 1, "", "HUMAN_BY_ACC"], [0, 4, 1, "", "HUMAN_BY_GENBANK"], [0, 4, 1, "", "HUMAN_BY_HUMAN_CHR"], [0, 4, 1, "", "HUMAN_BY_HUMAN_WO_CHR"], [0, 4, 1, "", "HUMAN_BY_REFSEQ"], [0, 2, 1, "", "toHumanName"]], "anacore.filters": [[0, 1, 1, "", "AbstractAtomicFilter"], [0, 1, 1, "", "EmptyIterFilter"], [0, 1, 1, "", "Filter"], [0, 1, 1, "", "FiltersCombiner"], [0, 3, 1, "", "filtersFromDict"]], "anacore.filters.AbstractAtomicFilter": [[0, 2, 1, "", "eval"], [0, 2, 1, "", "setGetter"], [0, 2, 1, "", "toDict"]], "anacore.filters.EmptyIterFilter": [[0, 2, 1, "", "eval"], [0, 2, 1, "", "fromDict"]], "anacore.filters.Filter": [[0, 5, 1, "", "aggregator"], [0, 2, 1, "", "eval"], [0, 2, 1, "", "fromDict"], [0, 5, 1, "", "operator"], [0, 2, 1, "", "setFct"], [0, 5, 1, "", "values"]], "anacore.filters.FiltersCombiner": [[0, 2, 1, "", "eval"]], "anacore.fusion": [[0, 1, 1, "", "ArribaIO"], [0, 1, 1, "", "BreakendVCFIO"], [0, 1, 1, "", "FusionCatcherIO"], [0, 1, 1, "", "FusionFileReader"], [0, 1, 1, "", "STARFusionIO"], [0, 3, 1, "", "getAltFromCoord"], [0, 3, 1, "", "getBNDInterval"], [0, 3, 1, "", "getCoordDictFromCoordStr"], [0, 3, 1, "", "getCoordStr"], [0, 3, 1, "", "getStrand"]], "anacore.fusion.ArribaIO": [[0, 2, 1, "", "isValid"], [0, 2, 1, "", "setVCFHeader"], [0, 4, 1, "", "titles"], [0, 2, 1, "", "write"]], "anacore.fusion.BreakendVCFIO": [[0, 2, 1, "", "close"], [0, 2, 1, "", "get"], [0, 2, 1, "", "getSub"], [0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "loadIndex"], [0, 2, 1, "", "write"]], "anacore.fusion.FusionCatcherIO": [[0, 2, 1, "", "isValid"], [0, 2, 1, "", "setVCFHeader"], [0, 4, 1, "", "titles"], [0, 2, 1, "", "write"]], "anacore.fusion.FusionFileReader": [[0, 2, 1, "", "factory"]], "anacore.fusion.STARFusionIO": [[0, 2, 1, "", "isValid"], [0, 2, 1, "", "setVCFHeader"], [0, 4, 1, "", "titles"], [0, 2, 1, "", "write"]], "anacore.genomicRegion": [[0, 1, 1, "", "CDS"], [0, 1, 1, "", "Exon"], [0, 1, 1, "", "Gene"], [0, 1, 1, "", "Intron"], [0, 1, 1, "", "Protein"], [0, 1, 1, "", "Transcript"]], "anacore.genomicRegion.Protein": [[0, 2, 1, "", "aaLength"], [0, 2, 1, "", "contains"], [0, 2, 1, "", "getCDSFromTranscript"], [0, 2, 1, "", "getCodonInfo"], [0, 2, 1, "", "getCodonRefPos"], [0, 2, 1, "", "getCodonSeqFromProtPos"], [0, 2, 1, "", "getNtPosFromRefPos"], [0, 2, 1, "", "getNtPosFromRegionPos"], [0, 2, 1, "", "getPosOnRef"], [0, 2, 1, "", "getPosOnRegion"], [0, 2, 1, "", "getSubFromRegionPos"], [0, 2, 1, "", "hasOverlap"], [0, 2, 1, "", "length"], [0, 5, 1, "", "transcript"]], "anacore.genomicRegion.Transcript": [[0, 2, 1, "", "addProtein"], [0, 2, 1, "", "delProtein"], [0, 2, 1, "", "getPosOnRef"], [0, 2, 1, "", "getPosOnRegion"], [0, 2, 1, "", "getSubFromRefPos"], [0, 2, 1, "", "getSubFromRegionPos"], [0, 2, 1, "", "length"], [0, 5, 1, "", "proteins"]], "anacore.getterPath": [[0, 1, 1, "", "GetterPath"], [0, 1, 1, "", "IterableToken"], [0, 1, 1, "", "KeyToken"], [0, 1, 1, "", "MethodToken"]], "anacore.getterPath.GetterPath": [[0, 2, 1, "", "fromDict"], [0, 2, 1, "", "fromStr"], [0, 2, 1, "", "get"], [0, 2, 1, "", "toDict"], [0, 2, 1, "", "toStr"]], "anacore.getterPath.IterableToken": [[0, 2, 1, "", "get"], [0, 2, 1, "", "toDict"], [0, 2, 1, "", "toStr"]], "anacore.getterPath.KeyToken": [[0, 2, 1, "", "get"], [0, 2, 1, "", "toDict"], [0, 2, 1, "", "toStr"]], "anacore.getterPath.MethodToken": [[0, 2, 1, "", "get"], [0, 2, 1, "", "toDict"], [0, 2, 1, "", "toStr"]], "anacore.gff": [[0, 1, 1, "", "GFF3IO"], [0, 1, 1, "", "GFF3Record"]], "anacore.gff.GFF3IO": [[0, 2, 1, "", "decodedValue"], [0, 2, 1, "", "encodedValue"], [0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "write"]], "anacore.gff.GFF3Record": [[0, 2, 1, "", "addToAnnot"], [0, 2, 1, "", "toGff"]], "anacore.gtf": [[0, 1, 1, "", "GTFIO"], [0, 3, 1, "", "loadModel"]], "anacore.gtf.GTFIO": [[0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "recordToLine"], [0, 2, 1, "", "write"]], "anacore.hgvs": [[0, 1, 1, "", "Accession"], [0, 1, 1, "", "HGVS"], [0, 1, 1, "", "HGVSProtChange"], [0, 1, 1, "", "MutalyzerBatch"], [0, 1, 1, "", "RunMutalyzerDescription"], [0, 1, 1, "", "RunMutalyzerLegend"]], "anacore.hgvs.HGVS": [[0, 2, 1, "", "fromStr"], [0, 2, 1, "", "isPredicted"]], "anacore.hgvs.HGVSProtChange": [[0, 5, 1, "", "end_aa"], [0, 2, 1, "", "fromStr"], [0, 2, 1, "", "insCouldBeIdentical"], [0, 2, 1, "", "isInFrameIns"], [0, 2, 1, "", "isValid"], [0, 5, 1, "", "new_aa"], [0, 5, 1, "", "start_aa"]], "anacore.hgvs.MutalyzerBatch": [[0, 2, 1, "", "getExecTime"], [0, 2, 1, "", "getParsedResponse"], [0, 2, 1, "", "getRequestURL"], [0, 2, 1, "", "getResponse"], [0, 2, 1, "", "submit"], [0, 2, 1, "", "syncRequest"], [0, 2, 1, "", "updateStatus"]], "anacore.hgvs.RunMutalyzerDescription": [[0, 2, 1, "", "getByAccession"]], "anacore.hgvs.RunMutalyzerLegend": [[0, 2, 1, "", "getIdByName"], [0, 2, 1, "", "getProtBytr"]], "anacore.illumina": [[0, 0, 0, "-", "base"], [0, 0, 0, "-", "demultiplex"], [0, 0, 0, "-", "run"], [0, 0, 0, "-", "samplesheet"]], "anacore.illumina.base": [[0, 3, 1, "", "getIlluminaName"], [0, 3, 1, "", "getInfFromSeqDesc"], [0, 3, 1, "", "getInfFromSeqID"], [0, 3, 1, "", "getLibNameFromReadsPath"], [0, 3, 1, "", "getPlatformFromSerialNumber"]], "anacore.illumina.demultiplex": [[0, 1, 1, "", "DemultStatFactory"], [0, 0, 0, "-", "base"], [0, 0, 0, "-", "bcl2fastq"], [0, 0, 0, "-", "bclconvert"]], "anacore.illumina.demultiplex.DemultStatFactory": [[0, 2, 1, "", "get"]], "anacore.illumina.demultiplex.base": [[0, 1, 1, "", "AbstractDemultStat"]], "anacore.illumina.demultiplex.base.AbstractDemultStat": [[0, 2, 1, "", "samplesCounts"], [0, 5, 1, "", "undetermined"], [0, 2, 1, "", "undeterminedCounts"], [0, 2, 1, "", "unexpectedBarcodes"]], "anacore.illumina.demultiplex.bcl2fastq": [[0, 1, 1, "", "DemultLog"], [0, 1, 1, "", "DemultStat"]], "anacore.illumina.demultiplex.bcl2fastq.DemultStat": [[0, 5, 1, "", "undetermined"]], "anacore.illumina.demultiplex.bclconvert": [[0, 1, 1, "", "DemultLog"], [0, 1, 1, "", "DemultStat"]], "anacore.illumina.demultiplex.bclconvert.DemultStat": [[0, 5, 1, "", "undetermined"]], "anacore.illumina.run": [[0, 1, 1, "", "CompletedJobInfo"], [0, 1, 1, "", "RTAComplete"], [0, 1, 1, "", "Run"], [0, 1, 1, "", "RunInfo"], [0, 1, 1, "", "RunParameters"], [0, 3, 1, "", "etreeToDict"], [0, 3, 1, "", "getRunFolderInfo"]], "anacore.illumina.run.Run": [[0, 2, 1, "", "isCopied"], [0, 2, 1, "", "isEnded"], [0, 2, 1, "", "isSequenced"], [0, 5, 1, "", "parameters"]], "anacore.illumina.samplesheet": [[0, 1, 1, "", "AbstractSampleSheet"], [0, 1, 1, "", "Sample"], [0, 1, 1, "", "SampleADS"], [0, 1, 1, "", "SampleSheetADS"], [0, 1, 1, "", "SampleSheetFactory"], [0, 1, 1, "", "SampleSheetV1"], [0, 1, 1, "", "SampleSheetV2"]], "anacore.illumina.samplesheet.AbstractSampleSheet": [[0, 2, 1, "", "cleanedEnd"]], "anacore.illumina.samplesheet.Sample": [[0, 5, 1, "", "basename"], [0, 5, 1, "", "library_basename"], [0, 2, 1, "", "toDict"]], "anacore.illumina.samplesheet.SampleADS": [[0, 5, 1, "", "basename"], [0, 2, 1, "", "findSplFiles"], [0, 2, 1, "", "setSplFiles"]], "anacore.illumina.samplesheet.SampleSheetADS": [[0, 2, 1, "", "filterPanels"], [0, 2, 1, "", "isValid"]], "anacore.illumina.samplesheet.SampleSheetFactory": [[0, 2, 1, "", "get"]], "anacore.illumina.samplesheet.SampleSheetV1": [[0, 4, 1, "", "DEMULTIPLEX_SECTION"], [0, 4, 1, "", "REQUIRED_SECTIONS"], [0, 4, 1, "", "SAMPLES_SECTION"], [0, 2, 1, "", "isValid"]], "anacore.illumina.samplesheet.SampleSheetV2": [[0, 4, 1, "", "DEMULTIPLEX_SECTION"], [0, 4, 1, "", "REQUIRED_SECTIONS"], [0, 4, 1, "", "SAMPLES_SECTION"], [0, 2, 1, "", "isValid"]], "anacore.maf": [[0, 1, 1, "", "MAFIO"], [0, 3, 1, "", "getName"]], "anacore.maf.MAFIO": [[0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "write"]], "anacore.matrix": [[0, 1, 1, "", "DistanceMatrixIO"]], "anacore.msi": [[0, 0, 0, "-", "annot"], [0, 0, 0, "-", "base"], [0, 0, 0, "-", "hubble"], [0, 0, 0, "-", "locus"], [0, 0, 0, "-", "msings"], [0, 0, 0, "-", "msisensorpro"], [0, 0, 0, "-", "reportIO"], [0, 0, 0, "-", "sample"]], "anacore.msi.annot": [[0, 1, 1, "", "MSIAnnot"], [0, 3, 1, "", "addLociResToSpl"], [0, 3, 1, "", "addLociResult"], [0, 3, 1, "", "getCastedValue"], [0, 3, 1, "", "getLocusAnnotDict"]], "anacore.msi.annot.MSIAnnot": [[0, 2, 1, "", "recordToLine"]], "anacore.msi.base": [[0, 1, 1, "", "LocusClassifier"], [0, 1, 1, "", "Status"], [0, 3, 1, "", "getNbSupporting"], [0, 3, 1, "", "toDict"]], "anacore.msi.base.LocusClassifier": [[0, 2, 1, "", "fit"], [0, 2, 1, "", "predict"], [0, 2, 1, "", "predict_proba"], [0, 2, 1, "", "set_status"]], "anacore.msi.base.Status": [[0, 2, 1, "", "authorizedValues"], [0, 4, 1, "", "none"], [0, 4, 1, "", "stable"], [0, 4, 1, "", "undetermined"], [0, 4, 1, "", "unstable"]], "anacore.msi.hubble": [[0, 1, 1, "", "HubbleDiff"], [0, 1, 1, "", "HubbleDist"], [0, 3, 1, "", "nbRepeatFromStart"], [0, 3, 1, "", "parseHubbleResults"]], "anacore.msi.hubble.HubbleDiff": [[0, 2, 1, "", "recordToLine"]], "anacore.msi.hubble.HubbleDist": [[0, 2, 1, "", "recordToLine"]], "anacore.msi.locus": [[0, 1, 1, "", "Locus"], [0, 1, 1, "", "LocusDataDistrib"], [0, 1, 1, "", "LocusRes"], [0, 3, 1, "", "getRefSeqInfo"]], "anacore.msi.locus.Locus": [[0, 2, 1, "", "delResult"], [0, 5, 1, "", "end"], [0, 2, 1, "", "fromDict"], [0, 5, 1, "", "length"], [0, 5, 1, "", "start"]], "anacore.msi.locus.LocusDataDistrib": [[0, 2, 1, "", "fromDense"], [0, 2, 1, "", "getCount"], [0, 2, 1, "", "getDenseCount"], [0, 2, 1, "", "getDensePrct"], [0, 2, 1, "", "getMaxLength"], [0, 2, 1, "", "getMinLength"], [0, 2, 1, "", "getMostRepresented"], [0, 2, 1, "", "getNbPeaks"], [0, 2, 1, "", "items"]], "anacore.msi.locus.LocusRes": [[0, 2, 1, "", "fromDict"]], "anacore.msi.msings": [[0, 1, 1, "", "MSINGSAnalysisIO"], [0, 1, 1, "", "MSINGSEval"], [0, 1, 1, "", "MSINGSReport"]], "anacore.msi.msings.MSINGSAnalysisIO": [[0, 2, 1, "", "recordToLine"]], "anacore.msi.msings.MSINGSEval": [[0, 2, 1, "", "getNbPeaks"], [0, 2, 1, "", "getThresholdFromNbPeaks"]], "anacore.msi.msisensorpro": [[0, 1, 1, "", "BaselineIO"], [0, 1, 1, "", "BaselineRecord"], [0, 1, 1, "", "DistIO"], [0, 1, 1, "", "ProEval"], [0, 1, 1, "", "ProIO"], [0, 3, 1, "", "binaryToString"], [0, 3, 1, "", "parseProResults"], [0, 3, 1, "", "stringToBinary"]], "anacore.msi.msisensorpro.BaselineIO": [[0, 2, 1, "", "recordToLine"]], "anacore.msi.msisensorpro.BaselineRecord": [[0, 2, 1, "", "fromModel"], [0, 5, 1, "", "left_flank_binary"], [0, 5, 1, "", "repeat_unit_binary"], [0, 5, 1, "", "repeat_unit_length"], [0, 5, 1, "", "right_flank_binary"]], "anacore.msi.msisensorpro.DistIO": [[0, 2, 1, "", "nextLocus"], [0, 2, 1, "", "recordToLines"], [0, 2, 1, "", "write"]], "anacore.msi.msisensorpro.ProEval": [[0, 2, 1, "", "getSlippageScores"], [0, 2, 1, "", "getThreshold"], [0, 2, 1, "", "getThresholdFromScores"]], "anacore.msi.msisensorpro.ProIO": [[0, 2, 1, "", "loci"], [0, 2, 1, "", "recordToLine"], [0, 2, 1, "", "write"]], "anacore.msi.reportIO": [[0, 1, 1, "", "ReportIO"]], "anacore.msi.reportIO.ReportIO": [[0, 2, 1, "", "getIncompleteModels"], [0, 2, 1, "", "parse"], [0, 2, 1, "", "write"]], "anacore.msi.sample": [[0, 1, 1, "", "MSISample"], [0, 1, 1, "", "MSISplRes"]], "anacore.msi.sample.MSISample": [[0, 2, 1, "", "addLocus"], [0, 2, 1, "", "delLoci"], [0, 2, 1, "", "delLocus"], [0, 2, 1, "", "fromDict"], [0, 2, 1, "", "getLociMethods"], [0, 2, 1, "", "getNbDetermined"], [0, 2, 1, "", "getNbLoci"], [0, 2, 1, "", "getNbProcessed"], [0, 2, 1, "", "getNbStable"], [0, 2, 1, "", "getNbUndetermined"], [0, 2, 1, "", "getNbUnstable"], [0, 2, 1, "", "setScore"], [0, 2, 1, "", "setStatusByInstabilityRatio"]], "anacore.msi.sample.MSISplRes": [[0, 2, 1, "", "fromDict"]], "anacore.node": [[0, 1, 1, "", "Node"]], "anacore.node.Node": [[0, 2, 1, "", "addChild"], [0, 2, 1, "", "fromClusterNode"], [0, 2, 1, "", "fromDict"], [0, 2, 1, "", "getAncestors"], [0, 2, 1, "", "getChildByName"], [0, 2, 1, "", "getDepth"], [0, 2, 1, "", "getDescendants"], [0, 2, 1, "", "getLeaves"], [0, 2, 1, "", "hasChild"], [0, 2, 1, "", "toDict"], [0, 2, 1, "", "toExtendedNewick"], [0, 2, 1, "", "toNewick"]], "anacore.picardIO": [[0, 1, 1, "", "PicardReader"], [0, 3, 1, "", "castCol"], [0, 3, 1, "", "getColType"], [0, 3, 1, "", "getValType"]], "anacore.region": [[0, 1, 1, "", "Region"], [0, 1, 1, "", "RegionList"], [0, 1, 1, "", "RegionTree"], [0, 3, 1, "", "consolidated"], [0, 3, 1, "", "iterOverlapped"], [0, 3, 1, "", "iterOverlappedByRegion"], [0, 3, 1, "", "mergedRegion"], [0, 3, 1, "", "splittedByRef"]], "anacore.region.Region": [[0, 2, 1, "", "contains"], [0, 2, 1, "", "fromStr"], [0, 2, 1, "", "getCoordinatesStr"], [0, 2, 1, "", "getMinDist"], [0, 2, 1, "", "getPosOnRef"], [0, 2, 1, "", "getPosOnRegion"], [0, 2, 1, "", "hasOverlap"], [0, 2, 1, "", "hasStrandedOverlap"], [0, 2, 1, "", "length"], [0, 2, 1, "", "setReference"], [0, 2, 1, "", "strandedContains"]], "anacore.region.RegionList": [[0, 2, 1, "", "getContainers"], [0, 2, 1, "", "getNearests"], [0, 2, 1, "", "getOverlapped"]], "anacore.region.RegionTree": [[0, 2, 1, "", "addChild"], [0, 2, 1, "", "sortChildren"]], "anacore.sequence": [[0, 1, 1, "", "AA3LettersAlphabet"], [0, 1, 1, "", "Alphabet"], [0, 1, 1, "", "CodonAlphabet"], [0, 1, 1, "", "DNAAlphabet"], [0, 1, 1, "", "RNAAlphabet"], [0, 1, 1, "", "Sequence"], [0, 3, 1, "", "getShortestRepeatUnit"]], "anacore.sequence.AA3LettersAlphabet": [[0, 4, 1, "", "codons_by_aa"], [0, 2, 1, "", "isValid"], [0, 4, 1, "", "one_by_three"], [0, 2, 1, "", "toOneLetter"], [0, 4, 1, "", "words"]], "anacore.sequence.CodonAlphabet": [[0, 4, 1, "", "aa_by_codon"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "translate"], [0, 4, 1, "", "words"]], "anacore.sequence.DNAAlphabet": [[0, 4, 1, "", "complement"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "revCom"], [0, 4, 1, "", "words"]], "anacore.sequence.RNAAlphabet": [[0, 4, 1, "", "complement"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "revCom"], [0, 4, 1, "", "words"]], "anacore.sequence.Sequence": [[0, 2, 1, "", "dnaRevCom"], [0, 2, 1, "", "rnaRevCom"]], "anacore.sequenceIO": [[0, 1, 1, "", "Faidx"], [0, 1, 1, "", "FaidxRecord"], [0, 1, 1, "", "FastaIO"], [0, 1, 1, "", "FastqIO"], [0, 1, 1, "", "IdxFastaIO"], [0, 1, 1, "", "SequenceFileReader"], [0, 3, 1, "", "getStrandedSeqFromPos"]], "anacore.sequenceIO.Faidx": [[0, 2, 1, "", "readById"]], "anacore.sequenceIO.FastaIO": [[0, 2, 1, "", "close"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "nbSeq"], [0, 2, 1, "", "nbSeqAndNt"], [0, 2, 1, "", "nextSeq"], [0, 2, 1, "", "seqToFastaLine"], [0, 2, 1, "", "write"]], "anacore.sequenceIO.FastqIO": [[0, 2, 1, "", "close"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "nbSeq"], [0, 2, 1, "", "nbSeqAndNt"], [0, 2, 1, "", "nextSeq"], [0, 2, 1, "", "qualOffset"], [0, 2, 1, "", "seqToFastqLine"], [0, 2, 1, "", "write"]], "anacore.sequenceIO.IdxFastaIO": [[0, 2, 1, "", "get"], [0, 2, 1, "", "getSub"]], "anacore.sequenceIO.SequenceFileReader": [[0, 2, 1, "", "factory"]], "anacore.sv": [[0, 1, 1, "", "HashedSVIO"], [0, 1, 1, "", "SVIO"]], "anacore.sv.HashedSVIO": [[0, 2, 1, "", "recordToLine"]], "anacore.sv.SVIO": [[0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "isValid"], [0, 2, 1, "", "recordToLine"], [0, 2, 1, "", "write"], [0, 2, 1, "", "writeHeader"]], "anacore.tophatFusion": [[0, 1, 1, "", "TopHatFusionIO"]], "anacore.vcf": [[0, 1, 1, "", "HeaderAttr"], [0, 1, 1, "", "HeaderDescAttr"], [0, 1, 1, "", "HeaderFilterAttr"], [0, 1, 1, "", "HeaderFormatAttr"], [0, 1, 1, "", "HeaderInfoAttr"], [0, 1, 1, "", "HeaderSampleAttr"], [0, 1, 1, "", "HeaderTypedAttr"], [0, 1, 1, "", "VCFIO"], [0, 1, 1, "", "VCFRecord"], [0, 1, 1, "", "VCFSymbAltRecord"], [0, 3, 1, "", "decodeInfoValue"], [0, 3, 1, "", "encodeInfoValue"], [0, 3, 1, "", "getAlleleRecord"], [0, 3, 1, "", "getFreqMatrix"], [0, 3, 1, "", "getHeaderAttr"]], "anacore.vcf.HeaderAttr": [[0, 2, 1, "", "items"], [0, 2, 1, "", "keys"]], "anacore.vcf.VCFIO": [[0, 2, 1, "", "copyHeader"], [0, 2, 1, "", "getSub"], [0, 2, 1, "", "isRecordLine"], [0, 2, 1, "", "recToVCFLine"], [0, 2, 1, "", "write"], [0, 2, 1, "", "writeHeader"]], "anacore.vcf.VCFRecord": [[0, 2, 1, "", "containsIndel"], [0, 2, 1, "", "fastDownstreamed"], [0, 2, 1, "", "fastStandardize"], [0, 2, 1, "", "getAD"], [0, 2, 1, "", "getAF"], [0, 2, 1, "", "getAFBySample"], [0, 2, 1, "", "getAltAD"], [0, 2, 1, "", "getAltAF"], [0, 2, 1, "", "getDP"], [0, 2, 1, "", "getEmptyAlleleMarker"], [0, 2, 1, "", "getMostDownstream"], [0, 2, 1, "", "getMostUpstream"], [0, 2, 1, "", "getName"], [0, 2, 1, "", "getPopAltAD"], [0, 2, 1, "", "getPopAltAF"], [0, 2, 1, "", "getPopDP"], [0, 2, 1, "", "getPopRefAD"], [0, 2, 1, "", "getPopRefAF"], [0, 2, 1, "", "isDeletion"], [0, 2, 1, "", "isIndel"], [0, 2, 1, "", "isInsAndDel"], [0, 2, 1, "", "isInsertion"], [0, 2, 1, "", "normalizeSingleAllele"], [0, 2, 1, "", "refEnd"], [0, 2, 1, "", "refStart"], [0, 2, 1, "", "type"]], "anacore.vcf.VCFSymbAltRecord": [[0, 2, 1, "", "containsIndel"], [0, 5, 1, "", "end"], [0, 2, 1, "", "fastDownstreamed"], [0, 2, 1, "", "fastStandardize"], [0, 2, 1, "", "fromStdRecord"], [0, 2, 1, "", "getMostDownstream"], [0, 2, 1, "", "getMostUpstream"], [0, 2, 1, "", "getName"], [0, 2, 1, "", "isDeletion"], [0, 2, 1, "", "isIndel"], [0, 2, 1, "", "isInsAndDel"], [0, 2, 1, "", "isInsertion"], [0, 2, 1, "", "isInversion"], [0, 2, 1, "", "normalizeSingleAllele"], [0, 2, 1, "", "refEnd"], [0, 2, 1, "", "refStart"], [0, 5, 1, "", "sv_len"], [0, 2, 1, "", "type"]]}, "objtypes": {"0": "py:module", "1": "py:class", "2": "py:method", "3": "py:function", "4": "py:attribute", "5": "py:property"}, "objnames": {"0": ["py", "module", "Python module"], "1": ["py", "class", "Python class"], "2": ["py", "method", "Python method"], "3": ["py", "function", "Python function"], "4": ["py", "attribute", "Python attribute"], "5": ["py", "property", "Python property"]}, "titleterms": {"anacor": [0, 1], "packag": 0, "abstractfil": 0, "annotvcf": 0, "bed": 0, "db": 0, "homo_sapien": 0, "access": 0, "filter": 0, "fusion": 0, "genomicregion": 0, "gff": 0, "gtf": 0, "hgv": 0, "illumina": 0, "maf": 0, "matrix": 0, "msi": 0, "annot": 0, "base": 0, "hubbl": 0, "locu": 0, "msing": 0, "msisensorpro": 0, "reportio": 0, "sampl": 0, "node": 0, "picardio": 0, "region": 0, "sequenc": 0, "sequenceio": 0, "starlog": 0, "sv": 0, "tophatfus": 0, "vcf": 0, "manag": 1, "standard": 1, "file": 1, "format": 1, "object": 1, "from": 1, "ng": 1, "descript": 1, "instal": 1, "content": 1, "copyright": 1, "indic": 1, "tabl": 1, "demultiplex": 0, "bcl2fastq": 0, "bclconvert": 0, "run": 0, "samplesheet": 0, "getterpath": 0}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx.ext.viewcode": 1, "sphinx": 57}, "alltitles": {"AnaCore - managing standard file formats and objects from NGS": [[1, "anacore-managing-standard-file-formats-and-objects-from-ngs"]], "Description": [[1, "description"]], "Installation": [[1, "installation"]], "Content": [[1, "content"]], "Copyright": [[1, "copyright"]], "Indices and tables": [[1, "indices-and-tables"]], "AnaCore package": [[0, "anacore-package"]], "anacore.abstractFile": [[0, "module-anacore.abstractFile"]], "anacore.annotVcf": [[0, "module-anacore.annotVcf"]], "anacore.bed": [[0, "module-anacore.bed"]], "anacore.db": [[0, "module-anacore.db"]], "anacore.db.homo_sapiens": [[0, "module-anacore.db.homo_sapiens"]], "anacore.db.homo_sapiens.accession": [[0, "module-anacore.db.homo_sapiens.accession"]], "anacore.filters": [[0, "module-anacore.filters"]], "anacore.fusion": [[0, "module-anacore.fusion"]], "anacore.genomicRegion": [[0, "module-anacore.genomicRegion"]], "anacore.getterPath": [[0, "module-anacore.getterPath"]], "anacore.gff": [[0, "module-anacore.gff"]], "anacore.gtf": [[0, "module-anacore.gtf"]], "anacore.hgvs": [[0, "module-anacore.hgvs"]], "anacore.illumina": [[0, "module-anacore.illumina"]], "anacore.illumina.base": [[0, "module-anacore.illumina.base"]], "anacore.illumina.demultiplex": [[0, "module-anacore.illumina.demultiplex"]], "anacore.illumina.demultiplex.base": [[0, "module-anacore.illumina.demultiplex.base"]], "anacore.illumina.demultiplex.bcl2fastq": [[0, "module-anacore.illumina.demultiplex.bcl2fastq"]], "anacore.illumina.demultiplex.bclconvert": [[0, "module-anacore.illumina.demultiplex.bclconvert"]], "anacore.illumina.run": [[0, "module-anacore.illumina.run"]], "anacore.illumina.samplesheet": [[0, "module-anacore.illumina.samplesheet"]], "anacore.maf": [[0, "module-anacore.maf"]], "anacore.matrix": [[0, "module-anacore.matrix"]], "anacore.msi": [[0, "module-anacore.msi"]], "anacore.msi.annot": [[0, "module-anacore.msi.annot"]], "anacore.msi.base": [[0, "module-anacore.msi.base"]], "anacore.msi.hubble": [[0, "module-anacore.msi.hubble"]], "anacore.msi.locus": [[0, "module-anacore.msi.locus"]], "anacore.msi.msings": [[0, "module-anacore.msi.msings"]], "anacore.msi.msisensorpro": [[0, "module-anacore.msi.msisensorpro"]], "anacore.msi.reportIO": [[0, "module-anacore.msi.reportIO"]], "anacore.msi.sample": [[0, "module-anacore.msi.sample"]], "anacore.node": [[0, "module-anacore.node"]], "anacore.picardIO": [[0, "module-anacore.picardIO"]], "anacore.region": [[0, "module-anacore.region"]], "anacore.sequence": [[0, "module-anacore.sequence"]], "anacore.sequenceIO": [[0, "module-anacore.sequenceIO"]], "anacore.STARLog": [[0, "module-anacore.STARLog"]], "anacore.sv": [[0, "module-anacore.sv"]], "anacore.tophatFusion": [[0, "module-anacore.tophatFusion"]], "anacore.vcf": [[0, "module-anacore.vcf"]]}, "indexentries": {"aa3lettersalphabet (class in anacore.sequence)": [[0, "anacore.sequence.AA3LettersAlphabet"]], "abstractatomicfilter (class in anacore.filters)": [[0, "anacore.filters.AbstractAtomicFilter"]], "abstractdemultstat (class in anacore.illumina.demultiplex.base)": [[0, "anacore.illumina.demultiplex.base.AbstractDemultStat"]], "abstractfile (class in anacore.abstractfile)": [[0, "anacore.abstractFile.AbstractFile"]], "abstractsamplesheet (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.AbstractSampleSheet"]], "accession (class in anacore.hgvs)": [[0, "anacore.hgvs.Accession"]], "alphabet (class in anacore.sequence)": [[0, "anacore.sequence.Alphabet"]], "annotvcfio (class in anacore.annotvcf)": [[0, "anacore.annotVcf.AnnotVCFIO"]], "arribaio (class in anacore.fusion)": [[0, "anacore.fusion.ArribaIO"]], "assemblyaccession (class in anacore.db.homo_sapiens.accession)": [[0, "anacore.db.homo_sapiens.accession.AssemblyAccession"]], "bedio (class in anacore.bed)": [[0, "anacore.bed.BEDIO"]], "bedrecord (class in anacore.bed)": [[0, "anacore.bed.BEDRecord"]], "bedrecordtobedline() (anacore.bed.bedio method)": [[0, "anacore.bed.BEDIO.BEDRecordToBEDLine"]], "baselineio (class in anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.BaselineIO"]], "baselinerecord (class in anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.BaselineRecord"]], "breakendvcfio (class in anacore.fusion)": [[0, "anacore.fusion.BreakendVCFIO"]], "cds (class in anacore.genomicregion)": [[0, "anacore.genomicRegion.CDS"]], "chraccession (class in anacore.db.homo_sapiens.accession)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession"]], "codonalphabet (class in anacore.sequence)": [[0, "anacore.sequence.CodonAlphabet"]], "completedjobinfo (class in anacore.illumina.run)": [[0, "anacore.illumina.run.CompletedJobInfo"]], "demultiplex_section (anacore.illumina.samplesheet.samplesheetv1 attribute)": [[0, "anacore.illumina.samplesheet.SampleSheetV1.DEMULTIPLEX_SECTION"]], "demultiplex_section (anacore.illumina.samplesheet.samplesheetv2 attribute)": [[0, "anacore.illumina.samplesheet.SampleSheetV2.DEMULTIPLEX_SECTION"]], "dnaalphabet (class in anacore.sequence)": [[0, "anacore.sequence.DNAAlphabet"]], "demultlog (class in anacore.illumina.demultiplex.bcl2fastq)": [[0, "anacore.illumina.demultiplex.bcl2fastq.DemultLog"]], "demultlog (class in anacore.illumina.demultiplex.bclconvert)": [[0, "anacore.illumina.demultiplex.bclconvert.DemultLog"]], "demultstat (class in anacore.illumina.demultiplex.bcl2fastq)": [[0, "anacore.illumina.demultiplex.bcl2fastq.DemultStat"]], "demultstat (class in anacore.illumina.demultiplex.bclconvert)": [[0, "anacore.illumina.demultiplex.bclconvert.DemultStat"]], "demultstatfactory (class in anacore.illumina.demultiplex)": [[0, "anacore.illumina.demultiplex.DemultStatFactory"]], "distio (class in anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.DistIO"]], "distancematrixio (class in anacore.matrix)": [[0, "anacore.matrix.DistanceMatrixIO"]], "emptyiterfilter (class in anacore.filters)": [[0, "anacore.filters.EmptyIterFilter"]], "exon (class in anacore.genomicregion)": [[0, "anacore.genomicRegion.Exon"]], "faidx (class in anacore.sequenceio)": [[0, "anacore.sequenceIO.Faidx"]], "faidxrecord (class in anacore.sequenceio)": [[0, "anacore.sequenceIO.FaidxRecord"]], "fastaio (class in anacore.sequenceio)": [[0, "anacore.sequenceIO.FastaIO"]], "fastqio (class in anacore.sequenceio)": [[0, "anacore.sequenceIO.FastqIO"]], "filter (class in anacore.filters)": [[0, "anacore.filters.Filter"]], "filterscombiner (class in anacore.filters)": [[0, "anacore.filters.FiltersCombiner"]], "fusioncatcherio (class in anacore.fusion)": [[0, "anacore.fusion.FusionCatcherIO"]], "fusionfilereader (class in anacore.fusion)": [[0, "anacore.fusion.FusionFileReader"]], "gff3io (class in anacore.gff)": [[0, "anacore.gff.GFF3IO"]], "gff3record (class in anacore.gff)": [[0, "anacore.gff.GFF3Record"]], "gtfio (class in anacore.gtf)": [[0, "anacore.gtf.GTFIO"]], "gene (class in anacore.genomicregion)": [[0, "anacore.genomicRegion.Gene"]], "getterpath (class in anacore.getterpath)": [[0, "anacore.getterPath.GetterPath"]], "hgvs (class in anacore.hgvs)": [[0, "anacore.hgvs.HGVS"]], "hgvsprotchange (class in anacore.hgvs)": [[0, "anacore.hgvs.HGVSProtChange"]], "human_by_acc (anacore.db.homo_sapiens.accession.assemblyaccession attribute)": [[0, "anacore.db.homo_sapiens.accession.AssemblyAccession.HUMAN_BY_ACC"]], "human_by_acc (anacore.db.homo_sapiens.accession.chraccession attribute)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession.HUMAN_BY_ACC"]], "human_by_genbank (anacore.db.homo_sapiens.accession.assemblyaccession attribute)": [[0, "anacore.db.homo_sapiens.accession.AssemblyAccession.HUMAN_BY_GENBANK"]], "human_by_genbank (anacore.db.homo_sapiens.accession.chraccession attribute)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession.HUMAN_BY_GENBANK"]], "human_by_human_chr (anacore.db.homo_sapiens.accession.chraccession attribute)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession.HUMAN_BY_HUMAN_CHR"]], "human_by_human_wo_chr (anacore.db.homo_sapiens.accession.chraccession attribute)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession.HUMAN_BY_HUMAN_WO_CHR"]], "human_by_refseq (anacore.db.homo_sapiens.accession.assemblyaccession attribute)": [[0, "anacore.db.homo_sapiens.accession.AssemblyAccession.HUMAN_BY_REFSEQ"]], "human_by_refseq (anacore.db.homo_sapiens.accession.chraccession attribute)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession.HUMAN_BY_REFSEQ"]], "hashedsvio (class in anacore.sv)": [[0, "anacore.sv.HashedSVIO"]], "headerattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderAttr"]], "headerdescattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderDescAttr"]], "headerfilterattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderFilterAttr"]], "headerformatattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderFormatAttr"]], "headerinfoattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderInfoAttr"]], "headersampleattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderSampleAttr"]], "headertypedattr (class in anacore.vcf)": [[0, "anacore.vcf.HeaderTypedAttr"]], "hubblediff (class in anacore.msi.hubble)": [[0, "anacore.msi.hubble.HubbleDiff"]], "hubbledist (class in anacore.msi.hubble)": [[0, "anacore.msi.hubble.HubbleDist"]], "idxfastaio (class in anacore.sequenceio)": [[0, "anacore.sequenceIO.IdxFastaIO"]], "intron (class in anacore.genomicregion)": [[0, "anacore.genomicRegion.Intron"]], "iterabletoken (class in anacore.getterpath)": [[0, "anacore.getterPath.IterableToken"]], "keytoken (class in anacore.getterpath)": [[0, "anacore.getterPath.KeyToken"]], "locus (class in anacore.msi.locus)": [[0, "anacore.msi.locus.Locus"]], "locusclassifier (class in anacore.msi.base)": [[0, "anacore.msi.base.LocusClassifier"]], "locusdatadistrib (class in anacore.msi.locus)": [[0, "anacore.msi.locus.LocusDataDistrib"]], "locusres (class in anacore.msi.locus)": [[0, "anacore.msi.locus.LocusRes"]], "mafio (class in anacore.maf)": [[0, "anacore.maf.MAFIO"]], "msiannot (class in anacore.msi.annot)": [[0, "anacore.msi.annot.MSIAnnot"]], "msingsanalysisio (class in anacore.msi.msings)": [[0, "anacore.msi.msings.MSINGSAnalysisIO"]], "msingseval (class in anacore.msi.msings)": [[0, "anacore.msi.msings.MSINGSEval"]], "msingsreport (class in anacore.msi.msings)": [[0, "anacore.msi.msings.MSINGSReport"]], "msisample (class in anacore.msi.sample)": [[0, "anacore.msi.sample.MSISample"]], "msisplres (class in anacore.msi.sample)": [[0, "anacore.msi.sample.MSISplRes"]], "methodtoken (class in anacore.getterpath)": [[0, "anacore.getterPath.MethodToken"]], "mutalyzerbatch (class in anacore.hgvs)": [[0, "anacore.hgvs.MutalyzerBatch"]], "node (class in anacore.node)": [[0, "anacore.node.Node"]], "picardreader (class in anacore.picardio)": [[0, "anacore.picardIO.PicardReader"]], "proeval (class in anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.ProEval"]], "proio (class in anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.ProIO"]], "protein (class in anacore.genomicregion)": [[0, "anacore.genomicRegion.Protein"]], "required_sections (anacore.illumina.samplesheet.samplesheetv1 attribute)": [[0, "anacore.illumina.samplesheet.SampleSheetV1.REQUIRED_SECTIONS"]], "required_sections (anacore.illumina.samplesheet.samplesheetv2 attribute)": [[0, "anacore.illumina.samplesheet.SampleSheetV2.REQUIRED_SECTIONS"]], "rnaalphabet (class in anacore.sequence)": [[0, "anacore.sequence.RNAAlphabet"]], "rtacomplete (class in anacore.illumina.run)": [[0, "anacore.illumina.run.RTAComplete"]], "region (class in anacore.region)": [[0, "anacore.region.Region"]], "regionlist (class in anacore.region)": [[0, "anacore.region.RegionList"]], "regiontree (class in anacore.region)": [[0, "anacore.region.RegionTree"]], "reportio (class in anacore.msi.reportio)": [[0, "anacore.msi.reportIO.ReportIO"]], "run (class in anacore.illumina.run)": [[0, "anacore.illumina.run.Run"]], "runinfo (class in anacore.illumina.run)": [[0, "anacore.illumina.run.RunInfo"]], "runmutalyzerdescription (class in anacore.hgvs)": [[0, "anacore.hgvs.RunMutalyzerDescription"]], "runmutalyzerlegend (class in anacore.hgvs)": [[0, "anacore.hgvs.RunMutalyzerLegend"]], "runparameters (class in anacore.illumina.run)": [[0, "anacore.illumina.run.RunParameters"]], "samples_section (anacore.illumina.samplesheet.samplesheetv1 attribute)": [[0, "anacore.illumina.samplesheet.SampleSheetV1.SAMPLES_SECTION"]], "samples_section (anacore.illumina.samplesheet.samplesheetv2 attribute)": [[0, "anacore.illumina.samplesheet.SampleSheetV2.SAMPLES_SECTION"]], "starfusionio (class in anacore.fusion)": [[0, "anacore.fusion.STARFusionIO"]], "starlog (class in anacore.starlog)": [[0, "anacore.STARLog.STARLog"]], "svio (class in anacore.sv)": [[0, "anacore.sv.SVIO"]], "sample (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.Sample"]], "sampleads (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.SampleADS"]], "samplesheetads (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.SampleSheetADS"]], "samplesheetfactory (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.SampleSheetFactory"]], "samplesheetv1 (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.SampleSheetV1"]], "samplesheetv2 (class in anacore.illumina.samplesheet)": [[0, "anacore.illumina.samplesheet.SampleSheetV2"]], "sequence (class in anacore.sequence)": [[0, "anacore.sequence.Sequence"]], "sequencefilereader (class in anacore.sequenceio)": [[0, "anacore.sequenceIO.SequenceFileReader"]], "status (class in anacore.msi.base)": [[0, "anacore.msi.base.Status"]], "tophatfusionio (class in anacore.tophatfusion)": [[0, "anacore.tophatFusion.TopHatFusionIO"]], "transcript (class in anacore.genomicregion)": [[0, "anacore.genomicRegion.Transcript"]], "vcfio (class in anacore.vcf)": [[0, "anacore.vcf.VCFIO"]], "vcfrecord (class in anacore.vcf)": [[0, "anacore.vcf.VCFRecord"]], "vcfsymbaltrecord (class in anacore.vcf)": [[0, "anacore.vcf.VCFSymbAltRecord"]], "vepvcfio (class in anacore.annotvcf)": [[0, "anacore.annotVcf.VEPVCFIO"]], "aalength() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.aaLength"]], "aa_by_codon (anacore.sequence.codonalphabet attribute)": [[0, "anacore.sequence.CodonAlphabet.aa_by_codon"]], "addchild() (anacore.node.node method)": [[0, "anacore.node.Node.addChild"]], "addchild() (anacore.region.regiontree method)": [[0, "anacore.region.RegionTree.addChild"]], "addlocirestospl() (in module anacore.msi.annot)": [[0, "anacore.msi.annot.addLociResToSpl"]], "addlociresult() (in module anacore.msi.annot)": [[0, "anacore.msi.annot.addLociResult"]], "addlocus() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.addLocus"]], "addprotein() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.addProtein"]], "addtoannot() (anacore.gff.gff3record method)": [[0, "anacore.gff.GFF3Record.addToAnnot"]], "aggregator (anacore.filters.filter property)": [[0, "anacore.filters.Filter.aggregator"]], "anacore.starlog": [[0, "module-anacore.STARLog"]], "anacore.abstractfile": [[0, "module-anacore.abstractFile"]], "anacore.annotvcf": [[0, "module-anacore.annotVcf"]], "anacore.bed": [[0, "module-anacore.bed"]], "anacore.db": [[0, "module-anacore.db"]], "anacore.db.homo_sapiens": [[0, "module-anacore.db.homo_sapiens"]], "anacore.db.homo_sapiens.accession": [[0, "module-anacore.db.homo_sapiens.accession"]], "anacore.filters": [[0, "module-anacore.filters"]], "anacore.fusion": [[0, "module-anacore.fusion"]], "anacore.genomicregion": [[0, "module-anacore.genomicRegion"]], "anacore.getterpath": [[0, "module-anacore.getterPath"]], "anacore.gff": [[0, "module-anacore.gff"]], "anacore.gtf": [[0, "module-anacore.gtf"]], "anacore.hgvs": [[0, "module-anacore.hgvs"]], "anacore.illumina": [[0, "module-anacore.illumina"]], "anacore.illumina.base": [[0, "module-anacore.illumina.base"]], "anacore.illumina.demultiplex": [[0, "module-anacore.illumina.demultiplex"]], "anacore.illumina.demultiplex.base": [[0, "module-anacore.illumina.demultiplex.base"]], "anacore.illumina.demultiplex.bcl2fastq": [[0, "module-anacore.illumina.demultiplex.bcl2fastq"]], "anacore.illumina.demultiplex.bclconvert": [[0, "module-anacore.illumina.demultiplex.bclconvert"]], "anacore.illumina.run": [[0, "module-anacore.illumina.run"]], "anacore.illumina.samplesheet": [[0, "module-anacore.illumina.samplesheet"]], "anacore.maf": [[0, "module-anacore.maf"]], "anacore.matrix": [[0, "module-anacore.matrix"]], "anacore.msi": [[0, "module-anacore.msi"]], "anacore.msi.annot": [[0, "module-anacore.msi.annot"]], "anacore.msi.base": [[0, "module-anacore.msi.base"]], "anacore.msi.hubble": [[0, "module-anacore.msi.hubble"]], "anacore.msi.locus": [[0, "module-anacore.msi.locus"]], "anacore.msi.msings": [[0, "module-anacore.msi.msings"]], "anacore.msi.msisensorpro": [[0, "module-anacore.msi.msisensorpro"]], "anacore.msi.reportio": [[0, "module-anacore.msi.reportIO"]], "anacore.msi.sample": [[0, "module-anacore.msi.sample"]], "anacore.node": [[0, "module-anacore.node"]], "anacore.picardio": [[0, "module-anacore.picardIO"]], "anacore.region": [[0, "module-anacore.region"]], "anacore.sequence": [[0, "module-anacore.sequence"]], "anacore.sequenceio": [[0, "module-anacore.sequenceIO"]], "anacore.sv": [[0, "module-anacore.sv"]], "anacore.tophatfusion": [[0, "module-anacore.tophatFusion"]], "anacore.vcf": [[0, "module-anacore.vcf"]], "authorizedvalues() (anacore.msi.base.status static method)": [[0, "anacore.msi.base.Status.authorizedValues"]], "basename (anacore.illumina.samplesheet.sample property)": [[0, "anacore.illumina.samplesheet.Sample.basename"]], "basename (anacore.illumina.samplesheet.sampleads property)": [[0, "anacore.illumina.samplesheet.SampleADS.basename"]], "binarytostring() (in module anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.binaryToString"]], "castcol() (in module anacore.picardio)": [[0, "anacore.picardIO.castCol"]], "cleanedend() (anacore.illumina.samplesheet.abstractsamplesheet static method)": [[0, "anacore.illumina.samplesheet.AbstractSampleSheet.cleanedEnd"]], "close() (anacore.abstractfile.abstractfile method)": [[0, "anacore.abstractFile.AbstractFile.close"]], "close() (anacore.fusion.breakendvcfio method)": [[0, "anacore.fusion.BreakendVCFIO.close"]], "close() (anacore.sequenceio.fastaio method)": [[0, "anacore.sequenceIO.FastaIO.close"]], "close() (anacore.sequenceio.fastqio method)": [[0, "anacore.sequenceIO.FastqIO.close"]], "closed() (anacore.abstractfile.abstractfile method)": [[0, "anacore.abstractFile.AbstractFile.closed"]], "codons_by_aa (anacore.sequence.aa3lettersalphabet attribute)": [[0, "anacore.sequence.AA3LettersAlphabet.codons_by_aa"]], "complement (anacore.sequence.dnaalphabet attribute)": [[0, "anacore.sequence.DNAAlphabet.complement"]], "complement (anacore.sequence.rnaalphabet attribute)": [[0, "anacore.sequence.RNAAlphabet.complement"]], "consolidated() (in module anacore.region)": [[0, "anacore.region.consolidated"]], "contains() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.contains"]], "contains() (anacore.region.region method)": [[0, "anacore.region.Region.contains"]], "containsindel() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.containsIndel"]], "containsindel() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.containsIndel"]], "copyheader() (anacore.annotvcf.annotvcfio method)": [[0, "anacore.annotVcf.AnnotVCFIO.copyHeader"]], "copyheader() (anacore.vcf.vcfio method)": [[0, "anacore.vcf.VCFIO.copyHeader"]], "decodeinfovalue() (in module anacore.vcf)": [[0, "anacore.vcf.decodeInfoValue"]], "decodedvalue() (anacore.gff.gff3io static method)": [[0, "anacore.gff.GFF3IO.decodedValue"]], "delloci() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.delLoci"]], "dellocus() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.delLocus"]], "delprotein() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.delProtein"]], "delresult() (anacore.msi.locus.locus method)": [[0, "anacore.msi.locus.Locus.delResult"]], "dnarevcom() (anacore.sequence.sequence method)": [[0, "anacore.sequence.Sequence.dnaRevCom"]], "encodeinfovalue() (in module anacore.vcf)": [[0, "anacore.vcf.encodeInfoValue"]], "encodedvalue() (anacore.gff.gff3io static method)": [[0, "anacore.gff.GFF3IO.encodedValue"]], "end (anacore.msi.locus.locus property)": [[0, "anacore.msi.locus.Locus.end"]], "end (anacore.vcf.vcfsymbaltrecord property)": [[0, "anacore.vcf.VCFSymbAltRecord.end"]], "end_aa (anacore.hgvs.hgvsprotchange property)": [[0, "anacore.hgvs.HGVSProtChange.end_aa"]], "etreetodict() (in module anacore.illumina.run)": [[0, "anacore.illumina.run.etreeToDict"]], "eval() (anacore.filters.abstractatomicfilter method)": [[0, "anacore.filters.AbstractAtomicFilter.eval"]], "eval() (anacore.filters.emptyiterfilter method)": [[0, "anacore.filters.EmptyIterFilter.eval"]], "eval() (anacore.filters.filter method)": [[0, "anacore.filters.Filter.eval"]], "eval() (anacore.filters.filterscombiner method)": [[0, "anacore.filters.FiltersCombiner.eval"]], "factory() (anacore.fusion.fusionfilereader static method)": [[0, "anacore.fusion.FusionFileReader.factory"]], "factory() (anacore.sequenceio.sequencefilereader static method)": [[0, "anacore.sequenceIO.SequenceFileReader.factory"]], "fastdownstreamed() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.fastDownstreamed"]], "fastdownstreamed() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.fastDownstreamed"]], "faststandardize() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.fastStandardize"]], "faststandardize() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.fastStandardize"]], "filterpanels() (anacore.illumina.samplesheet.samplesheetads method)": [[0, "anacore.illumina.samplesheet.SampleSheetADS.filterPanels"]], "filtersfromdict() (in module anacore.filters)": [[0, "anacore.filters.filtersFromDict"]], "findsplfiles() (anacore.illumina.samplesheet.sampleads method)": [[0, "anacore.illumina.samplesheet.SampleADS.findSplFiles"]], "fit() (anacore.msi.base.locusclassifier method)": [[0, "anacore.msi.base.LocusClassifier.fit"]], "fromclusternode() (anacore.node.node static method)": [[0, "anacore.node.Node.fromClusterNode"]], "fromdense() (anacore.msi.locus.locusdatadistrib static method)": [[0, "anacore.msi.locus.LocusDataDistrib.fromDense"]], "fromdict() (anacore.filters.emptyiterfilter static method)": [[0, "anacore.filters.EmptyIterFilter.fromDict"]], "fromdict() (anacore.filters.filter static method)": [[0, "anacore.filters.Filter.fromDict"]], "fromdict() (anacore.getterpath.getterpath static method)": [[0, "anacore.getterPath.GetterPath.fromDict"]], "fromdict() (anacore.msi.locus.locus static method)": [[0, "anacore.msi.locus.Locus.fromDict"]], "fromdict() (anacore.msi.locus.locusres static method)": [[0, "anacore.msi.locus.LocusRes.fromDict"]], "fromdict() (anacore.msi.sample.msisample static method)": [[0, "anacore.msi.sample.MSISample.fromDict"]], "fromdict() (anacore.msi.sample.msisplres static method)": [[0, "anacore.msi.sample.MSISplRes.fromDict"]], "fromdict() (anacore.node.node static method)": [[0, "anacore.node.Node.fromDict"]], "frommodel() (anacore.msi.msisensorpro.baselinerecord static method)": [[0, "anacore.msi.msisensorpro.BaselineRecord.fromModel"]], "fromstdrecord() (anacore.vcf.vcfsymbaltrecord static method)": [[0, "anacore.vcf.VCFSymbAltRecord.fromStdRecord"]], "fromstr() (anacore.getterpath.getterpath static method)": [[0, "anacore.getterPath.GetterPath.fromStr"]], "fromstr() (anacore.hgvs.hgvs static method)": [[0, "anacore.hgvs.HGVS.fromStr"]], "fromstr() (anacore.hgvs.hgvsprotchange static method)": [[0, "anacore.hgvs.HGVSProtChange.fromStr"]], "fromstr() (anacore.region.region static method)": [[0, "anacore.region.Region.fromStr"]], "get() (anacore.fusion.breakendvcfio method)": [[0, "anacore.fusion.BreakendVCFIO.get"]], "get() (anacore.getterpath.getterpath method)": [[0, "anacore.getterPath.GetterPath.get"]], "get() (anacore.getterpath.iterabletoken method)": [[0, "anacore.getterPath.IterableToken.get"]], "get() (anacore.getterpath.keytoken method)": [[0, "anacore.getterPath.KeyToken.get"]], "get() (anacore.getterpath.methodtoken method)": [[0, "anacore.getterPath.MethodToken.get"]], "get() (anacore.illumina.demultiplex.demultstatfactory static method)": [[0, "anacore.illumina.demultiplex.DemultStatFactory.get"]], "get() (anacore.illumina.samplesheet.samplesheetfactory static method)": [[0, "anacore.illumina.samplesheet.SampleSheetFactory.get"]], "get() (anacore.sequenceio.idxfastaio method)": [[0, "anacore.sequenceIO.IdxFastaIO.get"]], "getad() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getAD"]], "getaf() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getAF"]], "getafbysample() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getAFBySample"]], "getallelerecord() (in module anacore.vcf)": [[0, "anacore.vcf.getAlleleRecord"]], "getaltad() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getAltAD"]], "getaltaf() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getAltAF"]], "getaltfromcoord() (in module anacore.fusion)": [[0, "anacore.fusion.getAltFromCoord"]], "getancestors() (anacore.node.node method)": [[0, "anacore.node.Node.getAncestors"]], "getareas() (in module anacore.bed)": [[0, "anacore.bed.getAreas"]], "getareasbychr() (in module anacore.bed)": [[0, "anacore.bed.getAreasByChr"]], "getbndinterval() (in module anacore.fusion)": [[0, "anacore.fusion.getBNDInterval"]], "getbyaccession() (anacore.hgvs.runmutalyzerdescription method)": [[0, "anacore.hgvs.RunMutalyzerDescription.getByAccession"]], "getcdsfromtranscript() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getCDSFromTranscript"]], "getcastedvalue() (in module anacore.msi.annot)": [[0, "anacore.msi.annot.getCastedValue"]], "getchildbyname() (anacore.node.node method)": [[0, "anacore.node.Node.getChildByName"]], "getcodoninfo() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getCodonInfo"]], "getcodonrefpos() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getCodonRefPos"]], "getcodonseqfromprotpos() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getCodonSeqFromProtPos"]], "getcoltype() (in module anacore.picardio)": [[0, "anacore.picardIO.getColType"]], "getcontainers() (anacore.region.regionlist method)": [[0, "anacore.region.RegionList.getContainers"]], "getcoorddictfromcoordstr() (in module anacore.fusion)": [[0, "anacore.fusion.getCoordDictFromCoordStr"]], "getcoordstr() (in module anacore.fusion)": [[0, "anacore.fusion.getCoordStr"]], "getcoordinatesstr() (anacore.region.region method)": [[0, "anacore.region.Region.getCoordinatesStr"]], "getcount() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getCount"]], "getdp() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getDP"]], "getdensecount() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getDenseCount"]], "getdenseprct() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getDensePrct"]], "getdepth() (anacore.node.node method)": [[0, "anacore.node.Node.getDepth"]], "getdescendants() (anacore.node.node method)": [[0, "anacore.node.Node.getDescendants"]], "getemptyallelemarker() (anacore.vcf.vcfrecord static method)": [[0, "anacore.vcf.VCFRecord.getEmptyAlleleMarker"]], "getexectime() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.getExecTime"]], "getfreqmatrix() (in module anacore.vcf)": [[0, "anacore.vcf.getFreqMatrix"]], "getheaderattr() (in module anacore.vcf)": [[0, "anacore.vcf.getHeaderAttr"]], "getidbyname() (anacore.hgvs.runmutalyzerlegend method)": [[0, "anacore.hgvs.RunMutalyzerLegend.getIdByName"]], "getilluminaname() (in module anacore.illumina.base)": [[0, "anacore.illumina.base.getIlluminaName"]], "getincompletemodels() (anacore.msi.reportio.reportio static method)": [[0, "anacore.msi.reportIO.ReportIO.getIncompleteModels"]], "getinffromseqdesc() (in module anacore.illumina.base)": [[0, "anacore.illumina.base.getInfFromSeqDesc"]], "getinffromseqid() (in module anacore.illumina.base)": [[0, "anacore.illumina.base.getInfFromSeqID"]], "getleaves() (anacore.node.node method)": [[0, "anacore.node.Node.getLeaves"]], "getlibnamefromreadspath() (in module anacore.illumina.base)": [[0, "anacore.illumina.base.getLibNameFromReadsPath"]], "getlocimethods() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getLociMethods"]], "getlocusannotdict() (in module anacore.msi.annot)": [[0, "anacore.msi.annot.getLocusAnnotDict"]], "getmaxlength() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getMaxLength"]], "getmaxnbcol() (anacore.bed.bedio static method)": [[0, "anacore.bed.BEDIO.getMaxNbCol"]], "getmindist() (anacore.region.region method)": [[0, "anacore.region.Region.getMinDist"]], "getminlength() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getMinLength"]], "getmostdownstream() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getMostDownstream"]], "getmostdownstream() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.getMostDownstream"]], "getmostrepresented() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getMostRepresented"]], "getmostupstream() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getMostUpstream"]], "getmostupstream() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.getMostUpstream"]], "getname() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getName"]], "getname() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.getName"]], "getname() (in module anacore.maf)": [[0, "anacore.maf.getName"]], "getnbdetermined() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getNbDetermined"]], "getnbloci() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getNbLoci"]], "getnbpeaks() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.getNbPeaks"]], "getnbpeaks() (anacore.msi.msings.msingseval static method)": [[0, "anacore.msi.msings.MSINGSEval.getNbPeaks"]], "getnbprocessed() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getNbProcessed"]], "getnbstable() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getNbStable"]], "getnbsupporting() (in module anacore.msi.base)": [[0, "anacore.msi.base.getNbSupporting"]], "getnbundetermined() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getNbUndetermined"]], "getnbunstable() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.getNbUnstable"]], "getnearests() (anacore.region.regionlist method)": [[0, "anacore.region.RegionList.getNearests"]], "getntposfromrefpos() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getNtPosFromRefPos"]], "getntposfromregionpos() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getNtPosFromRegionPos"]], "getoverlapped() (anacore.region.regionlist method)": [[0, "anacore.region.RegionList.getOverlapped"]], "getparsedresponse() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.getParsedResponse"]], "getplatformfromserialnumber() (in module anacore.illumina.base)": [[0, "anacore.illumina.base.getPlatformFromSerialNumber"]], "getpopaltad() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getPopAltAD"]], "getpopaltaf() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getPopAltAF"]], "getpopdp() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getPopDP"]], "getpoprefad() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getPopRefAD"]], "getpoprefaf() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.getPopRefAF"]], "getposonref() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getPosOnRef"]], "getposonref() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.getPosOnRef"]], "getposonref() (anacore.region.region method)": [[0, "anacore.region.Region.getPosOnRef"]], "getposonregion() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getPosOnRegion"]], "getposonregion() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.getPosOnRegion"]], "getposonregion() (anacore.region.region method)": [[0, "anacore.region.Region.getPosOnRegion"]], "getprotbytr() (anacore.hgvs.runmutalyzerlegend method)": [[0, "anacore.hgvs.RunMutalyzerLegend.getProtBytr"]], "getrefseqinfo() (in module anacore.msi.locus)": [[0, "anacore.msi.locus.getRefSeqInfo"]], "getrequesturl() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.getRequestURL"]], "getresponse() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.getResponse"]], "getrunfolderinfo() (in module anacore.illumina.run)": [[0, "anacore.illumina.run.getRunFolderInfo"]], "getshortestrepeatunit() (in module anacore.sequence)": [[0, "anacore.sequence.getShortestRepeatUnit"]], "getslippagescores() (anacore.msi.msisensorpro.proeval static method)": [[0, "anacore.msi.msisensorpro.ProEval.getSlippageScores"]], "getsortedareasbychr() (in module anacore.bed)": [[0, "anacore.bed.getSortedAreasByChr"]], "getstrand() (in module anacore.fusion)": [[0, "anacore.fusion.getStrand"]], "getstrandedseqfrompos() (in module anacore.sequenceio)": [[0, "anacore.sequenceIO.getStrandedSeqFromPos"]], "getsub() (anacore.fusion.breakendvcfio method)": [[0, "anacore.fusion.BreakendVCFIO.getSub"]], "getsub() (anacore.sequenceio.idxfastaio method)": [[0, "anacore.sequenceIO.IdxFastaIO.getSub"]], "getsub() (anacore.vcf.vcfio method)": [[0, "anacore.vcf.VCFIO.getSub"]], "getsubfromrefpos() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.getSubFromRefPos"]], "getsubfromregionpos() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.getSubFromRegionPos"]], "getsubfromregionpos() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.getSubFromRegionPos"]], "getthreshold() (anacore.msi.msisensorpro.proeval static method)": [[0, "anacore.msi.msisensorpro.ProEval.getThreshold"]], "getthresholdfromnbpeaks() (anacore.msi.msings.msingseval static method)": [[0, "anacore.msi.msings.MSINGSEval.getThresholdFromNbPeaks"]], "getthresholdfromscores() (anacore.msi.msisensorpro.proeval static method)": [[0, "anacore.msi.msisensorpro.ProEval.getThresholdFromScores"]], "getvaltype() (in module anacore.picardio)": [[0, "anacore.picardIO.getValType"]], "haschild() (anacore.node.node method)": [[0, "anacore.node.Node.hasChild"]], "hasoverlap() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.hasOverlap"]], "hasoverlap() (anacore.region.region method)": [[0, "anacore.region.Region.hasOverlap"]], "hasstrandedoverlap() (anacore.region.region method)": [[0, "anacore.region.Region.hasStrandedOverlap"]], "inscouldbeidentical() (anacore.hgvs.hgvsprotchange static method)": [[0, "anacore.hgvs.HGVSProtChange.insCouldBeIdentical"]], "iscopied() (anacore.illumina.run.run method)": [[0, "anacore.illumina.run.Run.isCopied"]], "isdeletion() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.isDeletion"]], "isdeletion() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.isDeletion"]], "isempty() (in module anacore.abstractfile)": [[0, "anacore.abstractFile.isEmpty"]], "isended() (anacore.illumina.run.run method)": [[0, "anacore.illumina.run.Run.isEnded"]], "isgzip() (in module anacore.abstractfile)": [[0, "anacore.abstractFile.isGzip"]], "isinframeins() (anacore.hgvs.hgvsprotchange method)": [[0, "anacore.hgvs.HGVSProtChange.isInFrameIns"]], "isindel() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.isIndel"]], "isindel() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.isIndel"]], "isinsanddel() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.isInsAndDel"]], "isinsanddel() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.isInsAndDel"]], "isinsertion() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.isInsertion"]], "isinsertion() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.isInsertion"]], "isinversion() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.isInversion"]], "ispredicted() (anacore.hgvs.hgvs method)": [[0, "anacore.hgvs.HGVS.isPredicted"]], "isrecordline() (anacore.abstractfile.abstractfile method)": [[0, "anacore.abstractFile.AbstractFile.isRecordLine"]], "isrecordline() (anacore.bed.bedio method)": [[0, "anacore.bed.BEDIO.isRecordLine"]], "isrecordline() (anacore.fusion.breakendvcfio method)": [[0, "anacore.fusion.BreakendVCFIO.isRecordLine"]], "isrecordline() (anacore.gff.gff3io method)": [[0, "anacore.gff.GFF3IO.isRecordLine"]], "isrecordline() (anacore.gtf.gtfio method)": [[0, "anacore.gtf.GTFIO.isRecordLine"]], "isrecordline() (anacore.maf.mafio method)": [[0, "anacore.maf.MAFIO.isRecordLine"]], "isrecordline() (anacore.sv.svio method)": [[0, "anacore.sv.SVIO.isRecordLine"]], "isrecordline() (anacore.vcf.vcfio method)": [[0, "anacore.vcf.VCFIO.isRecordLine"]], "issequenced() (anacore.illumina.run.run method)": [[0, "anacore.illumina.run.Run.isSequenced"]], "isvalid() (anacore.bed.bedio static method)": [[0, "anacore.bed.BEDIO.isValid"]], "isvalid() (anacore.fusion.arribaio static method)": [[0, "anacore.fusion.ArribaIO.isValid"]], "isvalid() (anacore.fusion.breakendvcfio static method)": [[0, "anacore.fusion.BreakendVCFIO.isValid"]], "isvalid() (anacore.fusion.fusioncatcherio static method)": [[0, "anacore.fusion.FusionCatcherIO.isValid"]], "isvalid() (anacore.fusion.starfusionio static method)": [[0, "anacore.fusion.STARFusionIO.isValid"]], "isvalid() (anacore.hgvs.hgvsprotchange static method)": [[0, "anacore.hgvs.HGVSProtChange.isValid"]], "isvalid() (anacore.illumina.samplesheet.samplesheetads static method)": [[0, "anacore.illumina.samplesheet.SampleSheetADS.isValid"]], "isvalid() (anacore.illumina.samplesheet.samplesheetv1 static method)": [[0, "anacore.illumina.samplesheet.SampleSheetV1.isValid"]], "isvalid() (anacore.illumina.samplesheet.samplesheetv2 static method)": [[0, "anacore.illumina.samplesheet.SampleSheetV2.isValid"]], "isvalid() (anacore.sequence.aa3lettersalphabet static method)": [[0, "anacore.sequence.AA3LettersAlphabet.isValid"]], "isvalid() (anacore.sequence.codonalphabet static method)": [[0, "anacore.sequence.CodonAlphabet.isValid"]], "isvalid() (anacore.sequence.dnaalphabet static method)": [[0, "anacore.sequence.DNAAlphabet.isValid"]], "isvalid() (anacore.sequence.rnaalphabet static method)": [[0, "anacore.sequence.RNAAlphabet.isValid"]], "isvalid() (anacore.sequenceio.fastaio static method)": [[0, "anacore.sequenceIO.FastaIO.isValid"]], "isvalid() (anacore.sequenceio.fastqio static method)": [[0, "anacore.sequenceIO.FastqIO.isValid"]], "isvalid() (anacore.sv.svio static method)": [[0, "anacore.sv.SVIO.isValid"]], "items() (anacore.msi.locus.locusdatadistrib method)": [[0, "anacore.msi.locus.LocusDataDistrib.items"]], "items() (anacore.vcf.headerattr method)": [[0, "anacore.vcf.HeaderAttr.items"]], "iteroverlapped() (in module anacore.region)": [[0, "anacore.region.iterOverlapped"]], "iteroverlappedbyregion() (in module anacore.region)": [[0, "anacore.region.iterOverlappedByRegion"]], "keys() (anacore.vcf.headerattr method)": [[0, "anacore.vcf.HeaderAttr.keys"]], "left_flank_binary (anacore.msi.msisensorpro.baselinerecord property)": [[0, "anacore.msi.msisensorpro.BaselineRecord.left_flank_binary"]], "length (anacore.msi.locus.locus property)": [[0, "anacore.msi.locus.Locus.length"]], "length() (anacore.genomicregion.protein method)": [[0, "anacore.genomicRegion.Protein.length"]], "length() (anacore.genomicregion.transcript method)": [[0, "anacore.genomicRegion.Transcript.length"]], "length() (anacore.region.region method)": [[0, "anacore.region.Region.length"]], "library_basename (anacore.illumina.samplesheet.sample property)": [[0, "anacore.illumina.samplesheet.Sample.library_basename"]], "loadindex() (anacore.fusion.breakendvcfio method)": [[0, "anacore.fusion.BreakendVCFIO.loadIndex"]], "loadmodel() (in module anacore.gtf)": [[0, "anacore.gtf.loadModel"]], "loci() (anacore.msi.msisensorpro.proio static method)": [[0, "anacore.msi.msisensorpro.ProIO.loci"]], "mergedregion() (in module anacore.region)": [[0, "anacore.region.mergedRegion"]], "module": [[0, "module-anacore.STARLog"], [0, "module-anacore.abstractFile"], [0, "module-anacore.annotVcf"], [0, "module-anacore.bed"], [0, "module-anacore.db"], [0, "module-anacore.db.homo_sapiens"], [0, "module-anacore.db.homo_sapiens.accession"], [0, "module-anacore.filters"], [0, "module-anacore.fusion"], [0, "module-anacore.genomicRegion"], [0, "module-anacore.getterPath"], [0, "module-anacore.gff"], [0, "module-anacore.gtf"], [0, "module-anacore.hgvs"], [0, "module-anacore.illumina"], [0, "module-anacore.illumina.base"], [0, "module-anacore.illumina.demultiplex"], [0, "module-anacore.illumina.demultiplex.base"], [0, "module-anacore.illumina.demultiplex.bcl2fastq"], [0, "module-anacore.illumina.demultiplex.bclconvert"], [0, "module-anacore.illumina.run"], [0, "module-anacore.illumina.samplesheet"], [0, "module-anacore.maf"], [0, "module-anacore.matrix"], [0, "module-anacore.msi"], [0, "module-anacore.msi.annot"], [0, "module-anacore.msi.base"], [0, "module-anacore.msi.hubble"], [0, "module-anacore.msi.locus"], [0, "module-anacore.msi.msings"], [0, "module-anacore.msi.msisensorpro"], [0, "module-anacore.msi.reportIO"], [0, "module-anacore.msi.sample"], [0, "module-anacore.node"], [0, "module-anacore.picardIO"], [0, "module-anacore.region"], [0, "module-anacore.sequence"], [0, "module-anacore.sequenceIO"], [0, "module-anacore.sv"], [0, "module-anacore.tophatFusion"], [0, "module-anacore.vcf"]], "nbrepeatfromstart() (in module anacore.msi.hubble)": [[0, "anacore.msi.hubble.nbRepeatFromStart"]], "nbseq() (anacore.sequenceio.fastaio static method)": [[0, "anacore.sequenceIO.FastaIO.nbSeq"]], "nbseq() (anacore.sequenceio.fastqio static method)": [[0, "anacore.sequenceIO.FastqIO.nbSeq"]], "nbseqandnt() (anacore.sequenceio.fastaio static method)": [[0, "anacore.sequenceIO.FastaIO.nbSeqAndNt"]], "nbseqandnt() (anacore.sequenceio.fastqio static method)": [[0, "anacore.sequenceIO.FastqIO.nbSeqAndNt"]], "new_aa (anacore.hgvs.hgvsprotchange property)": [[0, "anacore.hgvs.HGVSProtChange.new_aa"]], "nextlocus() (anacore.msi.msisensorpro.distio method)": [[0, "anacore.msi.msisensorpro.DistIO.nextLocus"]], "nextseq() (anacore.sequenceio.fastaio method)": [[0, "anacore.sequenceIO.FastaIO.nextSeq"]], "nextseq() (anacore.sequenceio.fastqio method)": [[0, "anacore.sequenceIO.FastqIO.nextSeq"]], "none (anacore.msi.base.status attribute)": [[0, "anacore.msi.base.Status.none"]], "normalizesingleallele() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.normalizeSingleAllele"]], "normalizesingleallele() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.normalizeSingleAllele"]], "one_by_three (anacore.sequence.aa3lettersalphabet attribute)": [[0, "anacore.sequence.AA3LettersAlphabet.one_by_three"]], "operator (anacore.filters.filter property)": [[0, "anacore.filters.Filter.operator"]], "parameters (anacore.illumina.run.run property)": [[0, "anacore.illumina.run.Run.parameters"]], "parse() (anacore.msi.reportio.reportio static method)": [[0, "anacore.msi.reportIO.ReportIO.parse"]], "parsehubbleresults() (in module anacore.msi.hubble)": [[0, "anacore.msi.hubble.parseHubbleResults"]], "parseproresults() (in module anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.parseProResults"]], "predict() (anacore.msi.base.locusclassifier method)": [[0, "anacore.msi.base.LocusClassifier.predict"]], "predict_proba() (anacore.msi.base.locusclassifier method)": [[0, "anacore.msi.base.LocusClassifier.predict_proba"]], "proteins (anacore.genomicregion.transcript property)": [[0, "anacore.genomicRegion.Transcript.proteins"]], "qualoffset() (anacore.sequenceio.fastqio static method)": [[0, "anacore.sequenceIO.FastqIO.qualOffset"]], "read() (anacore.abstractfile.abstractfile method)": [[0, "anacore.abstractFile.AbstractFile.read"]], "readbyid() (anacore.sequenceio.faidx method)": [[0, "anacore.sequenceIO.Faidx.readById"]], "recfromregion() (anacore.bed.bedrecord static method)": [[0, "anacore.bed.BEDRecord.recFromRegion"]], "rectovcfline() (anacore.annotvcf.annotvcfio method)": [[0, "anacore.annotVcf.AnnotVCFIO.recToVCFLine"]], "rectovcfline() (anacore.vcf.vcfio method)": [[0, "anacore.vcf.VCFIO.recToVCFLine"]], "recordtoline() (anacore.gtf.gtfio method)": [[0, "anacore.gtf.GTFIO.recordToLine"]], "recordtoline() (anacore.msi.annot.msiannot method)": [[0, "anacore.msi.annot.MSIAnnot.recordToLine"]], "recordtoline() (anacore.msi.hubble.hubblediff method)": [[0, "anacore.msi.hubble.HubbleDiff.recordToLine"]], "recordtoline() (anacore.msi.hubble.hubbledist method)": [[0, "anacore.msi.hubble.HubbleDist.recordToLine"]], "recordtoline() (anacore.msi.msings.msingsanalysisio method)": [[0, "anacore.msi.msings.MSINGSAnalysisIO.recordToLine"]], "recordtoline() (anacore.msi.msisensorpro.baselineio method)": [[0, "anacore.msi.msisensorpro.BaselineIO.recordToLine"]], "recordtoline() (anacore.msi.msisensorpro.proio method)": [[0, "anacore.msi.msisensorpro.ProIO.recordToLine"]], "recordtoline() (anacore.sv.hashedsvio method)": [[0, "anacore.sv.HashedSVIO.recordToLine"]], "recordtoline() (anacore.sv.svio method)": [[0, "anacore.sv.SVIO.recordToLine"]], "recordtolines() (anacore.msi.msisensorpro.distio method)": [[0, "anacore.msi.msisensorpro.DistIO.recordToLines"]], "refend() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.refEnd"]], "refend() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.refEnd"]], "refstart() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.refStart"]], "refstart() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.refStart"]], "repeat_unit_binary (anacore.msi.msisensorpro.baselinerecord property)": [[0, "anacore.msi.msisensorpro.BaselineRecord.repeat_unit_binary"]], "repeat_unit_length (anacore.msi.msisensorpro.baselinerecord property)": [[0, "anacore.msi.msisensorpro.BaselineRecord.repeat_unit_length"]], "revcom() (anacore.sequence.dnaalphabet static method)": [[0, "anacore.sequence.DNAAlphabet.revCom"]], "revcom() (anacore.sequence.rnaalphabet static method)": [[0, "anacore.sequence.RNAAlphabet.revCom"]], "right_flank_binary (anacore.msi.msisensorpro.baselinerecord property)": [[0, "anacore.msi.msisensorpro.BaselineRecord.right_flank_binary"]], "rnarevcom() (anacore.sequence.sequence method)": [[0, "anacore.sequence.Sequence.rnaRevCom"]], "samplescounts() (anacore.illumina.demultiplex.base.abstractdemultstat method)": [[0, "anacore.illumina.demultiplex.base.AbstractDemultStat.samplesCounts"]], "seqtofastaline() (anacore.sequenceio.fastaio method)": [[0, "anacore.sequenceIO.FastaIO.seqToFastaLine"]], "seqtofastqline() (anacore.sequenceio.fastqio method)": [[0, "anacore.sequenceIO.FastqIO.seqToFastqLine"]], "setfct() (anacore.filters.filter method)": [[0, "anacore.filters.Filter.setFct"]], "setgetter() (anacore.filters.abstractatomicfilter method)": [[0, "anacore.filters.AbstractAtomicFilter.setGetter"]], "setreference() (anacore.region.region method)": [[0, "anacore.region.Region.setReference"]], "setscore() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.setScore"]], "setsplfiles() (anacore.illumina.samplesheet.sampleads method)": [[0, "anacore.illumina.samplesheet.SampleADS.setSplFiles"]], "setstatusbyinstabilityratio() (anacore.msi.sample.msisample method)": [[0, "anacore.msi.sample.MSISample.setStatusByInstabilityRatio"]], "setvcfheader() (anacore.fusion.arribaio static method)": [[0, "anacore.fusion.ArribaIO.setVCFHeader"]], "setvcfheader() (anacore.fusion.fusioncatcherio static method)": [[0, "anacore.fusion.FusionCatcherIO.setVCFHeader"]], "setvcfheader() (anacore.fusion.starfusionio static method)": [[0, "anacore.fusion.STARFusionIO.setVCFHeader"]], "set_status() (anacore.msi.base.locusclassifier method)": [[0, "anacore.msi.base.LocusClassifier.set_status"]], "sortchildren() (anacore.region.regiontree method)": [[0, "anacore.region.RegionTree.sortChildren"]], "splittedbyref() (in module anacore.region)": [[0, "anacore.region.splittedByRef"]], "stable (anacore.msi.base.status attribute)": [[0, "anacore.msi.base.Status.stable"]], "start (anacore.msi.locus.locus property)": [[0, "anacore.msi.locus.Locus.start"]], "start_aa (anacore.hgvs.hgvsprotchange property)": [[0, "anacore.hgvs.HGVSProtChange.start_aa"]], "strandedcontains() (anacore.region.region method)": [[0, "anacore.region.Region.strandedContains"]], "stringtobinary() (in module anacore.msi.msisensorpro)": [[0, "anacore.msi.msisensorpro.stringToBinary"]], "submit() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.submit"]], "sv_len (anacore.vcf.vcfsymbaltrecord property)": [[0, "anacore.vcf.VCFSymbAltRecord.sv_len"]], "syncrequest() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.syncRequest"]], "titles (anacore.fusion.arribaio attribute)": [[0, "anacore.fusion.ArribaIO.titles"]], "titles (anacore.fusion.fusioncatcherio attribute)": [[0, "anacore.fusion.FusionCatcherIO.titles"]], "titles (anacore.fusion.starfusionio attribute)": [[0, "anacore.fusion.STARFusionIO.titles"]], "todict() (anacore.filters.abstractatomicfilter method)": [[0, "anacore.filters.AbstractAtomicFilter.toDict"]], "todict() (anacore.getterpath.getterpath method)": [[0, "anacore.getterPath.GetterPath.toDict"]], "todict() (anacore.getterpath.iterabletoken method)": [[0, "anacore.getterPath.IterableToken.toDict"]], "todict() (anacore.getterpath.keytoken method)": [[0, "anacore.getterPath.KeyToken.toDict"]], "todict() (anacore.getterpath.methodtoken method)": [[0, "anacore.getterPath.MethodToken.toDict"]], "todict() (anacore.illumina.samplesheet.sample method)": [[0, "anacore.illumina.samplesheet.Sample.toDict"]], "todict() (anacore.node.node method)": [[0, "anacore.node.Node.toDict"]], "todict() (in module anacore.msi.base)": [[0, "anacore.msi.base.toDict"]], "toextendednewick() (anacore.node.node method)": [[0, "anacore.node.Node.toExtendedNewick"]], "togff() (anacore.gff.gff3record method)": [[0, "anacore.gff.GFF3Record.toGff"]], "tohumanname() (anacore.db.homo_sapiens.accession.assemblyaccession static method)": [[0, "anacore.db.homo_sapiens.accession.AssemblyAccession.toHumanName"]], "tohumanname() (anacore.db.homo_sapiens.accession.chraccession static method)": [[0, "anacore.db.homo_sapiens.accession.ChrAccession.toHumanName"]], "tonewick() (anacore.node.node method)": [[0, "anacore.node.Node.toNewick"]], "tooneletter() (anacore.sequence.aa3lettersalphabet static method)": [[0, "anacore.sequence.AA3LettersAlphabet.toOneLetter"]], "tostr() (anacore.getterpath.getterpath method)": [[0, "anacore.getterPath.GetterPath.toStr"]], "tostr() (anacore.getterpath.iterabletoken method)": [[0, "anacore.getterPath.IterableToken.toStr"]], "tostr() (anacore.getterpath.keytoken method)": [[0, "anacore.getterPath.KeyToken.toStr"]], "tostr() (anacore.getterpath.methodtoken method)": [[0, "anacore.getterPath.MethodToken.toStr"]], "transcript (anacore.genomicregion.protein property)": [[0, "anacore.genomicRegion.Protein.transcript"]], "translate() (anacore.sequence.codonalphabet static method)": [[0, "anacore.sequence.CodonAlphabet.translate"]], "type() (anacore.vcf.vcfrecord method)": [[0, "anacore.vcf.VCFRecord.type"]], "type() (anacore.vcf.vcfsymbaltrecord method)": [[0, "anacore.vcf.VCFSymbAltRecord.type"]], "undetermined (anacore.illumina.demultiplex.base.abstractdemultstat property)": [[0, "anacore.illumina.demultiplex.base.AbstractDemultStat.undetermined"]], "undetermined (anacore.illumina.demultiplex.bcl2fastq.demultstat property)": [[0, "anacore.illumina.demultiplex.bcl2fastq.DemultStat.undetermined"]], "undetermined (anacore.illumina.demultiplex.bclconvert.demultstat property)": [[0, "anacore.illumina.demultiplex.bclconvert.DemultStat.undetermined"]], "undetermined (anacore.msi.base.status attribute)": [[0, "anacore.msi.base.Status.undetermined"]], "undeterminedcounts() (anacore.illumina.demultiplex.base.abstractdemultstat method)": [[0, "anacore.illumina.demultiplex.base.AbstractDemultStat.undeterminedCounts"]], "unexpectedbarcodes() (anacore.illumina.demultiplex.base.abstractdemultstat method)": [[0, "anacore.illumina.demultiplex.base.AbstractDemultStat.unexpectedBarcodes"]], "unstable (anacore.msi.base.status attribute)": [[0, "anacore.msi.base.Status.unstable"]], "updatestatus() (anacore.hgvs.mutalyzerbatch method)": [[0, "anacore.hgvs.MutalyzerBatch.updateStatus"]], "values (anacore.filters.filter property)": [[0, "anacore.filters.Filter.values"]], "words (anacore.sequence.aa3lettersalphabet attribute)": [[0, "anacore.sequence.AA3LettersAlphabet.words"]], "words (anacore.sequence.codonalphabet attribute)": [[0, "anacore.sequence.CodonAlphabet.words"]], "words (anacore.sequence.dnaalphabet attribute)": [[0, "anacore.sequence.DNAAlphabet.words"]], "words (anacore.sequence.rnaalphabet attribute)": [[0, "anacore.sequence.RNAAlphabet.words"]], "write() (anacore.bed.bedio method)": [[0, "anacore.bed.BEDIO.write"]], "write() (anacore.fusion.arribaio method)": [[0, "anacore.fusion.ArribaIO.write"]], "write() (anacore.fusion.breakendvcfio method)": [[0, "anacore.fusion.BreakendVCFIO.write"]], "write() (anacore.fusion.fusioncatcherio method)": [[0, "anacore.fusion.FusionCatcherIO.write"]], "write() (anacore.fusion.starfusionio method)": [[0, "anacore.fusion.STARFusionIO.write"]], "write() (anacore.gff.gff3io method)": [[0, "anacore.gff.GFF3IO.write"]], "write() (anacore.gtf.gtfio method)": [[0, "anacore.gtf.GTFIO.write"]], "write() (anacore.maf.mafio method)": [[0, "anacore.maf.MAFIO.write"]], "write() (anacore.msi.msisensorpro.distio method)": [[0, "anacore.msi.msisensorpro.DistIO.write"]], "write() (anacore.msi.msisensorpro.proio method)": [[0, "anacore.msi.msisensorpro.ProIO.write"]], "write() (anacore.msi.reportio.reportio static method)": [[0, "anacore.msi.reportIO.ReportIO.write"]], "write() (anacore.sequenceio.fastaio method)": [[0, "anacore.sequenceIO.FastaIO.write"]], "write() (anacore.sequenceio.fastqio method)": [[0, "anacore.sequenceIO.FastqIO.write"]], "write() (anacore.sv.svio method)": [[0, "anacore.sv.SVIO.write"]], "write() (anacore.vcf.vcfio method)": [[0, "anacore.vcf.VCFIO.write"]], "writeheader() (anacore.annotvcf.annotvcfio method)": [[0, "anacore.annotVcf.AnnotVCFIO.writeHeader"]], "writeheader() (anacore.sv.svio method)": [[0, "anacore.sv.SVIO.writeHeader"]], "writeheader() (anacore.vcf.vcfio method)": [[0, "anacore.vcf.VCFIO.writeHeader"]]}})